<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Map range coordinates between transcripts and genome space</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for mapToTranscripts {OrganismDbi}"><tr><td>mapToTranscripts {OrganismDbi}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Map range coordinates between transcripts and genome space</h2>

<h3>Description</h3>

<p>Map range coordinates between features in the transcriptome and
genome (reference) space.
</p>
<p>See <code>?<a href="../../GenomicAlignments/html/coordinate-mapping-methods.html">mapToAlignments</a></code> in the 
<span class="pkg">GenomicAlignments</span> package for mapping coordinates between 
reads (local) and genome (reference) space using a CIGAR alignment.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'ANY,MultiDb'
mapToTranscripts(x, transcripts,
          ignore.strand = TRUE,
          extractor.fun = GenomicFeatures::transcripts, ...) 
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p><code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges-class</a></code> object of positions to be mapped.
<code>x</code> must have names when mapping to the genome.
</p>
</td></tr>
<tr valign="top"><td><code>transcripts</code></td>
<td>

<p>The <code>OrganismDb</code> object that will be used to extract features
using the <code>extractor.fun</code>.
</p>
</td></tr>
<tr valign="top"><td><code>ignore.strand</code></td>
<td>

<p>When TRUE, strand is ignored in overlap operations.
</p>
</td></tr>
<tr valign="top"><td><code>extractor.fun</code></td>
<td>

<p>Function to extract genomic features from a <code>TxDb</code> object.
</p>
<p>Valid <code>extractor</code> functions:
</p>

<ul>
<li><p>transcripts  ## default
</p>
</li>
<li><p>exons
</p>
</li>
<li><p>cds
</p>
</li>
<li><p>genes
</p>
</li>
<li><p>promoters
</p>
</li>
<li><p>disjointExons
</p>
</li>
<li><p>microRNAs
</p>
</li>
<li><p>tRNAs
</p>
</li>
<li><p>transcriptsBy
</p>
</li>
<li><p>exonsBy
</p>
</li>
<li><p>cdsBy
</p>
</li>
<li><p>intronsByTranscript
</p>
</li>
<li><p>fiveUTRsByTranscript
</p>
</li>
<li><p>threeUTRsByTranscript
</p>
</li></ul>
    
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments passed to <code>extractor.fun</code> functions. 
</p>
</td></tr>
</table>


<h3>Details</h3>


<ul>
<li><p><code>mapToTranscripts</code>
The genomic range in <code>x</code> is mapped to the local position in the
<code>transcripts</code> ranges. A successful mapping occurs when <code>x</code>
is completely within the <code>transcripts</code> range, equivalent to: 
</p>
<pre>
      findOverlaps(..., type="within")
      </pre>
<p>Transcriptome-based coordinates start counting at 1 at the beginning
of the <code>transcripts</code> range and return positions where <code>x</code>
was aligned. The seqlevels of the return object are taken from the
<code>transcripts</code> object and should be transcript names. In this
direction, mapping is attempted between all elements of <code>x</code> and
all elements of <code>transcripts</code>.

</p>
</li></ul>



<h3>Value</h3>

<p>An object the same class as <code>x</code>.
</p>
<p>Parallel methods return an object the same shape as <code>x</code>. Ranges that
cannot be mapped (out of bounds or strand mismatch) are returned as 
zero-width ranges starting at 0 with a seqname of &quot;UNMAPPED&quot;.
</p>
<p>Non-parallel methods return an object that varies in length similar to a
Hits object. The result only contains mapped records, strand mismatch
and out of bound ranges are not returned. <code>xHits</code> and 
<code>transcriptsHits</code> metadata columns indicate the elements of <code>x</code> 
and <code>transcripts</code> used in the mapping.
</p>
<p>When present, names from <code>x</code> are propagated to the output. When
mapping to transcript coordinates, seqlevels of the output are the names 
on the <code>transcripts</code> object; most often these will be transcript
names. When mapping to the genome, seqlevels of the output are the seqlevels 
of <code>transcripts</code> which are usually chromosome names.
</p>


<h3>Author(s)</h3>

<p>V. Obenchain, M. Lawrence and H. Pag√®s; ported to work with
OrganismDbi by Marc Carlson</p>


<h3>See Also</h3>


<ul>
<li><p><a href="../../GenomicFeatures/html/coordinate-mapping-methods.html">mapToTranscripts</a>.
</p>
</li></ul>



<h3>Examples</h3>

<pre>

## ---------------------------------------------------------------------
## A. Basic Use
## ---------------------------------------------------------------------

library(Homo.sapiens)
x &lt;- GRanges("chr5", 
             IRanges(c(173315331,174151575), width=400, 
                     names=LETTERS[1:2])) 

## Map to transcript coordinates:
mapToTranscripts(x, Homo.sapiens)
</pre>

<hr /><div style="text-align: center;">[Package <em>OrganismDbi</em> version 1.32.0 <a href="00Index.html">Index</a>]</div>
</body></html>
