<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Make an OrganismDb object from an existing TxDb object.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeOrganismDbFromTxDb {OrganismDbi}"><tr><td>makeOrganismDbFromTxDb {OrganismDbi}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Make an OrganismDb object from an existing TxDb object.
</h2>

<h3>Description</h3>

<p>The <code>makeOrganismDbFromTxDb</code> function allows the user
to make a <a href="OrganismDb.html">OrganismDb</a> object from an existing TxDb object.
</p>


<h3>Usage</h3>

<pre>
makeOrganismDbFromTxDb(txdb, keytype=NA, orgdb=NA)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>txdb</code></td>
<td>
<p>a <code>TxDb</code> object</p>
</td></tr></table>
<p>.
</p>
<table summary="R argblock">
<tr valign="top"><td><code>keytype</code></td>
<td>
<p>By default, <code>makeOrganismDbFromTxDb</code> will try to
guess this information based on the OrgDb object that is inferred to
go with your TxDb object...  But in some instances, you may need to
supply an over-ride and that is what this argument is for.  It is the
column name of the ID type that your OrgDb will use as a foreign key
when connecting to the data from the associated TxDb.  So for example,
if you looked at the Homo.sapiens package the keytype for
<code>org.Hs.eg.db</code>, would be 'ENTREZID' because that is the kind of
ID that matches up with it's TxDb GENEID. (Because the GENEID for that
specific TxDb is from UCSC and uses entrez gene IDs)</p>
</td></tr></table>
<p>.
</p>
<table summary="R argblock">
<tr valign="top"><td><code>orgdb</code></td>
<td>
<p>By default, <code>makeOrganismDbFromTxDb</code> will use the
taxonomyID from your txdb to lookup an appropriate matching
<code>OrgDb</code> object but using this you can supply a different
<code>OrgDb</code> object.</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>makeOrganismDbFromTxDb</code> is a convenience function that processes
a <code>TxDb</code> object and pairs it up with GO.db and an appropriate
<code>OrgDb</code> object to make a <code>OrganismDb</code> object.
See <code>?<a href="makeOrganismDbFromBiomart.html">makeOrganismDbFromBiomart</a></code> and
<code>?<a href="makeOrganismDbFromUCSC.html">makeOrganismDbFromUCSC</a></code> for a similar function that
feeds data from either a BioMart or UCSC.
</p>


<h3>Value</h3>

<p>A <a href="OrganismDb.html">OrganismDb</a> object.</p>


<h3>Author(s)</h3>

<p>M. Carlson
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="makeOrganismDbFromBiomart.html">makeOrganismDbFromBiomart</a></code> for convenient ways to make a
<a href="OrganismDb.html">OrganismDb</a> object from BioMart online resources.
</p>
</li>
<li><p> The <a href="OrganismDb.html">OrganismDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## Not run: 
## lets start with a txdb object
transcript_ids &lt;- c(
    "uc009uzf.1",
    "uc009uzg.1",
    "uc009uzh.1",
    "uc009uzi.1",
    "uc009uzj.1"
)
txdbMouse &lt;- makeTxDbFromUCSC(genome="mm9", tablename="knownGene",
                          transcript_ids=transcript_ids)

## Using that, we can call our function to promote it to an OrgDb object:
odb &lt;- makeOrganismDbFromTxDb(txdb=txdbMouse)

columns(odb)

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>OrganismDbi</em> version 1.32.0 <a href="00Index.html">Index</a>]</div>
</body></html>
