<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Make a OrganismDb object from annotations available on a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeOrganismDbFromBiomart {OrganismDbi}"><tr><td>makeOrganismDbFromBiomart {OrganismDbi}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Make a OrganismDb object from annotations available on a
BioMart database
</h2>

<h3>Description</h3>

<p>The <code>makeOrganismDbFromBiomart</code> function allows the user
to make a <a href="OrganismDb.html">OrganismDb</a> object from transcript annotations
available on a BioMart database.  This object has all the benefits of
a TxDb, plus an associated OrgDb and GODb object.
</p>


<h3>Usage</h3>

<pre>
makeOrganismDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",
                    dataset="hsapiens_gene_ensembl",
                    transcript_ids=NULL,
                    circ_seqs=NULL,
                    filter="",
                    id_prefix="ensembl_",
                    host="www.ensembl.org",
                    port=80,
                    miRBaseBuild=NA,
                    keytype = "ENSEMBL",
                    orgdb = NA)

</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>biomart</code></td>
<td>
<p>which BioMart database to use.
Get the list of all available BioMart databases with the
<code><a href="../../biomaRt/html/listMarts.html">listMarts</a></code> function from the biomaRt
package. See the details section below for a list of BioMart
databases with compatible transcript annotations.</p>
</td></tr>
<tr valign="top"><td><code>dataset</code></td>
<td>
<p>which dataset from BioMart. For example:
<code>"hsapiens_gene_ensembl"</code>, <code>"mmusculus_gene_ensembl"</code>,
<code>"dmelanogaster_gene_ensembl"</code>, <code>"celegans_gene_ensembl"</code>,
<code>"scerevisiae_gene_ensembl"</code>, etc in the ensembl database.
See the examples section below for how to discover which datasets
are available in a given BioMart database.</p>
</td></tr>
<tr valign="top"><td><code>transcript_ids</code></td>
<td>
<p>optionally, only retrieve transcript
annotation data for the specified set of transcript ids.
If this is used, then the meta information displayed for the
resulting <a href="OrganismDb.html">TxDb</a> object will say 'Full dataset: no'.
Otherwise it will say 'Full dataset: yes'.  This <a href="OrganismDb.html">TxDb</a> object
will be embedded in the resulting <a href="OrganismDb.html">OrganismDb</a> object.</p>
</td></tr>
<tr valign="top"><td><code>circ_seqs</code></td>
<td>
<p>a character vector to list out which chromosomes
should be marked as circular.</p>
</td></tr>
<tr valign="top"><td><code>filter</code></td>
<td>
<p>Additional filters to use in the BioMart query. Must be
a named list. An example is <code>filter=as.list(c(source="entrez"))</code></p>
</td></tr>
<tr valign="top"><td><code>host</code></td>
<td>
<p>The host URL of the BioMart. Defaults to www.ensembl.org.</p>
</td></tr>
<tr valign="top"><td><code>port</code></td>
<td>
<p>The port to use in the HTTP communication with the host.</p>
</td></tr>
<tr valign="top"><td><code>id_prefix</code></td>
<td>
<p>Specifies the prefix used in BioMart attributes. For
example, some BioMarts may have an attribute specified as
<code>"ensembl_transcript_id"</code> whereas others have the same attribute
specified as <code>"transcript_id"</code>. Defaults to <code>"ensembl_"</code>.</p>
</td></tr>
<tr valign="top"><td><code>miRBaseBuild</code></td>
<td>
<p>specify the string for the appropriate build
Information from mirbase.db to use for microRNAs.  This can be
learned by calling <code>supportedMiRBaseBuildValues</code>.  By default,
this value will be set to <code>NA</code>, which will inactivate the
<code>microRNAs</code> accessor.</p>
</td></tr>
<tr valign="top"><td><code>keytype</code></td>
<td>
<p>This indicates the kind of key that this database will
use as a foreign key between it's TxDb object and it's OrgDb
object. So basically whatever the column name is for the foreign key
from your OrgDb that your TxDb will need to map it's GENEID on to.
By default it is &quot;ENSEMBL&quot; since the GENEID's for most biomaRt based
TxDbs will be ensembl gene ids and therefore they will need to map
to ENSEMBL gene mappings from the associated OrgDb object.</p>
</td></tr>
<tr valign="top"><td><code>orgdb</code></td>
<td>
<p>By default, <code>makeOrganismDbFromBiomart</code> will use the
taxonomyID from your txdb to lookup an appropriate matching
<code>OrgDb</code> object but using this you can supply a different
<code>OrgDb</code> object.</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>makeOrganismDbFromBiomart</code> is a convenience function that feeds
data from a BioMart database to the lower level
<code><a href="OrganismDb.html">OrganismDb</a></code> constructor.
See <code>?<a href="makeOrganismDbFromUCSC.html">makeOrganismDbFromUCSC</a></code> for a similar function
that feeds data from the UCSC source.
</p>
<p>The <code>listMarts</code> function from the <span class="pkg">biomaRt</span> package can be
used to list all public BioMart databases.
Not all databases returned by this function contain datasets that
are compatible with (i.e. understood by) <code>makeOrganismDbFromBiomart</code>.
Here is a list of datasets known to be compatible (updated on Sep 24, 2014):
</p>

<ul>
<li><p> All the datasets in the main Ensembl database:
use <code>biomart="ensembl"</code>.
</p>
</li>
<li><p> All the datasets in the Ensembl Fungi database:
use <code>biomart="fungi_mart_XX"</code> where XX is the release
version of the database e.g. <code>"fungi_mart_22"</code>.
</p>
</li>
<li><p> All the datasets in the Ensembl Metazoa database:
use <code>biomart="metazoa_mart_XX"</code> where XX is the release
version of the database e.g. <code>"metazoa_mart_22"</code>.
</p>
</li>
<li><p> All the datasets in the Ensembl Plants database:
use <code>biomart="plants_mart_XX"</code> where XX is the release
version of the database e.g. <code>"plants_mart_22"</code>.
</p>
</li>
<li><p> All the datasets in the Ensembl Protists database:
use <code>biomart="protists_mart_XX"</code> where XX is the release
version of the database e.g. <code>"protists_mart_22"</code>.
</p>
</li>
<li><p> All the datasets in the Gramene Mart:
use <code>biomart="ENSEMBL_MART_PLANT"</code>.
</p>
</li></ul>

<p>Not all these datasets have CDS information.
</p>


<h3>Value</h3>

<p>A <a href="OrganismDb.html">OrganismDb</a> object.</p>


<h3>Author(s)</h3>

<p>M. Carlson
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="makeOrganismDbFromUCSC.html">makeOrganismDbFromUCSC</a></code> for convenient ways to make a
<a href="OrganismDb.html">OrganismDb</a> object from UCSC online resources.
</p>
</li>
<li><p> The <code><a href="../../biomaRt/html/listMarts.html">listMarts</a></code>, <code><a href="../../biomaRt/html/useMart.html">useMart</a></code>,
and <code><a href="../../biomaRt/html/listDatasets.html">listDatasets</a></code> functions in the
<span class="pkg">biomaRt</span> package.
</p>
</li>
<li><p> The <code><a href="../../GenomicFeatures/html/makeTxDbPackage.html">supportedMiRBaseBuildValues</a></code> function for
listing all the possible values for the <code>miRBaseBuild</code>
argument.
</p>
</li>
<li><p> The <a href="OrganismDb.html">OrganismDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## Discover which datasets are available in the "ensembl" BioMart
## database:
library(biomaRt)
head(listDatasets(useMart("ensembl")))

## Retrieving an incomplete transcript dataset for Human from the
## "ensembl" BioMart database:
transcript_ids &lt;- c(
    "ENST00000013894",
    "ENST00000268655",
    "ENST00000313243",
    "ENST00000435657",
    "ENST00000384428",
    "ENST00000478783"
)
odb &lt;- makeOrganismDbFromBiomart(transcript_ids=transcript_ids)
odb  # note that these annotations match the GRCh38 genome assembly

## Now what if we want to use another mirror?  We might make use of the
## new host argument.  But wait!  If we use biomaRt, we can see that
## this host has named the mart differently!
listMarts(host="useast.ensembl.org")
## Therefore we must also change the name passed into the "mart"
## argument thusly:
try(
    odb &lt;- makeOrganismDbFromBiomart(biomart="ENSEMBL_MART_ENSEMBL",
                                     transcript_ids=transcript_ids,
                                     host="useast.ensembl.org")
)
odb
</pre>

<hr /><div style="text-align: center;">[Package <em>OrganismDbi</em> version 1.32.0 <a href="00Index.html">Index</a>]</div>
</body></html>
