<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: BSgenomeViews objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for BSgenomeViews-class {BSgenome}"><tr><td>BSgenomeViews-class {BSgenome}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>BSgenomeViews objects</h2>

<h3>Description</h3>

<p>The BSgenomeViews class is a container for storing a set of genomic
positions on a <a href="BSgenome-class.html">BSgenome</a> object, called the &quot;subject&quot; in this
context.
</p>


<h3>Usage</h3>

<pre>
## Constructor
## ------------

BSgenomeViews(subject, granges)

## Accessors
## ---------

## S4 method for signature 'BSgenomeViews'
subject(x)
## S4 method for signature 'BSgenomeViews'
granges(x, use.mcols=FALSE)

## S4 method for signature 'BSgenomeViews'
length(x)
## S4 method for signature 'BSgenomeViews'
names(x)
## S4 method for signature 'BSgenomeViews'
seqnames(x)
## S4 method for signature 'BSgenomeViews'
start(x)
## S4 method for signature 'BSgenomeViews'
end(x)
## S4 method for signature 'BSgenomeViews'
width(x)
## S4 method for signature 'BSgenomeViews'
strand(x)
## S4 method for signature 'BSgenomeViews'
ranges(x, use.mcols=FALSE)
## S4 method for signature 'BSgenomeViews'
elementNROWS(x)
## S4 method for signature 'BSgenomeViews'
seqinfo(x)

## DNAStringSet methods
## --------------------

## S4 method for signature 'BSgenomeViews'
seqtype(x)

## S4 method for signature 'BSgenomeViews'
nchar(x, type="chars", allowNA=FALSE)

## S4 method for signature 'BSgenomeViews'
unlist(x, recursive=TRUE, use.names=TRUE)

## S4 method for signature 'BSgenomeViews'
alphabetFrequency(x, as.prob=FALSE, collapse=FALSE, baseOnly=FALSE)

## S4 method for signature 'BSgenomeViews'
hasOnlyBaseLetters(x)

## S4 method for signature 'BSgenomeViews'
uniqueLetters(x)

## S4 method for signature 'BSgenomeViews'
letterFrequency(x, letters, OR="|", as.prob=FALSE, collapse=FALSE)

## S4 method for signature 'BSgenomeViews'
oligonucleotideFrequency(x, width, step=1,
                         as.prob=FALSE, as.array=FALSE,
                         fast.moving.side="right", with.labels=TRUE, simplify.as="matrix")

## S4 method for signature 'BSgenomeViews'
nucleotideFrequencyAt(x, at, as.prob=FALSE, as.array=TRUE,
                      fast.moving.side="right", with.labels=TRUE)

## S4 method for signature 'BSgenomeViews'
consensusMatrix(x, as.prob=FALSE, shift=0L, width=NULL, baseOnly=FALSE)

## S4 method for signature 'BSgenomeViews'
consensusString(x, ambiguityMap=IUPAC_CODE_MAP, threshold=0.25,
                shift=0L, width=NULL)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>subject</code></td>
<td>

<p>A <a href="BSgenome-class.html">BSgenome</a> object or the name of a reference genome specified
in a way that is accepted by the <code><a href="available.genomes.html">getBSgenome</a></code> function.
In that case the corresponding BSgenome data package needs to be already
installed (see <code>?<a href="available.genomes.html">getBSgenome</a></code> for the details).
</p>
</td></tr>
<tr valign="top"><td><code>granges</code></td>
<td>

<p>A <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object containing ranges relative to
the genomic sequences stored in <code>subject</code>.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>A BSgenomeViews object.
</p>
</td></tr>
<tr valign="top"><td><code>use.mcols</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code> (the default).
Whether the metadata columns on <code>x</code> (accessible with <code>mcols(x)</code>)
should be propagated to the returned object or not.
</p>
</td></tr>
<tr valign="top"><td><code>type, allowNA, recursive, use.names</code></td>
<td>

<p>Ignored.
</p>
</td></tr>
<tr valign="top"><td><code>as.prob, letters, OR, width</code></td>
<td>

<p>See <code>?<a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a></code> and
<code>?<a href="../../Biostrings/html/oligonucleotideFrequency.html">oligonucleotideFrequency</a></code> in the
<span class="pkg">Biostrings</span> package.
</p>
</td></tr>
<tr valign="top"><td><code>collapse, baseOnly</code></td>
<td>

<p>See <code>?<a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a></code> in the
<span class="pkg">Biostrings</span> package.
</p>
</td></tr>
<tr valign="top"><td><code>step, as.array, fast.moving.side, with.labels, simplify.as, at</code></td>
<td>

<p>See <code>?<a href="../../Biostrings/html/oligonucleotideFrequency.html">oligonucleotideFrequency</a></code> in the
<span class="pkg">Biostrings</span> package.
</p>
</td></tr>
<tr valign="top"><td><code>shift, ambiguityMap, threshold</code></td>
<td>

<p>See <code>?<a href="../../Biostrings/html/consensusMatrix.html">consensusMatrix</a></code> in the
<span class="pkg">Biostrings</span> package.
</p>
</td></tr>
</table>


<h3>Constructors</h3>


<dl>
<dt></dt><dd>
<p><code>BSgenomeViews(subject, granges)</code>:
Make a BSgenomeViews object by putting the views specified by
<code>granges</code> on top of the genomic sequences stored in <code>subject</code>.
See above for how argument <code>subject</code> and <code>granges</code> should be
specified.
</p>
</dd>
<dt></dt><dd>
<p><code>Views(subject, granges)</code>: Equivalent to
<code>BSgenomeViews(subject, granges)</code>. Provided for convenience.
</p>
</dd>
</dl>



<h3>Accessors</h3>

<p>In the code snippets below, <code>x</code> is a BSgenomeViews object.
</p>

<dl>
<dt></dt><dd>
<p><code>subject(x)</code>: Return the <a href="BSgenome-class.html">BSgenome</a> object containing the
full genomic sequences on top of which the views in <code>x</code> are
defined.
</p>
</dd>
<dt></dt><dd>
<p><code>granges(x, use.mcols=FALSE)</code>: Return the genomic ranges of the
views as a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object. These ranges are
relative to the genomic sequences stored in <code>subject(x)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>length(x)</code>: The number of views in <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>names(x)</code>: The names of the views in <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>seqnames(x)</code>, <code>start(x)</code>, <code>end(x)</code>, <code>width(x)</code>,
<code>strand(x)</code>: Equivalent to <code>seqnames(granges(x))</code>,
<code>start(granges(x))</code>, <code>end(granges(x))</code>,
<code>width(granges(x))</code>, <code>strand(granges(x))</code>, respectively.
</p>
</dd>
<dt></dt><dd>
<p><code>ranges(x, use.mcols=FALSE)</code>: Equivalent to
<code>ranges(granges(x, use.mcols), use.mcols)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>elementNROWS(x)</code>: Equivalent to <code>width(x)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>seqinfo(x)</code>: Equivalent to <code>seqinfo(subject(x))</code> and to
<code>seqinfo(granges(x))</code> (both are guaranteed to be the same).
See <code>?<a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a></code> in the <span class="pkg">GenomeInfoDb</span>
package for more information.
</p>
</dd>
</dl>



<h3>Coercion</h3>

<p>In the code snippets below, <code>x</code> is a BSgenomeViews object.
</p>

<dl>
<dt></dt><dd>
<p><code>as(x, "DNAStringSet")</code>: Turn <code>x</code> into a
<a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object by extxracting the DNA sequence
corresponding to each view. Alternatively <code>as(x, "XStringSet")</code>
can be used for this, and is equivalent to <code>as(x, "DNAStringSet")</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>as.character(x)</code>: Equivalent to
<code>as.character(as(x, "DNAStringSet"))</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>as.data.frame(x)</code>: Turn <code>x</code> into a data.frame.
</p>
</dd>
</dl>



<h3>Subsetting</h3>


<dl>
<dt></dt><dd>
<p><code>x[i]</code>: Select the views specified by <code>i</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>x[[i]]</code>: Extract the one view specified by <code>i</code>.
</p>
</dd>
</dl>



<h3>DNAStringSet methods</h3>

<p>For convenience, some methods defined for <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a>
objects in the <span class="pkg">Biostrings</span> package can be used directly on a
BSgenomeViews object. In that case, everything happens like if the
BSgenomeViews object <code>x</code> was turned into a
<a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object (with <code>as(x, "DNAStringSet")</code>)
before it's passed to the method for <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> objects.
</p>
<p>At the moment, the list of such methods is:
<code><a href="../../Biostrings/html/seqtype.html">seqtype</a></code>,
<code><a href="../../Biostrings/html/nchar,XStringSet-method.html">nchar,XStringSet-method</a></code>,
<code><a href="../../Biostrings/html/unlist,XStringSet-method.html">unlist,XStringSet-method</a></code>,
<code><a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a></code>,
<code><a href="../../Biostrings/html/hasOnlyBaseLetters.html">hasOnlyBaseLetters</a></code>,
<code><a href="../../Biostrings/html/uniqueLetters.html">uniqueLetters</a></code>,
<code><a href="../../Biostrings/html/letterFrequency.html">letterFrequency</a></code>,
<code><a href="../../Biostrings/html/oligonucleotideFrequency.html">oligonucleotideFrequency</a></code>,
<code><a href="../../Biostrings/html/nucleotideFrequencyAt.html">nucleotideFrequencyAt</a></code>,
<code><a href="../../Biostrings/html/consensusMatrix.html">consensusMatrix</a></code>,
and <code><a href="../../Biostrings/html/consensusString.html">consensusString</a></code>.
</p>
<p>See the corresponding man page in the <span class="pkg">Biostrings</span> package for a
description of these methods.
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s</p>


<h3>See Also</h3>


<ul>
<li><p> The <a href="BSgenome-class.html">BSgenome</a> class.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> class in the <span class="pkg">GenomicRanges</span>
package.
</p>
</li>
<li><p> The <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> class in the <span class="pkg">Biostrings</span>
package.
</p>
</li>
<li><p> The <a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a> and related getters in the
<span class="pkg">GenomeInfoDb</span> package for getting the sequence information
stored in an object.
</p>
</li>
<li> <p><a href="../../GenomicFeatures/html/TxDb.html">TxDb</a> objects in the <span class="pkg">GenomicFeatures</span>
package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
library(BSgenome.Mmusculus.UCSC.mm10)
genome &lt;- BSgenome.Mmusculus.UCSC.mm10
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb &lt;- TxDb.Mmusculus.UCSC.mm10.knownGene
ex &lt;- exons(txdb, columns=c("exon_id", "tx_name", "gene_id"))
v &lt;- Views(genome, ex)
v

subject(v)
granges(v)
seqinfo(v)
as(v, "DNAStringSet")

v10 &lt;- v[1:10]  # select the first 10 views
subject(v10)    # same as subject(v)
granges(v10)
seqinfo(v10)    # same as seqinfo(v)
as(v10, "DNAStringSet")
alphabetFrequency(v10)
alphabetFrequency(v10, collapse=TRUE)

v12 &lt;- v[width(v) &lt;= 12]  # select the views of 12 nucleotides or less
head(as.data.frame(v12))
trinucleotideFrequency(v12, simplify.as="collapsed")

## BSgenomeViews objects are list-like objects. That is, the
## BSgenomeViews class derives from List and typical list/List
## operations (e.g. [[, elementNROWS(), unlist(), elementType(),
## etc...) work on these objects:
is(v12, "List")  # TRUE
v12[[2]]
head(elementNROWS(v12))  # elementNROWS(v) is the same as width(v)
unlist(v12) 
elementType(v12)
</pre>

<hr /><div style="text-align: center;">[Package <em>BSgenome</em> version 1.58.0 <a href="00Index.html">Index</a>]</div>
</body></html>
