<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: getSeq methods for BSgenome and XStringSet objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for getSeq-methods {BSgenome}"><tr><td>getSeq-methods {BSgenome}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>getSeq methods for BSgenome and XStringSet objects</h2>

<h3>Description</h3>

<p><code><a href="../../Biostrings/html/getSeq.html">getSeq</a></code> methods for extracting a set of
sequences (or subsequences) from a <a href="BSgenome-class.html">BSgenome</a> or
<a href="../../Biostrings/html/XStringSet.html">XStringSet</a> object. For XStringSets, there are also
convenience methods on <code>[</code> that delegate to <code>getSeq</code>.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'BSgenome'
getSeq(x, names, start=NA, end=NA, width=NA,
                 strand="+", as.character=FALSE) 
## S4 method for signature 'XStringSet'
getSeq(x, names)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="BSgenome-class.html">BSgenome</a> or <a href="../../Biostrings/html/XStringSet.html">XStringSet</a> object.
See the <code><a href="available.genomes.html">available.genomes</a></code> function for how
to install a genome.
</p>
</td></tr>
<tr valign="top"><td><code>names</code></td>
<td>

<p>When <code>x</code> is a <a href="BSgenome-class.html">BSgenome</a>, <code>names</code> must be a
character vector containing the names of the sequences
in <code>x</code> where to get the subsequences from,
or a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object,
or a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object,
or a named <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object,
or a named <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> object.
The <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> or <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a>
object must be named according to the sequences
in <code>x</code> where to get the subsequences from.
</p>
<p>If <code>names</code> is missing, then <code>seqnames(x)</code> is used.
</p>
<p>See <code>?`<a href="BSgenome-class.html">BSgenome-class</a>`</code> for details on how to get
the lists of single sequences and multiple sequences (respectively)
contained in a <a href="BSgenome-class.html">BSgenome</a> object.
</p>
<p>When <code>x</code> is a <a href="../../Biostrings/html/XStringSet.html">XStringSet</a> object, <code>names</code>
must be a character vector, <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>start, end, width</code></td>
<td>

<p>Vector of integers (eventually with NAs) specifying the locations
of the subsequences to extract.
These are not needed (and it's an error to supply them)
when <code>names</code> is a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a>,
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a>, <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a>,
or <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>strand</code></td>
<td>

<p>A vector containing <code>"+"</code>s or/and <code>"-"</code>s.
This is not needed (and it's an error to supply it)
when <code>names</code> is a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>as.character</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>. Should the extracted sequences
be returned in a standard character vector?
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>L, the number of sequences to extract, is determined as follow:
</p>

<ul>
<li><p> If <code>names</code> is a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> object then L = <code>length(names)</code>.
</p>
</li>
<li><p> If <code>names</code> is a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> or
<a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object then
L = <code>length(unlist(names))</code>.
</p>
</li>
<li><p> Otherwise, L is the length of the longest of <code>names</code>,
<code>start</code>, <code>end</code> and <code>width</code> and all these
arguments are recycled to this length.
<code>NA</code>s and negative values in these 3 arguments are
solved according to the rules of the SEW (Start/End/Width)
interface (see <code>?<a href="../../IRanges/html/solveUserSEW.html">solveUserSEW</a></code> for
the details).
</p>
</li></ul>

<p>If <code>names</code> is neither a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object, then the <code>strand</code>
argument is also recycled to length L.
</p>
<p>Here is how the names passed to the <code>names</code> argument are matched
to the names of the sequences in <a href="BSgenome-class.html">BSgenome</a> object <code>x</code>.
For each <code>name</code> in <code>names</code>:
</p>

<ul>
<li><p> (1): If <code>x</code> contains a single sequence with that name
then this sequence is used for extraction;
</p>
</li>
<li><p> (2): Otherwise the names of all the elements in all the
multiple sequences are searched. If the <code>names</code> argument
is a character vector then <code>name</code> is treated as a regular
expression and <code><a href="../../base/html/grep.html">grep</a></code> is used for this search,
otherwise (i.e. when the names are supplied via a higher level
object like <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a>) then <code>name</code> must match
exactly the name of the sequence. If exactly 1 sequence is found,
then it is used for extraction, otherwise (i.e. if no sequence or
more than 1 sequence is found) then an error is raised.
</p>
</li></ul>

<p>There are convenience methods for extracting sequences from
<a href="../../Biostrings/html/XStringSet.html">XStringSet</a> objects using a
<a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> or <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a>
subscript (character subscripts are implicitly supported). Both methods
are simple wrappers around <code>getSeq</code>, although the GRangesList method
differs from the <code>getSeq</code> behavior in that the within-element results
are concatenated and returned as an XStringSet, rather than an
XStringSetList. See the examples.
</p>


<h3>Value</h3>

<p>Normally a <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object (or character vector
if <code>as.character=TRUE</code>).
</p>
<p>With the 2 following exceptions:
</p>

<ol>
<li><p> A <a href="../../Biostrings/html/DNAStringSetList.html">DNAStringSetList</a> object (or
<a href="../../IRanges/html/CharacterList.html">CharacterList</a> object if <code>as.character=TRUE</code>)
of the same shape as <code>names</code> if <code>names</code> is a
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
</p>
</li>
<li><p> A <a href="../../Biostrings/html/DNAString.html">DNAString</a> object (or single character string
if <code>as.character=TRUE</code>) if L = 1 and <code>names</code>
is not a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a>,
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a>, <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a>,
or <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> object.
</p>
</li></ol>



<h3>Note</h3>

<p>Be aware that using <code>as.character=TRUE</code> can be very inefficient
when extracting a &quot;big&quot; amount of DNA sequences (e.g. millions of
short sequences or a small number of very long sequences).
</p>
<p>Note that the masks in <code>x</code>, if any, are always ignored. In other
words, masked regions in the genome are extracted in the same way as
unmasked regions (this is achieved by dropping the masks before extraction).
See <code>?`<a href="../../Biostrings/html/MaskedDNAString-class.html">MaskedDNAString-class</a>`</code> for more
information about masked DNA sequences.
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s; improvements suggested by Matt Settles and others</p>


<h3>See Also</h3>

<p><code><a href="../../Biostrings/html/getSeq.html">getSeq</a></code>,
<code><a href="available.genomes.html">available.genomes</a></code>,
<a href="BSgenome-class.html">BSgenome-class</a>,
<a href="../../Biostrings/html/DNAString-class.html">DNAString-class</a>,
<a href="../../Biostrings/html/DNAStringSet-class.html">DNAStringSet-class</a>,
<a href="../../Biostrings/html/MaskedDNAString-class.html">MaskedDNAString-class</a>,
<a href="../../GenomicRanges/html/GRanges-class.html">GRanges-class</a>,
<a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList-class</a>,
<a href="../../IRanges/html/IntegerRangesList-class.html">IntegerRangesList-class</a>,
<a href="../../IRanges/html/IntegerRanges-class.html">IntegerRanges-class</a>,
<code><a href="../../base/html/grep.html">grep</a></code>
</p>


<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. SIMPLE EXAMPLES
## ---------------------------------------------------------------------

## Load the Caenorhabditis elegans genome (UCSC Release ce2):
library(BSgenome.Celegans.UCSC.ce2)

## Look at the index of sequences:
Celegans

## Get chromosome V as a DNAString object:
getSeq(Celegans, "chrV")
## which is in fact the same as doing:
Celegans$chrV

## Not run: 
  ## Never try this:
  getSeq(Celegans, "chrV", as.character=TRUE)
  ## or this (even worse):
  getSeq(Celegans, as.character=TRUE)

## End(Not run)

## Get the first 20 bases of each chromosome:
getSeq(Celegans, end=20)

## Get the last 20 bases of each chromosome:
getSeq(Celegans, start=-20)

## ---------------------------------------------------------------------
## B. EXTRACTING SMALL SEQUENCES FROM DIFFERENT CHROMOSOMES
## ---------------------------------------------------------------------

myseqs &lt;- data.frame(
  chr=c("chrI", "chrX", "chrM", "chrM", "chrX", "chrI", "chrM", "chrI"),
  start=c(NA, -40, 8510, 301, 30001, 9220500, -2804, -30),
  end=c(50, NA, 8522, 324, 30011, 9220555, -2801, -11),
  strand=c("+", "-", "+", "+", "-", "-", "+", "-")
)
getSeq(Celegans, myseqs$chr,
       start=myseqs$start, end=myseqs$end)
getSeq(Celegans, myseqs$chr,
       start=myseqs$start, end=myseqs$end, strand=myseqs$strand)

## ---------------------------------------------------------------------
## C. USING A GRanges OBJECT
## ---------------------------------------------------------------------

gr1 &lt;- GRanges(seqnames=c("chrI", "chrI", "chrM"),
               ranges=IRanges(start=101:103, width=9))
gr1  # all strand values are "*"
getSeq(Celegans, gr1)  # treats strand values as if they were "+"

strand(gr1)[] &lt;- "-"
getSeq(Celegans, gr1)

strand(gr1)[1] &lt;- "+"
getSeq(Celegans, gr1)

strand(gr1)[2] &lt;- "*"
if (interactive())
  getSeq(Celegans, gr1)  # Error: cannot mix "*" with other strand values

gr2 &lt;- GRanges(seqnames=c("chrM", "NM_058280_up_1000"),
               ranges=IRanges(start=103:102, width=9))
gr2
if (interactive()) {
  ## Because the sequence names are supplied via a GRanges object, they
  ## are not treated as regular expressions:
  getSeq(Celegans, gr2)  # Error: sequence NM_058280_up_1000 not found
}

## ---------------------------------------------------------------------
## D. USING A GRangesList OBJECT
## ---------------------------------------------------------------------

gr1 &lt;- GRanges(seqnames=c("chrI", "chrII", "chrM", "chrII"),
               ranges=IRanges(start=101:104, width=12),
               strand="+")
gr2 &lt;- shift(gr1, 5)
gr3 &lt;- gr2
strand(gr3) &lt;- "-"

grl &lt;- GRangesList(gr1, gr2, gr3)
getSeq(Celegans, grl)

## ---------------------------------------------------------------------
## E. EXTRACTING A HIGH NUMBER OF RANDOM 40-MERS FROM A GENOME
## ---------------------------------------------------------------------

extractRandomReads &lt;- function(x, density, readlength)
{
    if (!is.integer(readlength))
        readlength &lt;- as.integer(readlength)
    start &lt;- lapply(seqnames(x),
                    function(name)
                    {
                      seqlength &lt;- seqlengths(x)[name]
                      sample(seqlength - readlength + 1L,
                             seqlength * density,
                             replace=TRUE)
                    })
    names &lt;- rep.int(seqnames(x), elementNROWS(start))
    ranges &lt;- IRanges(start=unlist(start), width=readlength)
    strand &lt;- strand(sample(c("+", "-"), length(names), replace=TRUE))
    gr &lt;- GRanges(seqnames=names, ranges=ranges, strand=strand)
    getSeq(x, gr)
}

## With a density of 1 read every 100 genome bases, the total number of
## extracted 40-mers is about 1 million:
rndreads &lt;- extractRandomReads(Celegans, 0.01, 40)

## Notes:
## - The short sequences in 'rndreads' can be seen as the result of a
##   simulated high-throughput sequencing experiment. A non-realistic
##   one though because:
##     (a) It assumes that the underlying technology is perfect (the
##         generated reads have no technology induced errors).
##     (b) It assumes that the sequenced genome is exactly the same as
##         the reference genome.
##     (c) The simulated reads can contain IUPAC ambiguity letters only
##         because the reference genome contains them. In a real
##         high-throughput sequencing experiment, the sequenced genome
##         of course doesn't contain those letters, but the sequencer
##         can introduce them in the generated reads to indicate
##         ambiguous base-calling.
## - Those reads are coming from the plus and minus strands of the
##   chromosomes.
## - With a density of 0.01 and the reads being only 40-base long, the
##   average coverage of the genome is only 0.4 which is low. The total
##   number of reads is about 1 million and it takes less than 10 sec.
##   to generate them.
## - A higher coverage can be achieved by using a higher density and/or
##   longer reads. For example, with a density of 0.1 and 100-base reads
##   the average coverage is 10. The total number of reads is about 10
##   millions and it takes less than 1 minute to generate them.
## - Those reads could easily be mapped back to the reference by using
##   an efficient matching tool like matchPDict() for performing exact
##   matching (see ?matchPDict for more information). Typically, a
##   small percentage of the reads (4 to 5% in our case) will hit the
##   reference at multiple locations. This is especially true for such
##   short reads, and, in a lower proportion, is still true for longer
##   reads, even for reads as long as 300 bases.

## ---------------------------------------------------------------------
## F. SEE THE BSgenome CACHE IN ACTION
## ---------------------------------------------------------------------

options(verbose=TRUE)
first20 &lt;- getSeq(Celegans, end=20)
first20
gc()
stopifnot(length(ls(Celegans@.seqs_cache)) == 0L)
## One more gc() call is needed in order to see the amount of memory in
## used after all the chromosomes have been removed from the cache:
gc()

## ---------------------------------------------------------------------
## G. USING '[' FOR CONVENIENT EXTRACTION
## ---------------------------------------------------------------------

seqs &lt;- getSeq(Celegans)
seqs[gr1]
seqs[grl]

</pre>

<hr /><div style="text-align: center;">[Package <em>BSgenome</em> version 1.58.0 <a href="00Index.html">Index</a>]</div>
</body></html>
