<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Convert genotype calls from a VCF file to a SnpMatrix object</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for genotypeToSnpMatrix {VariantAnnotation}"><tr><td>genotypeToSnpMatrix {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Convert genotype calls from a VCF file to a SnpMatrix object</h2>

<h3>Description</h3>

<p>Convert an array of genotype calls from the &quot;GT&quot;, &quot;GP&quot;, &quot;GL&quot; or &quot;PL&quot;
FORMAT field of a VCF file to a <a href="../../snpStats/html/SnpMatrix-class.html">SnpMatrix</a>.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'CollapsedVCF'
genotypeToSnpMatrix(x, uncertain=FALSE, ...)
## S4 method for signature 'array'
genotypeToSnpMatrix(x, ref, alt, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <code>CollapsedVCF</code> object or a <code>array</code> of genotype data 
from the &quot;GT&quot;, &quot;GP&quot;, &quot;GL&quot; or &quot;PL&quot; FORMAT field of a VCF file.  This <code>array</code> is created 
with a call to <code>readVcf</code> and can be accessed with <code>geno(&lt;VCF&gt;)</code>.
</p>
</td></tr>
<tr valign="top"><td><code>uncertain</code></td>
<td>

<p>A logical indicating whether the genotypes to convert
should come from the &quot;GT&quot; field (<code>uncertain=FALSE</code>) or the &quot;GP&quot;, 
&quot;GL&quot; or &quot;PL&quot; field (<code>uncertain=TRUE</code>).
</p>
</td></tr>
<tr valign="top"><td><code>ref</code></td>
<td>

<p>A <code>DNAStringSet</code> of reference alleles.
</p>
</td></tr>
<tr valign="top"><td><code>alt</code></td>
<td>

<p>A <code>DNAStringSetList</code> of alternate alleles. 
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments, passed to methods.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>genotypeToSnpMatrix</code> converts an array of genotype calls from the 
&quot;GT&quot;, &quot;GP&quot;, &quot;GL&quot; or &quot;PL&quot; FORMAT field of a VCF file into a 
<a href="../../snpStats/html/SnpMatrix-class.html">SnpMatrix</a>. The following caveats apply, 
</p>

<ul>
<li><p>no distinction is made between phased and unphased genotypes
</p>
</li>
<li><p>variants with &gt;1 ALT allele are set to NA
</p>
</li>
<li><p>only single nucleotide variants are included; others are set to NA
</p>
</li>
<li><p>only diploid calls are included; others are set to NA
</p>
</li></ul>

<p>In VCF files, 0 represents the reference allele and integers greater than 0
represent the alternate alleles (i.e., 2, 3, 4 would indicate the 2nd, 3rd
or 4th allele in the ALT field for a particular variant). This function only 
supports variants with a single alternate allele and therefore the alternate 
values will always be 1. Genotypes are stored in the SnpMatrix 
as 0, 1, 2 or 3 where 0 = missing, 1 = &quot;0/0&quot;, 2 = &quot;0/1&quot; or &quot;1/0&quot; and 
3 = &quot;1/1&quot;. In SnpMatrix terminology, &quot;A&quot; is the reference allele 
and &quot;B&quot; is the risk allele. Equivalent statements to those made with 0 and 1 
allele values would be 0 = missing, 1 = &quot;A/A&quot;, 2 = &quot;A/B&quot; or &quot;B/A&quot; and
3 = &quot;B/B&quot;.
</p>
<p>The genotype fields are defined as follows:
</p>

<ul>
<li><p>GT : genotype, encoded as allele values separated by either of
&quot;/&quot; or &quot;|&quot;. The allele values are 0 for the reference allele and 1
for the alternate allele.
</p>
</li>
<li><p>GL : genotype likelihoods comprised of comma separated
floating point log10-scaled likelihoods for all possible
genotypes.  In the case of a reference allele A and a single
alternate allele B, the likelihoods will be ordered &quot;A/A&quot;, &quot;A/B&quot;,
&quot;B/B&quot;.
</p>
</li>
<li><p>PL : the phred-scaled genotype likelihoods rounded to the
closest integer.  The ordering of values is the 
same as for the GL field.
</p>
</li>
<li><p>GP : the phred-scaled genotype posterior probabilities for all
possible genotypes; intended to store
imputed genotype probabilities.  The ordering of values is the
same as for the GL field.
</p>
</li></ul>

<p>If <code>uncertain=TRUE</code>, the posterior probabilities of the three
genotypes (&quot;A/A&quot;, &quot;A/B&quot;, &quot;B/B&quot;) are encoded (approximately) as byte
values.  This encoding allows uncertain genotypes to be used in
<a href="../../snpStats/html/snpStats.html">snpStats</a> functions, which in some cases may be more
appropriate than using only the called genotypes.  The byte encoding
conserves memory by allowing the uncertain genotypes to be stored in a
two-dimensional raw matrix.
See the <a href="../../snpStats/html/snpStats.html">snpStats</a> documentation for more details.
</p>


<h3>Value</h3>

<p>A list with the following elements,
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>genotypes</code></td>
<td>

<p>The output genotype data as an object of class 
<code>"SnpMatrix"</code>. The columns are snps and the rows are the samples. 
See ?<code>SnpMatrix</code> details of the class structure.
</p>
</td></tr>
<tr valign="top"><td><code>map</code></td>
<td>
 
<p>A <code>DataFrame</code> giving the snp names and alleles at each locus.
The <code>ignore</code> column indicates which variants were set to <code>NA</code>
(see <code>NA</code> criteria in 'details' section).
</p>
</td></tr> 
</table>


<h3>Author(s)</h3>

<p>Stephanie Gogarten and Valerie Obenchain
</p>


<h3>References</h3>

<p><a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41</a> 
</p>


<h3>See Also</h3>

<p><a href="readVcf-methods.html">readVcf</a>,
<a href="VCF-class.html">VCF</a>,
<a href="../../snpStats/html/SnpMatrix-class.html">SnpMatrix</a>
</p>


<h3>Examples</h3>

<pre>
  ## ----------------------------------------------------------------
  ## Non-probability based snp encoding using "GT"
  ## ----------------------------------------------------------------
  fl &lt;- system.file("extdata", "ex2.vcf", package="VariantAnnotation") 
  vcf &lt;- readVcf(fl, "hg19")

  ## This file has no "GL" or "GP" field so we use "GT".
  geno(vcf)

  ## Convert the "GT" FORMAT field to a SnpMatrix.
  mat &lt;- genotypeToSnpMatrix(vcf)

  ## The result is a list of length 2.
  names(mat)

  ## Compare coding in the VCF file to the SnpMatrix.
  geno(vcf)$GT
  t(as(mat$genotype, "character"))

  ## The 'ignore' column in 'map' indicates which variants 
  ## were set to NA. Variant rs6040355 was ignored because 
  ## it has multiple alternate alleles, microsat1 is not a 
  ## snp, and chr20:1230237 has no alternate allele.
  mat$map

  ## ----------------------------------------------------------------
  ## Probability-based encoding using "GL", "PL" or "GP"
  ## ----------------------------------------------------------------
  ## Read a vcf file with a "GL" field.
  fl &lt;- system.file("extdata", "gl_chr1.vcf", package="VariantAnnotation") 
  vcf &lt;- readVcf(fl, "hg19")
  geno(vcf)

  ## Convert the "GL" FORMAT field to a SnpMatrix
  mat &lt;- genotypeToSnpMatrix(vcf, uncertain=TRUE)

  ## Only 3 of the 9 variants passed the filters.  The
  ## other 6 variants had no alternate alleles.
  mat$map

  ## Compare genotype representations for a subset of
  ## samples in variant rs180734498.
  ## Original called genotype
  geno(vcf)$GT["rs180734498", 14:16]

  ## Original genotype likelihoods
  geno(vcf)$GL["rs180734498", 14:16]

  ## Posterior probability (computed inside genotypeToSnpMatrix)
  GLtoGP(geno(vcf)$GL["rs180734498", 14:16, drop=FALSE])[1,]

  ## SnpMatrix coding.
  t(as(mat$genotype, "character"))["rs180734498", 14:16]
  t(as(mat$genotype, "numeric"))["rs180734498", 14:16]

  ## For samples NA11829 and NA11830, one probability is significantly
  ## higher than the others, so SnpMatrix calls the genotype.  These
  ## calls match the original coding: "0|1" -&gt; "A/B", "0|0" -&gt; "A/A".
  ## Sample NA11831 was originally called as "0|1" but the probability
  ## of "0|0" is only a factor of 3 lower, so SnpMatrix calls it as
  ## "Uncertain" with an appropriate byte-level encoding.
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
