<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: PolyPhenDb objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for PolyPhenDb-class {VariantAnnotation}"><tr><td>PolyPhenDb-class {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>PolyPhenDb objects</h2>

<h3>Description</h3>

<p>The PolyPhenDb class is a container for storing a connection to a PolyPhen 
sqlite database.
</p>


<h3>Details</h3>

<p>PolyPhen (Polymorphism Phenotyping) is a tool which predicts the possible 
impact of an amino acid substitution on the structure and function of a 
human protein by applying empirical rules to the sequence, phylogenetic 
and structural information characterizing the substitution. 
</p>
<p>PolyPhen makes its predictions using UniProt features, PSIC profiles scores 
derived from multiple alignment and matches to PDP or PQS structural 
databases. The procedure can be roughly outlined in the following 
steps, see the references for complete details,
</p>

<ul>
<li><p> sequence-based characterization of substitution site 
</p>
</li>
<li><p> calculation of PSIC profile scores for two amino acid variants 
</p>
</li>
<li><p> calculation of structural parameters and contacts 
</p>
</li>
<li><p> prediction 
</p>
</li></ul>

<p>PolyPhen uses empirically derived rules to predict that a non-synonymous 
SNP is
</p>

<ul>
<li><p> probably damaging : it is with high confidence supposed to affect
protein function or structure
</p>
</li>
<li><p> possibly damaging : it is supposed to affect protein function or
structure
</p>
</li>
<li><p> benign : most likely lacking any phenotypic effect
</p>
</li>
<li><p> unknown : when in some rare cases, the lack of data do not allow PolyPhen to
make a prediction 
</p>
</li></ul>



<h3>Methods</h3>

<p>In the code below, <code>x</code> is a <code>PolyPhenDb</code> object.
</p>

<dl>
<dt></dt><dd>
<p><code>metadata(x)</code>:
Returns <code>x</code>'s metadata in a data frame.
</p>
</dd>
<dt></dt><dd>
<p><code>columns(x)</code>:
Returns the names of the <code>columns</code> that can be used to subset the
data columns. For column descriptions see <code>?PolyPhenDbColumns</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>keys(x)</code>:
Returns the names of the <code>keys</code> that can be used to subset the
data rows. The <code>keys</code> values are the rsid's.
</p>
</dd>
<dt></dt><dd>
<p><code>select(x, keys = NULL, columns = NULL, ...)</code>:
Returns a subset of data defined by the character vectors <code>keys</code> 
and <code>columns</code>. If no <code>keys</code> are supplied, all rows are
returned. If no <code>columns</code> are supplied, all columns
are returned. See <code>?PolyPhenDbColumns</code> for column descriptions.
</p>
</dd> 
<dt></dt><dd>
<p><code>duplicateRSID(x)</code>:
Returns a named list of duplicate rsid groups. The names are the 
<code>keys</code>, the list elements are the rsid's that have been 
reported as having identical chromosome position and alleles and 
therefore translating into the same amino acid residue substitution.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Valerie Obenchain</p>


<h3>References</h3>

<p>PolyPhen Home:
<a href="http://genetics.bwh.harvard.edu/pph2/dokuwiki/">http://genetics.bwh.harvard.edu/pph2/dokuwiki/</a>
</p>
<p>Adzhubei IA, Schmidt S, Peshkin L, Ramensky VE, Gerasimova A, Bork P,
Kondrashov AS, Sunyaev SR. Nat Methods 7(4):248-249 (2010).
</p>
<p>Ramensky V, Bork P, Sunyaev S. Human non-synonymous SNPs: server and
survey. Nucleic Acids Res 30(17):3894-3900 (2002).
</p>
<p>Sunyaev SR, Eisenhaber F, Rodchenkov IV, Eisenhaber B, Tumanyan VG,
Kuznetsov EN. PSIC: profile extraction from sequence alignments with
position-specific counts of independent observations. Protein
Eng 12(5):387-394 (1999).
</p>


<h3>See Also</h3>

<p><code>?PolyPhenDbColumns</code>
</p>


<h3>Examples</h3>

<pre>
library(PolyPhen.Hsapiens.dbSNP131)

## metadata
metadata(PolyPhen.Hsapiens.dbSNP131)

## available rsid's 
head(keys(PolyPhen.Hsapiens.dbSNP131))

## column descriptions found at ?PolyPhenDbColumns
columns(PolyPhen.Hsapiens.dbSNP131)

## subset on keys and columns 
subst &lt;- c("AA1", "AA2", "PREDICTION")
rsids &lt;- c("rs2142947", "rs4995127", "rs3026284")
select(PolyPhen.Hsapiens.dbSNP131, keys=rsids, columns=subst)

## retrieve substitution scores 
subst &lt;- c("IDPMAX", "IDPSNP", "IDQMIN")
select(PolyPhen.Hsapiens.dbSNP131, keys=rsids, columns=subst)

## retrieve the PolyPhen-2 classifiers 
subst &lt;- c("PPH2CLASS", "PPH2PROB", "PPH2FPR", "PPH2TPR", "PPH2FDR")
select(PolyPhen.Hsapiens.dbSNP131, keys=rsids, columns=subst)

## duplicate groups of rsid's
duplicateRSID(PolyPhen.Hsapiens.dbSNP131, c("rs71225486", "rs1063796"))
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
