<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: PROVEANDb objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for PROVEANDb-class {VariantAnnotation}"><tr><td>PROVEANDb-class {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>PROVEANDb objects</h2>

<h3>Description</h3>

<p>The PROVEANDb class is a container for storing a connection to a PROVEAN 
sqlite database.
</p>


<h3>Details</h3>

<p>The SIFT tool is no longer actively maintained. A few of the
orginal authors have started the PROVEAN (Protein Variation
Effect Analyzer) project. PROVEAN is a software tool which predicts
whether an amino acid substitution or indel has an impact on the
biological function of a protein. PROVEAN is useful for filtering
sequence variants to identify nonsynonymous or indel variants that
are predicted to be functionally important.
</p>
<p>See the web pages for a complete description of the methods.
</p>

<ul>
<li><p> PROVEAN Home: <a href="http://provean.jcvi.org/index.php/">http://provean.jcvi.org/index.php/</a>
</p>
</li>
<li><p> SIFT Home: <a href="http://sift.jcvi.org/">http://sift.jcvi.org/</a>
</p>
</li></ul>

<p>Though SIFT is not under active development, the PROVEAN team still
provids the SIFT scores in the pre-computed downloads. This package,
<code>SIFT.Hsapiens.dbSNP137</code>, contains both SIFT and PROVEAN scores.
One notable difference between this and the previous SIFT database
package is that <code>keys</code> in <code>SIFT.Hsapiens.dbSNP132</code> are
rs IDs whereas in <code>SIFT.Hsapiens.dbSNP137</code> they are NCBI dbSNP IDs.
</p>


<h3>Methods</h3>

<p>In the code below, <code>x</code> is a <code>PROVEANDb</code> object.
</p>

<dl>
<dt></dt><dd>
<p><code>metadata(x)</code>:
Returns <code>x</code>'s metadata in a data frame.
</p>
</dd>
<dt></dt><dd>
<p><code>columns(x)</code>:
Returns the names of the <code>columns</code> that can be used to subset the
data columns.
</p>
</dd>
<dt></dt><dd>
<p><code>keys(x, keytype="DBSNPID", ...)</code>:
Returns the names of the <code>keys</code> that can be used to subset the
data rows. For SIFT.Hsapiens.dbSNP137 the <code>keys</code> are NCBI dbSNP ids.
</p>
</dd>
<dt></dt><dd>
<p><code>keytypes(x)</code>:
Returns the names of the <code>columns</code> that can be used as <code>keys</code>.
For SIFT.Hsapiens.dbSNP137 the NCBI dbSNP ids are the only keytype.
</p>
</dd>
<dt></dt><dd>
<p><code>select(x, keys = NULL, columns = NULL, keytype = "DBSNPID", ...)</code>:
Returns a subset of data defined by the character vectors <code>keys</code> 
and <code>columns</code>. If no <code>keys</code> are supplied, all rows are
returned. If no <code>columns</code> are supplied, all columns
are returned.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Valerie Obenchain</p>


<h3>References</h3>

<p>The PROVEAN tool has replaced SIFT:
<a href="http://provean.jcvi.org/about.php">http://provean.jcvi.org/about.php</a>
</p>
<p>Choi Y, Sims GE, Murphy S, Miller JR, Chan AP (2012) Predicting the
Functional Effect of Amino Acid Substitutions and Indels.
PLoS ONE 7(10): e46688.
</p>
<p>Choi Y (2012) A Fast Computation of Pairwise Sequence Alignment Scores
Between a Protein and a Set of Single-Locus Variants of Another Protein.
In Proceedings of the ACM Conference on Bioinformatics,
Computational Biology and Biomedicine (BCB '12). ACM, New York, NY, USA,
414-417.
</p>
<p>Kumar P, Henikoff S, Ng PC. Predicting the effects of coding non-synonymous
variants on protein function using the SIFT algorithm. Nat Protoc.
2009;4(7):1073-81
</p>
<p>Ng PC, Henikoff S. Predicting the Effects of Amino Acid Substitutions on
Protein Function Annu Rev Genomics Hum Genet. 2006;7:61-80.
</p>
<p>Ng PC, Henikoff S. SIFT: predicting amino acid changes that affect protein
function. Nucleic Acids Res. 2003 Jul 1;31(13):3812-4.
</p>


<h3>Examples</h3>

<pre>
  if (require(SIFT.Hsapiens.dbSNP137)) {
      ## metadata
      metadata(SIFT.Hsapiens.dbSNP137)

      ## keys are the DBSNPID (NCBI dbSNP ID)
      dbsnp &lt;- keys(SIFT.Hsapiens.dbSNP137)
      head(dbsnp)
      columns(SIFT.Hsapiens.dbSNP137)

      ## Return all columns. Note that the key, DBSNPID,
      ## is always returned. 
      select(SIFT.Hsapiens.dbSNP137, dbsnp[10])
      ## subset on keys and cols 
      cols &lt;- c("VARIANT", "PROVEANPRED", "SIFTPRED")
      select(SIFT.Hsapiens.dbSNP137, dbsnp[20:23], cols)
  }
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
