<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Import VCF files</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for scanVcf {VariantAnnotation}"><tr><td>scanVcf {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Import VCF files
</h2>

<h3>Description</h3>

<p>Import Variant Call Format (VCF) files in text or binary format
</p>


<h3>Usage</h3>

<pre>
scanVcfHeader(file, ...)
## S4 method for signature 'character'
scanVcfHeader(file, ...)

scanVcf(file, ..., param)
## S4 method for signature 'character,ScanVcfParam'
scanVcf(file, ..., param)
## S4 method for signature 'character,missing'
scanVcf(file, ..., param)
## S4 method for signature 'connection,missing'
scanVcf(file, ..., param)

## S4 method for signature 'TabixFile'
scanVcfHeader(file, ...)
## S4 method for signature 'TabixFile,missing'
scanVcf(file, ..., param)
## S4 method for signature 'TabixFile,ScanVcfParam'
scanVcf(file, ..., param)
## S4 method for signature 'TabixFile,GRanges'
scanVcf(file, ..., param)
## S4 method for signature 'TabixFile,IntegerRangesList'
scanVcf(file, ..., param)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>file</code></td>
<td>
<p>For <code>scanVcf</code> and <code>scanVcfHeader</code>, the character() 
file name, <code><a href="../../Rsamtools/html/TabixFile-class.html">TabixFile</a></code>, or class <code>connection</code> 
(<code>file()</code> or <code>bgzip()</code>) of the &lsquo;VCF&rsquo; file to be 
processed.
</p>
</td></tr>
<tr valign="top"><td><code>param</code></td>
<td>
<p>A instance of <code><a href="ScanVcfParam-class.html">ScanVcfParam</a></code> influencing 
which records are parsed and the &lsquo;INFO&rsquo; and &lsquo;GENO&rsquo; information
returned.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments for methods
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The argument <code>param</code> allows portions of the file to be input, but 
requires that the file be bgzip'd and indexed as a 
<code><a href="../../Rsamtools/html/TabixFile-class.html">TabixFile</a></code>.
</p>
<p><code>scanVcf</code> with <code>param="missing"</code> and <code>file="character"</code>
or <code>file="connection"</code> scan the entire file. With
<code>file="connection"</code>, an argument <code>n</code> indicates the number of
lines of the VCF file to input; a connection open at the beginning of
the call is open and incremented by <code>n</code> lines at the end of the
call, providing a convenient way to stream through large VCF files.
</p>
<p>The INFO field of the scanned VCF file is returned as a single
&lsquo;packed&rsquo; vector, as in the VCF file. The GENO field is a list of 
matrices, each matrix corresponds to a field as defined in the FORMAT 
field of the VCF header. Each matrix has as many rows as
scanned in the VCF file, and as many columns as there are samples. As
with the INFO field, the elements of the matrix are
&lsquo;packed&rsquo;. The reason that INFO and GENO are returned packed is
to facilitate manipulation, e.g., selecting particular rows or
samples in a consistent manner across elements.
</p>


<h3>Value</h3>

<p><code>scanVcfHeader</code> returns a <code>VCFHeader</code> object with
header information parsed into five categories, <code>samples</code>, 
<code>meta</code>, <code>fixed</code>, <code>info</code> and <code>geno</code>. Each
can be accessed with a &lsquo;getter&rsquo; of the same name
(e.g., info(&lt;VCFHeader&gt;)). If the file header has multiple rows 
with the same name (e.g., 'source') the row names of the DataFrame
are made unique in the usual way, 'source', 'source.1' etc. 
</p>
<p><code>scanVcf</code> returns a list, with one element per range. Each list
has 7 elements, obtained from the columns of the VCF specification:
</p>

<dl>
<dt>rowRanges</dt><dd>
<p><code>GRanges</code> instance derived from <code>CHROM</code>, <code>POS</code>, 
<code>ID</code>, and the width of <code>REF</code>
</p>
</dd>
<dt>REF</dt><dd>
<p>reference allele
</p>
</dd>
<dt>ALT</dt><dd>
<p>alternate allele
</p>
</dd>
<dt>QUAL</dt><dd>
<p>phred-scaled quality score for the assertion made in ALT
</p>
</dd>
<dt>FILTER</dt><dd>
<p>indicator of whether or not the position passed all filters applied
</p>
</dd>
<dt>INFO</dt><dd>
<p>additional information
</p>
</dd>
<dt>GENO</dt><dd>
<p>genotype information immediately following the FORMAT field in the VCF
</p>
</dd>
</dl>

<p>The <code>GENO</code> element is itself a list, with elements corresponding
to those defined in the VCF file header. For <code>scanVcf</code>, elements
of GENO are returned as a matrix of records x samples; if the
description of the element in the file header indicated multiplicity
other than 1 (e.g., variable number for &ldquo;A&rdquo;, &ldquo;G&rdquo;, or
&ldquo;.&rdquo;), then each entry in the matrix is a character string with
sub-entries comma-delimited.
</p>


<h3>Author(s)</h3>

<p>Martin Morgan and Valerie Obenchain&gt;
</p>


<h3>References</h3>

<p><a href="http://vcftools.sourceforge.net/specs.html">http://vcftools.sourceforge.net/specs.html</a> outlines the VCF
specification.
</p>
<p><a href="http://samtools.sourceforge.net/mpileup.shtml">http://samtools.sourceforge.net/mpileup.shtml</a> contains
information on the portion of the specification implemented by
<code>bcftools</code>.
</p>
<p><a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a> provides information on
<code>samtools</code>.
</p>


<h3>See Also</h3>

<p><code><a href="readVcf-methods.html">readVcf</a></code> 
<code><a href="../../Rsamtools/html/BcfFile-class.html">BcfFile</a></code> 
<code><a href="../../Rsamtools/html/TabixFile-class.html">TabixFile</a></code>
</p>


<h3>Examples</h3>

<pre>
  fl &lt;- system.file("extdata", "ex2.vcf", package="VariantAnnotation")
  scanVcfHeader(fl)
  vcf &lt;- scanVcf(fl)
  ## value: list-of-lists
  str(vcf)
  names(vcf[[1]][["GENO"]])
  vcf[[1]][["GENO"]][["GT"]]
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
