<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: SIFTDb objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for SIFTDb-class {VariantAnnotation}"><tr><td>SIFTDb-class {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>SIFTDb objects</h2>

<h3>Description</h3>

<p>The SIFTDb class is a container for storing a connection to a SIFT 
sqlite database.
</p>


<h3>Details</h3>

<p>SIFT is a sequence homology-based tool that sorts intolerant from tolerant 
amino acid substitutions and predicts whether an amino acid substitution 
in a protein will have a phenotypic effect. SIFT is based on the premise 
that protein evolution is correlated with protein function. Positions 
important for function should be conserved in an alignment of the protein 
family, whereas unimportant positions should appear diverse in an alignment.
</p>
<p>SIFT uses multiple alignment information to predict tolerated 
and deleterious substitutions for every position of the query sequence. 
The procedure can be outlined in the following steps, 
</p>

<ul>
<li><p> search for similar sequences
</p>
</li>
<li><p> choose closely related sequences that may share similar
function to the query sequence
</p>
</li>
<li><p> obtain the alignment of the chosen sequences
</p>
</li>
<li><p> calculate normalized probabilities for all possible
substitutions from the alignment.
</p>
</li></ul>
<p> Positions with normalized probabilities less than 0.05 are predicted
to be deleterious, those greater than or equal to 0.05 are predicted to be
tolerated.
</p>


<h3>Methods</h3>

<p>In the code below, <code>x</code> is a <code>SIFTDb</code> object.
</p>

<dl>
<dt></dt><dd>
<p><code>metadata(x)</code>:
Returns <code>x</code>'s metadata in a data frame.
</p>
</dd>
<dt></dt><dd>
<p><code>columns(x)</code>:
Returns the names of the <code>columns</code> that can be used to subset the
data columns.
</p>
</dd>
<dt></dt><dd>
<p><code>keys(x)</code>:
Returns the names of the <code>keys</code> that can be used to subset the
data rows. The <code>keys</code> values are the rsid's.
</p>
</dd>
<dt></dt><dd>
<p><code>select(x, keys = NULL, columns = NULL, ...)</code>:
Returns a subset of data defined by the character vectors <code>keys</code> 
and <code>columns</code>. If no <code>keys</code> are supplied, all rows are
returned. If no <code>columns</code> are supplied, all columns
are returned. For column descriptions see <code>?SIFTDbColumns</code>.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Valerie Obenchain</p>


<h3>References</h3>

<p>SIFT Home:
<a href="http://sift.jcvi.org/">http://sift.jcvi.org/</a>
</p>
<p>Kumar P, Henikoff S, Ng PC. Predicting the effects of coding non-synonymous
variants on protein function using the SIFT algorithm. Nat Protoc.
2009;4(7):1073-81
</p>
<p>Ng PC, Henikoff S. Predicting the Effects of Amino Acid Substitutions on
Protein Function Annu Rev Genomics Hum Genet. 2006;7:61-80.
</p>
<p>Ng PC, Henikoff S. SIFT: predicting amino acid changes that affect protein
function. Nucleic Acids Res. 2003 Jul 1;31(13):3812-4.
</p>


<h3>Examples</h3>

<pre>
if (interactive()) {
    library(SIFT.Hsapiens.dbSNP132)
    
    ## metadata
    metadata(SIFT.Hsapiens.dbSNP132)
    
    ## available rsid's 
    head(keys(SIFT.Hsapiens.dbSNP132))
    
    ## for column descriptions see ?SIFTDbColumns
    columns(SIFT.Hsapiens.dbSNP132)
    
    ## subset on keys and columns 
    rsids &lt;- c("rs2142947", "rs17970171", "rs8692231", "rs3026284") 
    subst &lt;- c("RSID", "PREDICTION", "SCORE")
    select(SIFT.Hsapiens.dbSNP132, keys=rsids, columns=subst)
    select(SIFT.Hsapiens.dbSNP132, keys=rsids[1:2])
}
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
