<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Counts and distribution statistics for SNPs in a VCF object</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for snpSummary {VariantAnnotation}"><tr><td>snpSummary {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Counts and distribution statistics for SNPs in a VCF object</h2>

<h3>Description</h3>

<p>Counts and distribution statistics for SNPs in a VCF object
</p>


<h3>Usage</h3>

<pre>
  ## S4 method for signature 'CollapsedVCF'
snpSummary(x, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="VCF-class.html">CollapsedVCF</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments to methods.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Genotype counts, allele counts and Hardy Weinberg equilibrium
(HWE) statistics are calculated for single nucleotide variants
in a <a href="VCF-class.html">CollapsedVCF</a> object. HWE has been established as a 
useful quality filter on genotype data. This equilibrium should 
be attained in a single generation of random mating. Departures
from HWE are indicated by small p values and are almost invariably 
indicative of a problem with genotype calls.
</p>
<p>The following caveats apply:
</p>

<ul>
<li><p> No distinction is made between phased and unphased genotypes. 
</p>
</li>
<li><p> Only diploid calls are included.
</p>
</li>
<li><p> Only &lsquo;valid&rsquo; SNPs are included. A &lsquo;valid&rsquo; SNP is defined
as having a reference allele of length 1 and a single 
alternate allele of length 1.
</p>
</li></ul>

<p>Variants that do not meet these criteria are set to NA. 
</p>


<h3>Value</h3>

<p>The object returned is a <code>data.frame</code> with seven columns.
</p>

<dl>
<dt>g00</dt><dd>
<p>Counts for genotype 00 (homozygous reference).
</p>
</dd>
<dt>g01</dt><dd>
<p>Counts for genotype 01 or 10 (heterozygous).
</p>
</dd>
<dt>g11</dt><dd>
<p>Counts for genotype 11 (homozygous alternate).
</p>
</dd>
<dt>a0Freq</dt><dd>
<p>Frequency of the reference allele.
</p>
</dd>
<dt>a1Freq</dt><dd>
<p>Frequency of the alternate allele.
</p>
</dd>
<dt>HWEzscore</dt><dd>
<p>Z-score for departure from a null hypothesis of Hardy Weinberg equilibrium.
</p>
</dd>
<dt>HWEpvalue</dt><dd>
<p>p-value for departure from a null hypothesis of Hardy Weinberg equilibrium.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Chris Wallace &lt;cew54@cam.ac.uk&gt;
</p>


<h3>See Also</h3>

<p><a href="genotypeToSnpMatrix-methods.html">genotypeToSnpMatrix</a>,
<a href="probabilityToSnpMatrix.html">probabilityToSnpMatrix</a>
</p>


<h3>Examples</h3>

<pre>
  fl &lt;- system.file("extdata", "ex2.vcf", package="VariantAnnotation")
  vcf &lt;- readVcf(fl, "hg19")

  ## The return value is a data.frame with genotype counts
  ## and allele frequencies.
  df &lt;- snpSummary(vcf)
  df

  ## Compare to ranges in the VCF object:
  rowRanges(vcf)

  ## No statistics were computed for the variants in rows 3, 4 
  ## and 5. They were omitted because row 3 has two alternate 
  ## alleles, row 4 has none and row 5 is not a SNP.
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
