<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: VariantType subclasses</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for VariantType-class {VariantAnnotation}"><tr><td>VariantType-class {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>VariantType subclasses</h2>

<h3>Description</h3>

<p><code>VariantType</code> subclasses specify the type of variant to be located with
<code>locateVariants</code>.
</p>


<h3>Usage</h3>

<pre>
    CodingVariants()
    IntronVariants()
    FiveUTRVariants()
    ThreeUTRVariants()
    SpliceSiteVariants()
    IntergenicVariants(upstream = 1e+06L, downstream = 1e+06L,
                       idType=c("gene", "tx"))
    PromoterVariants(upstream = 2000L, downstream = 200L)
    AllVariants(promoter = PromoterVariants(),
                intergenic = IntergenicVariants())
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>upstream, downstream</code></td>
<td>

<p>Single <code>integer</code> values representing the number of base pairs 
upstream of the 5'-end and downstream of the 3'-end. Used in contructing 
<code>PromoterVariants()</code> and <code>IntergenicVariants()</code> objects only.
</p>
</td></tr>
<tr valign="top"><td><code>idType</code></td>
<td>

<p><code>character</code> indicating if the ids in the PRECEDEID and FOLLOWID 
metadata columns should be gene ids (&quot;gene&quot;) or transcript ids (&quot;tx&quot;). 
Applicable to <code>IntergenicVariants()</code> objects only.
</p>
</td></tr>
<tr valign="top"><td><code>promoter</code></td>
<td>

<p><code>PromoterVariants</code> object with appropriate <code>upstream</code> and 
<code>downstream</code> values. Used when constructing <code>AllVariants</code> 
objects only.
</p>
</td></tr>
<tr valign="top"><td><code>intergenic</code></td>
<td>

<p><code>IntergenicVariants</code> object with appropriate <code>upstream</code> and 
<code>downstream</code> values. Used when constructing <code>AllVariants</code> 
objects only.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>VariantType</code> is a virtual class inherited by the <code>CodingVariants</code>,
<code>IntronVariants</code>, <code>FiveUTRVariants</code>, <code>ThreeUTRVariants</code>,
<code>SpliceSiteVariants</code>, <code>IntergenicVariants</code> and <code>AllVariants</code>
subclasses.
</p>
<p>The subclasses are used as the <code>region</code> argument to
<code>locateVariants</code>. They designate the type of variant (i.e., region of
the annotation to match) when calling <code>locateVariants</code>. 
</p>
<p>The majority of subclasses have no slots and require no arguments for an 
instance to be created. <code>PromoterVariants</code> and <code>IntergenicVariants</code> 
and accept <code>upstream</code> and <code>downstream</code> arguments that define
the number of base pairs upstream from the 5'-end and downstream from
the 3'-end of the transcript region. See the ?<code>locateVariants</code> man
page for details. <code>IntergenicVariants</code> also accepts a
<code>idType</code> that controls what IDs are returned in the
PRECEDEID and FOLLOWID metadata columns.
</p>
<p><code>AllVariants</code> accepts <code>promoter</code> and
<code>intergenic</code> arguments which are <code>PromoterVariants()</code> and
<code>IntergenicVariants()</code> objects with the appropriate 
<code>upstream</code> and <code>downstream</code> values. 
</p>


<h3>Accessors</h3>

<p>In the following code, <code>x</code> is a <code>PromoterVariants</code> or a
<code>AllVariants</code> object.
</p>

<dl>
<dt></dt><dd>
<p><code>upstream(x)</code>, <code>upstream(x) &lt;- value</code>:
Gets or sets the number of base pairs defining a range
upstream of the 5' end (excludes 5' start value).
</p>
</dd>
<dt></dt><dd>
<p><code>downstream(x)</code>, <code>downstream(x) &lt;- value</code>:
Gets or sets the number of base pairs defining a range
downstream of the 3' end (excludes 3' end value).
</p>
</dd>
<dt></dt><dd>
<p><code>idType(x)</code>, <code>idType(x) &lt;- value</code>:
Gets or sets the <code>character()</code> which controls the id returned 
in the PRECEDEID and FOLLOWID output columns. Possible values are
&quot;gene&quot; and &quot;tx&quot;.
</p>
</dd>
<dt></dt><dd>
<p><code>promoters(x)</code>, <code>promoters(x) &lt;- value</code>:
Gets or sets the <code>PromoterVariants</code> in the 
<code>AllVariants</code> object.
</p>
</dd>
<dt></dt><dd>
<p><code>intergenic(x)</code>, <code>intergenic(x) &lt;- value</code>:
Gets or sets the <code>IntergenicVariants</code> in the 
<code>AllVariants</code> object.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Valerie Obenchain</p>


<h3>See Also</h3>


<ul>
<li><p> The promoters function on the 
<a href="../../GenomicRanges/html/intra-range-methods.html">intra-range-methods</a> man page in the 
GenomicRanges package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
  CodingVariants()
  SpliceSiteVariants()
  PromoterVariants(upstream=1000, downstream=10000)

  ## Default values for PromoterVariants and IntergenicVariants
  AllVariants()
  ## Modify 'upstream' and 'downstream' for IntergenicVariants
  AllVariants(intergenic=IntergenicVariants(500, 100))
  ## Reset PromoterVariants on existing AllVariants object
  av &lt;- AllVariants()
  av
  promoter(av) &lt;- PromoterVariants(100, 50)
  av
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
