<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: VRanges objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for VRanges-class {VariantAnnotation}"><tr><td>VRanges-class {VariantAnnotation}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>VRanges objects</h2>

<h3>Description</h3>

<p>The VRanges class is a container for variant calls, including SNVs and
indels. It extends <code><a href="../../GenomicRanges/html/GRanges.html">GRanges</a></code> to provide
special semantics on top of a simple vector of genomic locations. While
it is not as expressive as the <code><a href="VCF-class.html">VCF</a></code> object, it is
a simpler alternative that may be convenient for variant
calling/filtering and similar exercises.
</p>


<h3>Details</h3>

<p>VRanges extends GRanges to store the following components. Except
where noted, the components are considered columns in the dataset,
i.e., their lengths match the number of variants. Many columns can be
stored as either an atomic vector or an Rle. 
</p>

<dl>
<dt><code>ref</code></dt><dd><p>(<code>character</code>), the reference
allele. The range (start/end/width) should always correspond to
this sequence.</p>
</dd>
<dt><code>alt</code></dt><dd><p>(<code>character/Rle</code>),
the alternative allele (NA allowed). By convention there is only
a single alt allele per element (row) of the VRanges. Many methods,
like <code>match</code>, make this assumption. 
</p>
</dd>
<dt><code>refCount</code></dt><dd><p>(<code>integer/Rle</code>), read count for the
reference allele (NA allowed)</p>
</dd>
<dt><code>altCount</code></dt><dd><p>(<code>integer/Rle</code>), read count for the
alternative allele (NA allowed)</p>
</dd>
<dt><code>totalCount</code></dt><dd><p>(<code>integer/Rle</code>), total read count at the
position, must be at least <code>refCount+altCount</code> (NA allowed)</p>
</dd>
<dt><code>sampleNames</code></dt><dd><p>(<code>factor/Rle</code>), name of the sample -
results from multiple samplse can be combined into the same object
(NA allowed)</p>
</dd>
<dt><code>softFilterMatrix</code></dt><dd><p>(<code>matrix/FilterMatrix</code>),
variant by filter matrix, <code>TRUE</code> where variant passed the
filter; use a <code><a href="../../S4Vectors/html/FilterMatrix.html">FilterMatrix</a></code> to store the
actual <code>FilterRules</code> object that was applied</p>
</dd>
<dt><code>hardFilters</code></dt><dd><p>(<code>FilterRules</code>) record of hard
filters applied, i.e., only the variants that passed the filters
are present in this object; this is the only component that is not
a column, i.e., its length does not match the number of variants</p>
</dd>
</dl>

<p>Except in the special circumstances described here, a <code>VRanges</code>
may be treated like a <code>GRanges</code>. The range should span the
sequence in <code>ref</code>. Indels are typically represented by the VCF
convention, i.e., the start position is one upstream of the event. The
strand is always constrained to be positive (+).
</p>
<p>Indels, by convention, should be encoded VCF-style, with the upstream
reference base prepended to the indel sequence. The ref/alt for a
deletion of GCGT before A might be AGCGT/A and for an insertion might
be A/AGCGT. Since the range always matches the <code>ref</code> sequence,
this means a deletion will be the width of the deletion + 1, and an
insertion is always of width 1.
</p>
<p>VRanges and the VCF class:
The VRanges and VCF classes encode different types of information and
are semantically incompatible. While methods exist for converting
a VCF object to a VRanges and vice versa, information is lost in the
transformation. There is no way to collapse multiple rows of a VRanges
at the same genomic position and accurately represent missing data.
For this reason, it is not reasonable to assume that an object resulting
from multiple conversions (VRanges -&gt; VCF -&gt; VRanges) will be equivalent to 
the original.
</p>


<h3>Constructors</h3>


<dl>
<dt></dt><dd>
<p><code>VRanges(seqnames = Rle(), ranges = IRanges(), ref = character(), 
        alt = NA_character_, totalDepth = NA_integer_, refDepth = NA_integer_, 
        altDepth = NA_integer_, ..., sampleNames = NA_character_, 
        softFilterMatrix = FilterMatrix(matrix(nrow = length(gr), 
        ncol = 0L), FilterRules()), hardFilters = FilterRules())</code>:
Creates a VRanges object.
</p>

<dl>
<dt><code>seqnames</code></dt><dd><p>Rle object, character vector, or factor
containing the sequence names.</p>
</dd>
<dt><code>ranges</code></dt><dd><p>IRanges object containing the ranges.</p>
</dd>
<dt><code>ref</code></dt><dd><p>character vector, containing the reference allele.</p>
</dd>
<dt><code>alt</code></dt><dd><p>character vector/Rle,
containing the alternative allele (NA allowed).</p>
</dd>
<dt><code>totalDepth</code></dt><dd><p>integer vector/Rle, containing the
total read depth (NA allowed).</p>
</dd>
<dt><code>refDepth</code></dt><dd><p>integer vector/Rle, containing the
reference read depth (NA allowed).</p>
</dd>
<dt><code>altDepth</code></dt><dd><p>integer vector/Rle, containing the
reference read depth (NA allowed).</p>
</dd>
<dt><code>...</code></dt><dd><p>Arguments passed to the <code>GRanges</code>
constructor.</p>
</dd>
<dt><code>sampleNames</code></dt><dd><p>character/factor vector/Rle, containing the
sample names (NA allowed).</p>
</dd>
<dt><code>softFilterMatrix</code></dt><dd><p>a matrix (typically
a <code><a href="../../S4Vectors/html/FilterMatrix.html">FilterMatrix</a></code>) of dimension variant by
filter, with logical values indicating whether a variant
passed the filter.</p>
</dd>
<dt><code>hardFilters</code></dt><dd><p>a <code><a href="../../S4Vectors/html/FilterRules.html">FilterRules</a></code>,
containing the filters that have already been applied to
subset the object to its current state.</p>
</dd>
</dl>

</dd>
<dt></dt><dd>
<p><code>makeVRangesFromGRanges(gr, 
                                   ref.field="ref",
                                   alt.field="alt",
                                   totalDepth.field="totalDepth", 
                                   refDepth.field="refDepth", 
                                   altDepth.field="altDepth", 
                                   sampleNames.field="sampleNames",
                                   keep.extra.columns=TRUE)</code>:
Creates a VRanges object from a GRanges.
</p>

<dl>
<dt><code>gr</code></dt><dd>
<p>A <a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> object.
</p>
</dd>
<dt><code>ref.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>ref</code> field.
</p>
</dd>
<dt><code>alt.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>alt</code> field.
</p>
</dd>
<dt><code>totalDepth.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>totalDepth</code> field.
</p>
</dd>
<dt><code>refDepth.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>refDepth</code> field.
</p>
</dd>
<dt><code>altDepth.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>altDepth</code> field.
</p>
</dd>
<dt><code>sampleNames.field</code></dt><dd>
<p>The <code>character(1)</code> name of the GRanges metadata column to be 
used as the VRanges <code>sampleNames</code> field.
</p>
</dd>
<dt><code>keep.extra.columns</code></dt><dd>
<p><code>TRUE</code> (the default) or <code>FALSE</code>.
If <code>TRUE</code>, then the columns in <code>gr</code> that are not used to 
form the VRanges are retained as metadata columns. Otherwise, they 
will be ignored.
</p>
</dd>
</dl>

</dd>
</dl>



<h3>Coercion</h3>

<p>These functions/methods coerce objects to and from <code>VRanges</code>:
</p>

<dl>
<dt></dt><dd>
<p><code>asVCF(x, info = character(), filter = character(), meta =
        character())</code>: Creates a VCF object from a VRanges object. The
following gives the mapping from VRanges components to VCF:
</p>

<dl>
<dt>seqnames(x)</dt><dd><p>CHROM column</p>
</dd>
<dt>start(x)</dt><dd><p>POS column</p>
</dd>
<dt>names(x)</dt><dd><p>ID column</p>
</dd>
<dt>ref(x)</dt><dd><p>REF column</p>
</dd>
<dt>alt(x)</dt><dd><p>ALT column</p>
</dd>
<dt>totalDepth(x)</dt><dd><p>DP in FORMAT column</p>
</dd>
<dt>altDepth(x), refDepth(x)</dt><dd><p>AD in FORMAT column</p>
</dd>
<dt>sampleNames(x)</dt><dd><p>Names the sample columns</p>
</dd>
<dt>softFilterMatrix(x)</dt><dd><p>FT in FORMAT column, except filters
named in <code>filter</code> argument, which are considered
per-position and placed in the FILTER column</p>
</dd>
<dt>hardFilters(x)</dt><dd><p>Not yet exported</p>
</dd>
<dt>mcols(x)</dt><dd><p>Become fields in the FORMAT column; unless they
are named in the <code>info</code> argument, in which case they
are considered per-position and placed in the INFO column</p>
</dd>
<dt>metadata(x)</dt><dd><p>If named in the <code>meta</code> argument, output
in the VCF header; a component is required to be coercible to
a character vector of length one.</p>
</dd>
</dl>

<p>Note that <code>identical(x, as(as(x, "VCF"), "VRanges"))</code>
generally return <code>FALSE</code>.  During coercion to VCF, the &quot;geno&quot;
components are reshaped into matrix form, with NAs filling the
empty cells. The reverse coercion will not drop the NA values, so
rows are added to the new VRanges. All logical values will become
integers in VCF, and there is no automatic way of regenerating the
logical column with the reverse coercion. There are many other
cases of irreversibility.
</p>
</dd>
<dt></dt><dd>
<p><code>as(from, "VCF")</code>: Like calling <code>asVCF(from)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>as(from, "VRanges")</code>:
When <code>from</code> is a <code>VCF</code> this coercion is essentially
the inverse of <code>asVCF</code>. Information missing in the VCF
is imputed as NA.
</p>
<p>When <code>from</code> is a <code>GRanges</code>, metadata columns of
<code>ref</code>, <code>alt</code>, <code>refDepth</code>, <code>altDepth</code>,
<code>totalDepth</code> and <code>sampleNames</code> are transfered to 
the <code>VRanges</code> object. Additional metadata columns in
the <code>GRanges</code> can be retained or dropped with
<code>keep.extra.columns</code>. See also <code>makeVRangesFromGRanges</code>.
</p>
</dd>
</dl>



<h3>Accessors</h3>

<p>In addition to all of the <code>GRanges</code> accessors, <code>VRanges</code>
provides the following, where <code>x</code> is a VRanges object.
</p>

<dl>
<dt></dt><dd>
<p><code>alt(x), alt(x) &lt;- value</code>: Get or set the alt allele (character).
</p>
</dd>
<dt></dt><dd>
<p><code>ref(x), ref(x) &lt;- value</code>: Get or set the ref allele (character).
</p>
</dd>
<dt></dt><dd>
<p><code>altDepth(x), altDepth(x) &lt;- value</code>: Get or set the alt allele
read depth (integer).
</p>
</dd>
<dt></dt><dd>
<p><code>refDepth(x), refDepth(x) &lt;- value</code>: Get or set the ref
allele read depth (integer).
</p>
</dd>
<dt></dt><dd>
<p><code>totalDepth(x), totalDepth(x) &lt;- value</code>: Get or set the total
read depth (integer).
</p>
</dd>
<dt></dt><dd>
<p><code>altFraction(x)</code>: Returns <code>altDepth(x)/totalDepth(x)</code> (numeric).
</p>
</dd>
<dt></dt><dd>
<p><code>sampleNames(x), sampleNames(x) &lt;- value</code>: Get or set the
sample names (character/factor).
</p>
</dd>
<dt></dt><dd>
<p><code>softFilterMatrix(x), softFilterMatrix(x) &lt;- value</code>: Gets or
sets the soft filter matrix (any matrix, but ideally a
<code>FilterMatrix</code>).
</p>
</dd>
<dt></dt><dd>
<p><code>resetFilter(x)</code>: Removes all columns from <code>softFilterMatrix</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>called(x)</code>: Returns whether all filter results in
<code>softFilterMatrix(x)</code> are <code>TRUE</code> for each variant.
</p>
</dd>
<dt></dt><dd>
<p><code>hardFilters(x), hardFilters(x) &lt;- value</code>: Gets or
sets the hard filters (those applied to yield the current subset).
</p>
</dd>
</dl>



<h3>Utilities and Conveniences</h3>


<dl>
<dt></dt><dd>
<p><code>match(x)</code>: Like GRanges <code>match</code>, except matches on the
combination of chromosome, start, width, and <strong>alt</strong>.
</p>
</dd>
<dt></dt><dd>
<p><code>tabulate(bin)</code>: Finds <code>unique(bin)</code> and counts how many
times each unique element occurs in <code>bin</code>. The result is
stored in <code>mcols(bin)$sample.count</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>softFilter(x, filters, ...)</code>: applies the <code>FilterRules</code>
in <code>filters</code> to <code>x</code>, storing the results in
<code>softFilterMatrix</code>.
</p>
</dd>
</dl>
  


<h3>Input/Output to/from VCF</h3>


<dl>
<dt></dt><dd>
<p><code>writeVcf(obj, filename, ...)</code>: coerces to a VCF object and
writes it to a file; see <code><a href="writeVcf-methods.html">writeVcf</a></code>.
</p>
</dd>
<dt></dt><dd>
<p><code>readVcfAsVRanges(x, genome, param = ScanVcfParam(), ...)</code>:
Reads a VCF <code>x</code> directly into a <code>VRanges</code>;
see <code><a href="readVcf-methods.html">readVcf</a></code> for details on the arguments.
<code>readVcfAsVRanges</code> is an alternative syntax to
</p>
<pre>  as(readVcf(), "VRanges") </pre>
<p>NOTE: By default all INFO and FORMAT fields are read in with
<code>ScanVcfParam()</code>. The minimal information needed to create 
the <code>VRanges</code> can be specified as follows: 
</p>
<pre>  ScanVcfParam(fixed = "ALT", info = NA, geno = "AD")) </pre>
</dd>
</dl>



<h3>Variant Type</h3>

<p>Functions to identify variant type include <a href="isSNV-methods.html">isSNV</a>, 
<a href="isSNV-methods.html">isInsertion</a>, <a href="isSNV-methods.html">isDeletion</a>, <a href="isSNV-methods.html">isIndel</a>, 
<a href="isSNV-methods.html">isSubstitution</a> and <a href="isSNV-methods.html">isTransition</a>. See the ?<code>isSNV</code> 
man page for details.
</p>


<h3>Author(s)</h3>

<p>Michael Lawrence. <code>makeVRangesFromGRanges</code> was contributed 
by Thomas Sandmann.</p>


<h3>See Also</h3>

<p><a href="VRangesList-class.html">VRangesList</a>, a list of <code>VRanges</code>; <code>bam_tally</code> in the
gmapR package, which generates a <code>VRanges</code>.
</p>


<h3>Examples</h3>

<pre>
## construction
vr &lt;- VRanges(seqnames = c("chr1", "chr2"),
              ranges = IRanges(c(1, 10), c(5, 20)),
              ref = c("T", "A"), alt = c("C", "T"),
              refDepth = c(5, 10), altDepth = c(7, 6),
              totalDepth = c(12, 17), sampleNames = letters[1:2],
              hardFilters =
                FilterRules(list(coverage = function(x) totalDepth &gt; 10)),
              softFilterMatrix =
                FilterMatrix(matrix = cbind(depth = c(TRUE, FALSE)),
                             FilterRules(depth = function(x) altDepth(x) &gt; 6)),
              tumorSpecific = c(FALSE, TRUE))

## simple accessors
ref(vr)
alt(vr)
altDepth(vr)
vr$tumorSpecific
called(vr)

## coerce to VCF and write
vcf &lt;- as(vr, "VCF")
## writeVcf(vcf, "example.vcf")
## or just
## writeVcf(vr, "example.vcf")

## other utilities
match(vr, vr[2:1])
</pre>

<hr /><div style="text-align: center;">[Package <em>VariantAnnotation</em> version 1.36.0 <a href="00Index.html">Index</a>]</div>
</body></html>
