<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Filters for annotation objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for AnnotationFilter {AnnotationFilter}"><tr><td>AnnotationFilter {AnnotationFilter}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Filters for annotation objects</h2>

<h3>Description</h3>

<p>The filters extending the base <code>AnnotationFilter</code> class
represent a simple filtering concept for annotation resources.
Each filter object is thought to filter on a single (database)
table column using the provided values and the defined condition.
</p>
<p>Filter instances created using the constructor functions (e.g.
<code>GeneIdFilter</code>).
</p>
<p><code>supportedFilters()</code> lists all defined filters. It returns a two column
<code>data.frame</code> with the filter class name and its default field.
Packages using <code>AnnotationFilter</code> should implement the
<code>supportedFilters</code> for their annotation resource object (e.g. for
<code>object = "EnsDb"</code> in the <code>ensembldb</code> package) to list all
supported filters for the specific resource.
</p>
<p><code>condition()</code> get the <code>condition</code> value for
the filter <code>object</code>.
</p>
<p><code>value()</code> get the <code>value</code> for the filter
<code>object</code>.
</p>
<p><code>field()</code> get the <code>field</code> for the filter
<code>object</code>.
</p>
<p><code>not()</code> get the <code>not</code> for the filter <code>object</code>.
</p>
<p><code>feature()</code> get the <code>feature</code> for the
<code>GRangesFilter</code> <code>object</code>.
</p>
<p>Converts an <code>AnnotationFilter</code> object to a 
<code>character(1)</code> giving an equation that can be used as input to
a <code>dplyr</code> filter.
</p>
<p><code>AnnotationFilter</code> <em>translates</em> a filter
expression such as <code>~ gene_id == "BCL2"</code> into a filter object
extending the <code><a href="AnnotationFilter.html">AnnotationFilter</a></code> class (in the example a
<code><a href="AnnotationFilter.html">GeneIdFilter</a></code> object) or an
<code><a href="AnnotationFilterList.html">AnnotationFilterList</a></code> if the expression contains multiple
conditions (see examples below). Filter expressions have to be written
in the form <code>~ &lt;field&gt; &lt;condition&gt; &lt;value&gt;</code>, with <code>&lt;field&gt;</code>
being the default field of the filter class (use the
<code>supportedFilter</code> function to list all fields and filter classes),
<code>&lt;condition&gt;</code> the logical expression and <code>&lt;value&gt;</code> the value
for the filter.
</p>


<h3>Usage</h3>

<pre>
CdsStartFilter(value, condition = "==", not = FALSE)
CdsEndFilter(value, condition = "==", not = FALSE)
ExonIdFilter(value, condition = "==", not = FALSE)
ExonNameFilter(value, condition = "==", not = FALSE)
ExonRankFilter(value, condition = "==", not = FALSE)
ExonStartFilter(value, condition = "==", not = FALSE)
ExonEndFilter(value, condition = "==", not = FALSE)
GeneIdFilter(value, condition = "==", not = FALSE)
GeneNameFilter(value, condition = "==", not = FALSE)
GeneBiotypeFilter(value, condition = "==", not = FALSE)
GeneStartFilter(value, condition = "==", not = FALSE)
GeneEndFilter(value, condition = "==", not = FALSE)
EntrezFilter(value, condition = "==", not = FALSE)
SymbolFilter(value, condition = "==", not = FALSE)
TxIdFilter(value, condition = "==", not = FALSE)
TxNameFilter(value, condition = "==", not = FALSE)
TxBiotypeFilter(value, condition = "==", not = FALSE)
TxStartFilter(value, condition = "==", not = FALSE)
TxEndFilter(value, condition = "==", not = FALSE)
ProteinIdFilter(value, condition = "==", not = FALSE)
UniprotFilter(value, condition = "==", not = FALSE)
SeqNameFilter(value, condition = "==", not = FALSE)
SeqStrandFilter(value, condition = "==", not = FALSE)

## S4 method for signature 'AnnotationFilter'
condition(object)

## S4 method for signature 'AnnotationFilter'
value(object)

## S4 method for signature 'AnnotationFilter'
field(object)

## S4 method for signature 'AnnotationFilter'
not(object)

GRangesFilter(value, feature = "gene", type = c("any", "start", "end",
  "within", "equal"))

feature(object)

## S4 method for signature 'AnnotationFilter,missing'
convertFilter(object)

## S4 method for signature 'missing'
supportedFilters(object)

AnnotationFilter(expr)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p>An <code>AnnotationFilter</code> object.</p>
</td></tr>
<tr valign="top"><td><code>value</code></td>
<td>
<p><code>character()</code>, <code>integer()</code>, or
<code>GRanges()</code> value for the filter</p>
</td></tr>
<tr valign="top"><td><code>feature</code></td>
<td>
<p><code>character(1)</code> defining on what feature the
<code>GRangesFilter</code> should be applied. Choices could be
<code>"gene"</code>, <code>"tx"</code> or <code>"exon"</code>.</p>
</td></tr>
<tr valign="top"><td><code>type</code></td>
<td>
<p><code>character(1)</code> indicating how overlaps are to be
filtered. See <code>findOverlaps</code> in the IRanges package for a
description of this argument.</p>
</td></tr>
<tr valign="top"><td><code>expr</code></td>
<td>
<p>A filter expression, written as a <code>formula</code>, to be
converted to an <code>AnnotationFilter</code> or <code>AnnotationFilterList</code>
class. See below for examples.</p>
</td></tr>
<tr valign="top"><td><code>condition</code></td>
<td>
<p><code>character(1)</code> defining the condition to be
used in the filter. For <code>IntegerFilter</code> or <code>DoubleFilter</code>,
one of <code>"=="</code>, <code>"!="</code>, <code>"&gt;"</code>, <code>"&lt;"</code>, <code>"&gt;="</code>
or <code>"&lt;="</code>. For <code>CharacterFilter</code>, one of <code>"=="</code>,
<code>"!="</code>, <code>"startsWith"</code>, <code>"endsWith"</code> or <code>"contains"</code>.
Default condition is <code>"=="</code>.</p>
</td></tr>
<tr valign="top"><td><code>not</code></td>
<td>
<p><code>logical(1)</code> whether the <code>AnnotationFilter</code> is negated.
<code>TRUE</code> indicates is negated (!). <code>FALSE</code> indicates not
negated. Default not is <code>FALSE</code>.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>By default filters are only available for tables containing the
field on which the filter acts (i.e. that contain a column with the
name matching the value of the <code>field</code> slot of the
object). See the vignette for a description to use filters for
databases in which the database table column name differs from the
default <code>field</code> of the filter.
</p>
<p>Filter expressions for the <code>AnnotationFilter</code> class have to be
written as formulas, i.e. starting with a <code>~</code>.
</p>


<h3>Value</h3>

<p>The constructor function return an object extending
<code>AnnotationFilter</code>. For the return value of the other methods see
the methods' descriptions.
</p>
<p><code>character(1)</code> that can be used as input to a <code>dplyr</code> 
filter.
</p>
<p><code>AnnotationFilter</code> returns an
<code><a href="AnnotationFilter.html">AnnotationFilter</a></code> or an <code><a href="AnnotationFilterList.html">AnnotationFilterList</a></code>.
</p>


<h3>Note</h3>

<p>Translation of nested filter expressions using the
<code>AnnotationFilter</code> function is not yet supported.
</p>


<h3>See Also</h3>

<p><code><a href="AnnotationFilterList.html">AnnotationFilterList</a></code> for combining
<code>AnnotationFilter</code> objects.
</p>


<h3>Examples</h3>

<pre>
## filter by GRanges
GRangesFilter(GenomicRanges::GRanges("chr10:87869000-87876000"))
## Create a SymbolFilter to filter on a gene's symbol.
sf &lt;- SymbolFilter("BCL2")
sf

## Create a GeneStartFilter to filter based on the genes' chromosomal start
## coordinates
gsf &lt;- GeneStartFilter(10000, condition = "&gt;")
gsf

filter &lt;- SymbolFilter("ADA", "==")
result &lt;- convertFilter(filter)
result
supportedFilters()

## Convert a filter expression based on a gene ID to a GeneIdFilter
gnf &lt;- AnnotationFilter(~ gene_id == "BCL2")
gnf

## Same conversion but for two gene IDs.
gnf &lt;- AnnotationFilter(~ gene_id %in% c("BCL2", "BCL2L11"))
gnf

## Converting an expression that combines multiple filters. As a result we
## get an AnnotationFilterList containing the corresponding filters.
## Be aware that nesting of expressions/filters does not work.
flt &lt;- AnnotationFilter(~ gene_id %in% c("BCL2", "BCL2L11") &amp;
                        tx_biotype == "nonsense_mediated_decay" |
                        seq_name == "Y")
flt

</pre>

<hr /><div style="text-align: center;">[Package <em>AnnotationFilter</em> version 1.14.0 <a href="00Index.html">Index</a>]</div>
</body></html>
