<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Coverage of a GAlignments, GAlignmentPairs, or...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for coverage-methods {GenomicAlignments}"><tr><td>coverage-methods {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Coverage of a GAlignments, GAlignmentPairs, or GAlignmentsList object</h2>

<h3>Description</h3>

<p><code><a href="../../IRanges/html/coverage.html">coverage</a></code> methods for <a href="GAlignments-class.html">GAlignments</a>,
<a href="GAlignmentPairs-class.html">GAlignmentPairs</a>, <a href="GAlignmentsList-class.html">GAlignmentsList</a>, and
<a href="../../Rsamtools/html/BamFile.html">BamFile</a> objects.
</p>
<p>NOTE: The <code><a href="../../IRanges/html/coverage.html">coverage</a></code> generic function and methods
for <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> and <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a>
objects are defined and documented in the <span class="pkg">IRanges</span> package.
Methods for <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> and
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> objects are defined and
documented in the <span class="pkg">GenomicRanges</span> package.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'GAlignments'
coverage(x, shift=0L, width=NULL, weight=1L,
         method=c("auto", "sort", "hash", "naive"), drop.D.ranges=FALSE)

## S4 method for signature 'GAlignmentPairs'
coverage(x, shift=0L, width=NULL, weight=1L,
         method=c("auto", "sort", "hash", "naive"), drop.D.ranges=FALSE)

## S4 method for signature 'GAlignmentsList'
coverage(x, shift=0L, width=NULL, weight=1L, ...)

## S4 method for signature 'BamFile'
coverage(x, shift=0L, width=NULL, weight=1L, ...,
         param=ScanBamParam())

## S4 method for signature 'character'
coverage(x, shift=0L, width=NULL, weight=1L, ...,
         yieldSize=2500000L)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="GAlignments-class.html">GAlignments</a>, <a href="GAlignmentPairs-class.html">GAlignmentPairs</a>, <a href="GAlignmentsList-class.html">GAlignmentsList</a>,
or <a href="../../Rsamtools/html/BamFile.html">BamFile</a> object, or the path to a BAM file.
</p>
</td></tr>
<tr valign="top"><td><code>shift, width, weight</code></td>
<td>

<p>See <code>coverage</code> method for <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> objects
in the <span class="pkg">GenomicRanges</span> package.
</p>
</td></tr>
<tr valign="top"><td><code>method</code></td>
<td>

<p>See <code>?<a href="../../IRanges/html/coverage.html">coverage</a></code> in the <span class="pkg">IRanges</span> package
for a description of this argument.
</p>
</td></tr>
<tr valign="top"><td><code>drop.D.ranges</code></td>
<td>

<p>Whether the coverage calculation should ignore ranges corresponding
to D (deletion) in the CIGAR string.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments passed to the <code>coverage</code> method for
<a href="GAlignments-class.html">GAlignments</a> objects.
</p>
</td></tr>
<tr valign="top"><td><code>param</code></td>
<td>

<p>An optional <a href="../../Rsamtools/html/ScanBamParam.html">ScanBamParam</a> object passed to
<code><a href="readGAlignments.html">readGAlignments</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>yieldSize</code></td>
<td>

<p>An optional argument controlling how many records are input
when iterating through a <a href="../../Rsamtools/html/BamFile.html">BamFile</a>.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The methods for <a href="GAlignments-class.html">GAlignments</a> and <a href="GAlignmentPairs-class.html">GAlignmentPairs</a> objects
do:
</p>
<pre>  coverage(grglist(x, drop.D.ranges=drop.D.ranges), ...)
</pre>
<p>The method for <a href="GAlignmentsList-class.html">GAlignmentsList</a> objects does:
</p>
<pre>  coverage(unlist(x), ...)
</pre>
<p>The method for <a href="../../Rsamtools/html/BamFile.html">BamFile</a> objects iterates through a
BAM file, reading <code>yieldSize(x)</code> records (or all records, if
<code>is.na(yieldSize(x))</code>) and calculating:
</p>
<pre>  gal &lt;- readGAlignments(x, param=param)
  coverage(gal, shift=shift, width=width, weight=weight, ...)
</pre>
<p>The method for <code>character</code> vectors of length 1 creates a
<a href="../../Rsamtools/html/BamFile.html">BamFile</a> object from <code>x</code> and performs the
calculation for <code>coverage,BamFile-method</code>.
</p>


<h3>Value</h3>

<p>A named <a href="../../IRanges/html/RleList.html">RleList</a> object with one coverage vector per
seqlevel in <code>x</code>.
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="../../IRanges/html/coverage.html">coverage</a></code> in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><a href="../../GenomicRanges/html/coverage-methods.html">coverage-methods</a> in the
<span class="pkg">GenomicRanges</span> package.
</p>
</li>
<li> <p><a href="../../IRanges/html/RleList.html">RleList</a> objects in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><a href="GAlignments-class.html">GAlignments</a> and <a href="GAlignmentPairs-class.html">GAlignmentPairs</a> objects.
</p>
</li>
<li> <p><code><a href="readGAlignments.html">readGAlignments</a></code>.
</p>
</li>
<li> <p><a href="../../Rsamtools/html/BamFile.html">BamFile</a> objects in the <span class="pkg">Rsamtools</span> package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. EXAMPLE WITH TOY DATA
## ---------------------------------------------------------------------

ex1_file &lt;- system.file("extdata", "ex1.bam", package="Rsamtools")

## Coverage of a GAlignments object:
gal &lt;- readGAlignments(ex1_file)
cvg1 &lt;- coverage(gal)
cvg1

## Coverage of a GAlignmentPairs object:
galp &lt;- readGAlignmentPairs(ex1_file)
cvg2 &lt;- coverage(galp)
cvg2

## Coverage of a GAlignmentsList object:
galist &lt;- readGAlignmentsList(ex1_file)
cvg3 &lt;- coverage(galist)
cvg3

table(mcols(galist)$mate_status)
mated_idx &lt;- which(mcols(galist)$mate_status == "mated")
mated_galist &lt;- galist[mated_idx]
mated_cvg3 &lt;- coverage(mated_galist)
mated_cvg3

## Sanity checks:
stopifnot(identical(cvg1, cvg3))
stopifnot(identical( cvg2, mated_cvg3))

## ---------------------------------------------------------------------
## B. EXAMPLE WITH REAL DATA
## ---------------------------------------------------------------------

library(pasillaBamSubset)
## See '?pasillaBamSubset' for more information about the 2 BAM files
## included in this package.
reads &lt;- readGAlignments(untreated3_chr4())
table(njunc(reads))  # data contains junction reads

## Junctions do NOT contribute to the coverage:
read1 &lt;- reads[which(njunc(reads) != 0L)[1]]  # 1st read with a junction
read1  # cigar shows a "skipped region" of length 15306
grglist(read1)[[1]]  # the junction is between pos 4500 and 19807
coverage(read1)$chr4  # junction is not covered

## Sanity checks:
cvg &lt;- coverage(reads)
read_chunks &lt;- unlist(grglist(reads), use.names=FALSE)
read_chunks_per_chrom &lt;- split(read_chunks, seqnames(read_chunks))
stopifnot(identical(sum(cvg), sum(width(read_chunks_per_chrom))))

galist &lt;- readGAlignmentsList(untreated3_chr4())
stopifnot(identical(cvg, coverage(galist)))
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
