<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Encode the overlaps between RNA-seq reads and the transcripts...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for encodeOverlaps-methods {GenomicAlignments}"><tr><td>encodeOverlaps-methods {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Encode the overlaps between RNA-seq reads and the transcripts of
a gene model</h2>

<h3>Description</h3>

<p>In the context of an RNA-seq experiment, encoding the overlaps between
the aligned reads and the transcripts of a given gene model can be used
for detecting those overlaps that are <em>compatible</em> with the splicing
of the transcript.
</p>
<p>The central tool for this is the <code>encodeOverlaps</code> method for
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> objects, which computes the &quot;overlap
encodings&quot; between a <code>query</code> and a <code>subject</code>, both list-like
objects with list elements containing multiple ranges.
</p>
<p>Other related utilities are also documented in this man page.
</p>


<h3>Usage</h3>

<pre>
encodeOverlaps(query, subject, hits=NULL, ...)

## S4 method for signature 'GRangesList,GRangesList'
encodeOverlaps(query, subject, hits=NULL,
               flip.query.if.wrong.strand=FALSE)

## Related utilities:

flipQuery(x, i)

selectEncodingWithCompatibleStrand(ovencA, ovencB,
                                   query.strand, subject.strand, hits=NULL)

isCompatibleWithSkippedExons(x, max.skipped.exons=NA)

extractSteppedExonRanks(x, for.query.right.end=FALSE)
extractSpannedExonRanks(x, for.query.right.end=FALSE)
extractSkippedExonRanks(x, for.query.right.end=FALSE)

extractQueryStartInTranscript(query, subject, hits=NULL, ovenc=NULL,
                              flip.query.if.wrong.strand=FALSE,
                              for.query.right.end=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>query, subject</code></td>
<td>

<p>Typically <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> objects representing the
the aligned reads and the transcripts of a given gene model, respectively.
If the 2 objects don't have the same length, and if the <code>hits</code>
argument is not supplied, then the shortest is recycled to the length
of the longest (the standard recycling rules apply).
</p>
<p>More generally speaking, <code>query</code> and <code>subject</code> must be
list-like objects with list elements containing multiple ranges e.g.
<a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> or <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a>
objects.
</p>
</td></tr>
<tr valign="top"><td><code>hits</code></td>
<td>

<p>An optional <a href="../../S4Vectors/html/Hits.html">Hits</a> object typically obtained from a
previous call to <code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a>(query, subject)</code>.
</p>
<p>Strictly speaking, <code>hits</code> only needs to be compatible with
<code>query</code> and <code>subject</code>, that is,
<code><a href="../../S4Vectors/html/queryLength.html">queryLength</a>(hits)</code> and
<code><a href="../../S4Vectors/html/subjectLength.html">subjectLength</a>(hits)</code> must be equal to
<code>length(query)</code> and <code>length(subject)</code>, respectively.
</p>
<p>Supplying <code>hits</code> is a convenient way to do
<code>encodeOverlaps(query[queryHits(hits)], subject[subjectHits(hits)])</code>,
that is, calling <code>encodeOverlaps(query, subject, hits)</code> is equivalent
to the above, but is much more efficient, especially when <code>query</code>
and/or <code>subject</code> are big. Of course, when <code>hits</code> is supplied,
<code>query</code> and <code>subject</code> are not expected to have the same length
anymore.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments for methods.
</p>
</td></tr>
<tr valign="top"><td><code>flip.query.if.wrong.strand</code></td>
<td>

<p>See the &quot;OverlapEncodings&quot; vignette located in this package
(<span class="pkg">GenomicAlignments</span>).
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>For <code>flipQuery</code>: a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
</p>
<p>For <code>isCompatibleWithSkippedExons</code>, <code>extractSteppedExonRanks</code>,
<code>extractSpannedExonRanks</code>, and <code>extractSkippedExonRanks</code>:
an <a href="OverlapEncodings-class.html">OverlapEncodings</a> object, a factor, or a character vector.
</p>
</td></tr>
<tr valign="top"><td><code>i</code></td>
<td>

<p>Subscript specifying the elements in <code>x</code> to flip. If missing, all
the elements are flipped.
</p>
</td></tr>
<tr valign="top"><td><code>ovencA, ovencB, ovenc</code></td>
<td>

<p><a href="OverlapEncodings-class.html">OverlapEncodings</a> objects.
</p>
</td></tr>
<tr valign="top"><td><code>query.strand, subject.strand</code></td>
<td>

<p>Vector-like objects containing the strand of the query and subject,
respectively.
</p>
</td></tr>
<tr valign="top"><td><code>max.skipped.exons</code></td>
<td>

<p>Not supported yet. If <code>NA</code> (the default), the number of skipped
exons must be 1 or more (there is no max).
</p>
</td></tr>
<tr valign="top"><td><code>for.query.right.end</code></td>
<td>

<p>If <code>TRUE</code>, then the information reported in the output is for
the right ends of the paired-end reads.
Using <code>for.query.right.end=TRUE</code> with single-end reads is an error.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>See <code>?OverlapEncodings</code> for a short introduction to &quot;overlap encodings&quot;.
</p>
<p>The topic of working with overlap encodings is covered in details
in the &quot;OverlapEncodings&quot; vignette located this package
(<span class="pkg">GenomicAlignments</span>) and accessible with
<code>vignette("OverlapEncodings")</code>.
</p>


<h3>Value</h3>

<p>For <code>encodeOverlaps</code>: An <a href="OverlapEncodings-class.html">OverlapEncodings</a> object.
If <code>hits</code> is not supplied, this object is <em>parallel</em> to the
longest of <code>query</code> and <code>subject</code>, that is, it has the length
of the longest and the i-th encoding in it corresponds to the i-th element
in the longest.
If <code>hits</code> is supplied, then the returned object is <em>parallel</em>
to it, that is, it has one encoding per hit.
</p>
<p>For <code>flipQuery</code>: TODO
</p>
<p>For <code>selectEncodingWithCompatibleStrand</code>: TODO
</p>
<p>For <code>isCompatibleWithSkippedExons</code>: A logical vector <em>parallel</em>
to <code>x</code>.
</p>
<p>For <code>extractSteppedExonRanks</code>, <code>extractSpannedExonRanks</code>, and
<code>extractSkippedExonRanks</code>: TODO
</p>
<p>For <code>extractQueryStartInTranscript</code>: TODO
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li><p> The <a href="OverlapEncodings-class.html">OverlapEncodings</a> class for a brief introduction to
&quot;overlap encodings&quot;.
</p>
</li>
<li><p> The <a href="../../S4Vectors/html/Hits.html">Hits</a> class defined and documented in the
<span class="pkg">S4Vectors</span> package.
</p>
</li>
<li><p> The &quot;OverlapEncodings&quot; vignette in this package.
</p>
</li>
<li> <p><code><a href="findCompatibleOverlaps-methods.html">findCompatibleOverlaps</a></code> for a specialized version
of <code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code> that uses
<code>encodeOverlaps</code> internally to keep only the hits where
the junctions in the aligned read are <em>compatible</em> with
the splicing of the annotated transcript.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> class defined and documented
in the <span class="pkg">GenomicRanges</span> package.
</p>
</li>
<li><p> The <code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code> generic function defined
in the <span class="pkg">IRanges</span> package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. BETWEEN 2 IntegerRangesList OBJECTS
## ---------------------------------------------------------------------
## In the context of an RNA-seq experiment, encoding the overlaps
## between 2 GRangesList objects, one containing the reads (the query),
## and one containing the transcripts (the subject), can be used for
## detecting hits between reads and transcripts that are "compatible"
## with the splicing of the transcript. Here we illustrate this with 2
## IntegerRangesList objects, in order to keep things simple:

## 4 aligned reads in the query:
read1 &lt;- IRanges(c(7, 15, 22), c(9, 19, 23))  # 2 junctions
read2 &lt;- IRanges(c(5, 15), c(9, 17))  # 1 junction
read3 &lt;- IRanges(c(16, 22), c(19, 24))  # 1 junction
read4 &lt;- IRanges(c(16, 23), c(19, 24))  # 1 junction
query &lt;- IRangesList(read1, read2, read3, read4)

## 1 transcript in the subject:
tx &lt;- IRanges(c(1, 4, 15, 22, 38), c(2, 9, 19, 25, 47))  # 5 exons
subject &lt;- IRangesList(tx)

## Encode the overlaps:
ovenc &lt;- encodeOverlaps(query, subject)
ovenc
encoding(ovenc)

## ---------------------------------------------------------------------
## B. BETWEEN 2 GRangesList OBJECTS
## ---------------------------------------------------------------------
## With real RNA-seq data, the reads and transcripts will typically be
## stored in GRangesList objects. Please refer to the "OverlapEncodings"
## vignette in this package for realistic examples.
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
