<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Finding overlapping genomic alignments</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for findOverlaps-methods {GenomicAlignments}"><tr><td>findOverlaps-methods {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Finding overlapping genomic alignments</h2>

<h3>Description</h3>

<p>Finds range overlaps between a <a href="GAlignments-class.html">GAlignments</a>, <a href="GAlignmentPairs-class.html">GAlignmentPairs</a>, 
or <a href="GAlignmentsList-class.html">GAlignmentsList</a> object, and another range-based object.
</p>
<p>NOTE: The <code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code> generic function and methods
for <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> and <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a>
objects are defined and documented in the <span class="pkg">IRanges</span> package.
The methods for <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> and
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> objects are defined and
documented in the <span class="pkg">GenomicRanges</span> package.
</p>
<p><a href="GAlignments-class.html">GAlignments</a>, <a href="GAlignmentPairs-class.html">GAlignmentPairs</a>, and <a href="GAlignmentsList-class.html">GAlignmentsList</a>
objects also support <code>countOverlaps</code>, <code>overlapsAny</code>, and
<code>subsetByOverlaps</code> thanks to the default methods defined in the
<span class="pkg">IRanges</span> package and to the <code>findOverlaps</code> method defined in
this package and documented below.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'GAlignments,GAlignments'
findOverlaps(query, subject,
    maxgap=-1L, minoverlap=0L,
    type=c("any", "start", "end", "within"),
    select=c("all", "first", "last", "arbitrary"),
    ignore.strand=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>query, subject</code></td>
<td>

<p>A <a href="GAlignments-class.html">GAlignments</a>, <a href="GAlignmentPairs-class.html">GAlignmentPairs</a>, or <a href="GAlignmentsList-class.html">GAlignmentsList</a>
object for either <code>query</code> or <code>subject</code>.
A vector-like object containing ranges for the other one.
</p>
</td></tr>
<tr valign="top"><td><code>maxgap, minoverlap, type, select</code></td>
<td>

<p>See <code>?<a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code> in the <span class="pkg">IRanges</span> package
for a description of these arguments.
</p>
</td></tr>
<tr valign="top"><td><code>ignore.strand</code></td>
<td>

<p>When set to <code>TRUE</code>, the strand information is ignored in the
overlap calculations.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>When the query or the subject (or both) is a <a href="GAlignments-class.html">GAlignments</a>
object, it is first turned into a <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a> object (with
<code>as( , "GRangesList")</code>) and then the rules described previously
apply. <a href="GAlignmentsList-class.html">GAlignmentsList</a> objects are coerced to <a href="GAlignments-class.html">GAlignments</a>
then to a <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a>. Feature indices are mapped back to the 
original <a href="GAlignmentsList-class.html">GAlignmentsList</a> list elements.
</p>
<p>When the query is a <a href="GAlignmentPairs-class.html">GAlignmentPairs</a> object, it is first
turned into a <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a> object (with <code>as( , "GRangesList")</code>)
and then the rules described previously apply.
</p>


<h3>Value</h3>

<p>A <a href="../../S4Vectors/html/Hits.html">Hits</a> object when <code>select="all"</code> or an integer
vector otherwise. 
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code>.
</p>
</li>
<li> <p><a href="../../S4Vectors/html/Hits-class.html">Hits-class</a>.
</p>
</li>
<li> <p><a href="../../GenomicRanges/html/GRanges-class.html">GRanges-class</a>.
</p>
</li>
<li> <p><a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList-class</a>.
</p>
</li>
<li> <p><a href="GAlignments-class.html">GAlignments-class</a>.
</p>
</li>
<li> <p><a href="GAlignmentPairs-class.html">GAlignmentPairs-class</a>.
</p>
</li>
<li> <p><a href="GAlignmentsList-class.html">GAlignmentsList-class</a>.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
ex1_file &lt;- system.file("extdata", "ex1.bam", package="Rsamtools")
galn &lt;- readGAlignments(ex1_file)

subject &lt;- granges(galn)[1]

## Note the absence of query no. 9 (i.e. 'galn[9]') in this result:
as.matrix(findOverlaps(galn, subject))

## This is because, by default, findOverlaps()/countOverlaps() are
## strand specific:
galn[8:10]
countOverlaps(galn[8:10], subject)
countOverlaps(galn[8:10], subject, ignore.strand=TRUE)

## Count alignments in 'galn' that DO overlap with 'subject' vs those
## that do NOT:
table(overlapsAny(galn, subject))
## Extract those that DO:
subsetByOverlaps(galn, subject)

## GAlignmentsList
galist &lt;- GAlignmentsList(galn[8:10], galn[3000:3002])
gr &lt;- GRanges(c("seq1", "seq1", "seq2"), 
              IRanges(c(15, 18, 1233), width=1),
              strand=c("-", "+", "+"))

countOverlaps(galist, gr)
countOverlaps(galist, gr, ignore.strand=TRUE)
findOverlaps(galist, gr)
findOverlaps(galist, gr, ignore.strand=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
