<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: GAlignmentsList objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for GAlignmentsList-class {GenomicAlignments}"><tr><td>GAlignmentsList-class {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>GAlignmentsList objects</h2>

<h3>Description</h3>

<p>The GAlignmentsList class is a container for storing a collection of
<a href="GAlignments-class.html">GAlignments</a> objects.
</p>


<h3>Details</h3>

<p>A GAlignmentsList object contains a list of <a href="GAlignments-class.html">GAlignments</a> objects.
The majority of operations on this page are described in more detail
on the GAlignments man page, see ?<code>GAlignments</code>.
</p>


<h3>Constructor</h3>


<dl>
<dt></dt><dd>
<p><code>GAlignmentsList(...)</code>:
Creates a GAlignmentsList from a list of <a href="GAlignments-class.html">GAlignments</a> objects.
</p>
</dd>
</dl>



<h3>Accessors</h3>

<p>In the code snippets below, <code>x</code> is a GAlignmentsList object.
</p>

<dl>
<dt></dt><dd>
<p><code>length(x)</code>:
Return the number of elements in <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>names(x)</code>, <code>names(x) &lt;- value</code>:
Get or set the names of the elements of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>seqnames(x)</code>, <code>seqnames(x) &lt;- value</code>:
Get or set the name of the reference sequences of the
alignments in each element of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>rname(x)</code>, <code>rname(x) &lt;- value</code>:
Same as <code>seqnames(x)</code> and <code>seqnames(x) &lt;- value</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>strand(x)</code>, <code>strand(x) &lt;- value</code>:
Get or set the strand of the alignments in each element 
of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>cigar(x)</code>:
Returns a character list of length <code>length(x)</code>
containing the CIGAR string for the alignments in
each element of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>qwidth(x)</code>:
Returns an integer list of length <code>length(x)</code>
containing the length of the alignments in each element of
<code>x</code> *after* hard clipping (i.e. the length of the 
query sequence that is stored in the corresponding SAM/BAM record).
</p>
</dd>
<dt></dt><dd>
<p><code>start(x)</code>, <code>end(x)</code>:
Returns an integer list of length <code>length(x)</code>
containing the &quot;start&quot; and &quot;end&quot; (respectively) of the 
alignments in each element of <code>x</code>. 
</p>
</dd>
<dt></dt><dd>
<p><code>width(x)</code>:
Returns an integer list of length <code>length(x)</code> containing
the &quot;width&quot; of the alignments in each element of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>njunc(x)</code>:
Returns an integer list of length <code>x</code> containing the number
of junctions (i.e. N operations in the CIGAR) for the alignments
in each element of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>seqinfo(x)</code>, <code>seqinfo(x) &lt;- value</code>:
Get or set the information about the underlying sequences.
<code>value</code> must be a <a href="../../GenomeInfoDb/html/Seqinfo.html">Seqinfo</a> object.
</p>
</dd>
<dt></dt><dd>
<p><code>seqlevels(x)</code>, <code>seqlevels(x) &lt;- value</code>:
Get or set the sequence levels of the alignments in each element
of <code>x</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>seqlengths(x)</code>, <code>seqlengths(x) &lt;- value</code>:
Get or set the sequence lengths for each element of <code>x</code>.
<code>seqlengths(x)</code> is equivalent to <code>seqlengths(seqinfo(x))</code>.
<code>value</code> can be a named non-negative integer or numeric vector
eventually with NAs.
</p>
</dd>
<dt></dt><dd>
<p><code>isCircular(x)</code>, <code>isCircular(x) &lt;- value</code>:
Get or set the circularity flags for the alignments in each
element in <code>x</code>. <code>value</code> must be a named logical list 
eventually with NAs.
</p>
</dd>
<dt></dt><dd>
<p><code>genome(x)</code>, <code>genome(x) &lt;- value</code>:
Get or set the genome identifier or assembly name for the alignments 
in each element of <code>x</code>. <code>value</code> must be a named character 
list eventually with NAs.
</p>
</dd>
<dt></dt><dd>
<p><code>seqnameStyle(x)</code>:
Get or set the seqname style for alignments in each element of <code>x</code>.
</p>
</dd>
</dl>



<h3>Coercion</h3>

<p>In the code snippets below, <code>x</code> is a GAlignmentsList object.
</p>

<dl>
<dt></dt><dd>
<p><code>granges(x, use.names=TRUE, use.mcols=FALSE, ignore.strand=FALSE)</code>, 
<code>ranges(x, use.names=TRUE, use.mcols=FALSE)</code>:
</p>
<p>Return either a <a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a> or a <a href="../../IRanges/html/IRanges.html">IRanges</a>
object of length <code>length(x)</code>. Note this coercion IGNORES 
the cigar information. The resulting ranges span the entire
range, including any junctions or spaces between paired-end reads.
</p>
<p>If <code>use.names</code> is TRUE, then the names on <code>x</code>
(if any) are propagated to the returned object.
If <code>use.mcols</code> is TRUE, then the metadata columns on <code>x</code>
(if any) are propagated to the returned object.
</p>
<p><code>granges</code> coercion supports <code>ignore.strand</code> to allow 
ranges of opposite strand to be combined (see examples). All
ranges in the resulting GRanges will have strand &lsquo;*&rsquo;.
</p>
</dd>
<dt></dt><dd>
<p><code>grglist(x, use.names=TRUE, use.mcols=FALSE, ignore.strand=FALSE)</code>, 
<code>rglist(x, use.names=TRUE, use.mcols=FALSE)</code>:
</p>
<p>Return either a <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a> or an <a href="../../IRanges/html/IRangesList.html">IRangesList</a>
object of length <code>length(x)</code>. This coercion RESPECTS the cigar 
information. The resulting ranges are fragments of the original ranges 
that do not include junctions or spaces between paired-end reads.
</p>
<p>If <code>use.names</code> is TRUE, then the names on <code>x</code>
(if any) are propagated to the returned object.
If <code>use.mcols</code> is TRUE, then the metadata columns on <code>x</code>
(if any) are propagated to the returned object.
</p>
<p><code>grglist</code> coercion supports <code>ignore.strand</code> to allow 
ranges of opposite strand to be combined (see examples). When
<code>ignore.strand</code> is TRUE all ranges in the resulting 
GRangesList have strand &lsquo;*&rsquo;.
</p>
</dd>
<dt></dt><dd>
<p><code>as(x, "GRanges")</code>, <code>as(x, "IntegerRanges")</code>,
<code>as(x, "GRangesList")</code>, <code>as(x, "IntegerRangesList")</code>:
Alternate ways of doing
<code>granges(x, use.names=TRUE, use.mcols=TRUE)</code>,
<code>ranges(x, use.names=TRUE, use.mcols=TRUE)</code>,
<code>grglist(x, use.names=TRUE, use.mcols=TRUE)</code>, and
<code>rglist(x, use.names=TRUE, use.mcols=TRUE)</code>, respectively.
</p>
</dd>
<dt></dt><dd>
<p><code>as.data.frame(x, row.names = NULL, optional = FALSE,
         ..., value.name = "value", use.outer.mcols = FALSE, 
         group_name.as.factor = FALSE)</code>: 
Coerces <code>x</code> to a <code>data.frame</code>. See as.data.frame on the 
<code>List</code> man page for details (?<code>List</code>). 
</p>
</dd>
<dt></dt><dd>
<p><code>as(x, "GALignmentsList")</code>: Here <code>x</code> is a
<a href="GAlignmentPairs-class.html">GAlignmentPairs</a> object. Return a GAlignmentsList object of length
<code>length(x)</code> where the i-th list element represents the ranges of
the i-th alignment pair in <code>x</code>.
</p>
</dd>
</dl>



<h3>Subsetting and related operations</h3>

<p>In the code snippets below, <code>x</code> is a GAlignmentsList object.
</p>

<dl>
<dt></dt><dd>
<p><code>x[i]</code>, <code>x[i] &lt;- value</code>:
Get or set list elements <code>i</code>. <code>i</code> can be a numeric 
or logical vector. <code>value</code> must be a GAlignments.
</p>
</dd>
<dt></dt><dd>
<p><code>x[[i]]</code>, <code>x[[i]] &lt;- value</code>:
Same as <code>x[i]</code>, <code>x[i] &lt;- value</code>. 
</p>
</dd>
<dt></dt><dd>
<p><code>x[i, j]</code>, <code>x[i, j] &lt;- value</code>:
Get or set list elements <code>i</code> with optional metadata columns
<code>j</code>. <code>i</code> can be a numeric, logical or missing. 
<code>value</code> must be a GAlignments.
</p>
</dd>
</dl>



<h3>Concatenation</h3>


<dl>
<dt></dt><dd>
<p><code>c(x, ..., ignore.mcols=FALSE)</code>:
Concatenate GAlignmentsList object <code>x</code> and the GAlignmentsList
objects in <code>...</code> together.
See <code>?<a href="../../S4Vectors/html/c.html">c</a></code> in the <span class="pkg">S4Vectors</span> package for
more information about concatenating Vector derivatives.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Valerie Obenchain</p>


<h3>References</h3>

<p><a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a>
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="readGAlignments.html">readGAlignmentsList</a></code> for reading genomic alignments
from a file (typically a BAM file) into a GAlignmentsList object.
</p>
</li>
<li> <p><a href="GAlignments-class.html">GAlignments</a> and <a href="GAlignmentPairs-class.html">GAlignmentPairs</a> objects for handling
aligned single- and paired-end reads, respectively.
</p>
</li>
<li> <p><a href="junctions-methods.html">junctions-methods</a> for extracting and summarizing junctions
from a GAlignmentsList object.
</p>
</li>
<li> <p><a href="findOverlaps-methods.html">findOverlaps-methods</a> for finding range
overlaps between a GAlignmentsList object and another range-based
object.
</p>
</li>
<li> <p><code><a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a></code> in the <span class="pkg">GenomeInfoDb</span>
package for getting/setting/modifying the sequence information
stored in an object.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> and
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> classes defined and documented
in the <span class="pkg">GenomicRanges</span> package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
gal1 &lt;- GAlignments(
    seqnames=Rle(factor(c("chr1", "chr2", "chr1", "chr3")),
        c(1, 3, 2, 4)),
    pos=1:10, cigar=paste0(10:1, "M"),
    strand=Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    names=head(letters, 10), score=1:10)

gal2 &lt;- GAlignments(
    seqnames=Rle(factor(c("chr2", "chr4")), c(3, 4)), pos=1:7,
    cigar=c("5M", "3M2N3M2N3M", "5M", "10M", "5M1N4M", "8M2N1M", "5M"),
    strand=Rle(strand(c("-", "+")), c(4, 3)),
    names=tail(letters, 7), score=1:7)

galist &lt;- GAlignmentsList(noGaps=gal1, Gaps=gal2)


## ---------------------------------------------------------------------
## A. BASIC MANIPULATION
## ---------------------------------------------------------------------

length(galist)
names(galist)
seqnames(galist)
strand(galist)
head(cigar(galist))
head(qwidth(galist))
head(start(galist))
head(end(galist))
head(width(galist))
head(njunc(galist))
seqlevels(galist)

## Rename the reference sequences:
seqlevels(galist) &lt;- sub("chr", "seq", seqlevels(galist))
seqlevels(galist)

grglist(galist)  # a GRangesList object
rglist(galist)   # an IRangesList object

## ---------------------------------------------------------------------
## B. SUBSETTING
## ---------------------------------------------------------------------

galist[strand(galist) == "-"]
has_junctions &lt;- sapply(galist,
                        function(x) any(grepl("N", cigar(x), fixed=TRUE)))
galist[has_junctions]

## Different ways to subset:
galist[2]             # a GAlignments object of length 1
galist[[2]]           # a GAlignments object of length 1
grglist(galist[2])  # a GRangesList object of length 1
rglist(galist[2])   # a NormalIRangesList object of length 1

## ---------------------------------------------------------------------
## C. mcols()/elementMetadata()
## ---------------------------------------------------------------------

## Metadata can be defined on the individual GAlignment elements
## and the overall GAlignmentsList object. By default, 'level=between' 
## extracts the GALignmentsList metadata. Using 'level=within' 
## will extract the metadata on the individual GAlignments objects.

mcols(galist) ## no metadata on the GAlignmentsList object
mcols(galist, level="within")


## ---------------------------------------------------------------------
## D. readGAlignmentsList()
## ---------------------------------------------------------------------

library(pasillaBamSubset)

## 'file' as character.
fl &lt;- untreated3_chr4()
galist1 &lt;- readGAlignmentsList(fl)

galist1[1:3]
length(galist1)
table(elementNROWS(galist1))

## When 'file' is a BamFile, 'asMates' must be TRUE. If FALSE,
## the data are treated as single-end and each list element of the
## GAlignmentsList will be of length 1. For single-end data 
## use readGAlignments() instead of readGAlignmentsList().
bf &lt;- BamFile(fl, yieldSize=3, asMates=TRUE)
readGAlignmentsList(bf)

## Use a 'param' to fine tune the results.
param &lt;- ScanBamParam(flag=scanBamFlag(isProperPair=TRUE))
galist2 &lt;- readGAlignmentsList(fl, param=param)
length(galist2)


## ---------------------------------------------------------------------
## E. COERCION 
## ---------------------------------------------------------------------

## The granges() and grlist() coercions support 'ignore.strand' to 
## allow ranges from different strands to be combined. In this example 
## paired-end reads aligned to opposite strands were read into a 
## GAlignmentsList. If the desired operation is to combine these ranges, 
## regardless of junctions or the space between pairs, 'ignore.strand'
## must be TRUE.
granges(galist[1])
granges(galist[1], ignore.strand=TRUE)

## grglist()
galist &lt;- GAlignmentsList(noGaps=gal1, Gaps=gal2)
grglist(galist)
grglist(galist, ignore.strand=TRUE)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
