<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Intra range transformations of a GAlignments or...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for intra-range-methods {GenomicAlignments}"><tr><td>intra-range-methods {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Intra range transformations of a GAlignments or GAlignmentsList object</h2>

<h3>Description</h3>

<p>This man page documents intra range transformations of a
<a href="GAlignments-class.html">GAlignments</a> or <a href="GAlignmentsList-class.html">GAlignmentsList</a> object.
</p>
<p>See <code>?`<a href="../../IRanges/html/intra-range-methods.html">intra-range-methods</a>`</code> and
<code>?`<a href="../../IRanges/html/inter-range-methods.html">inter-range-methods</a>`</code> in the <span class="pkg">IRanges</span>
package for a quick introduction to intra range and inter range
transformations.
</p>
<p>Intra range methods for <a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a> and <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a>
objects are defined and documented in the <span class="pkg">GenomicRanges</span> package.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'GAlignments'
qnarrow(x, start=NA, end=NA, width=NA)
## S4 method for signature 'GAlignmentsList'
qnarrow(x, start=NA, end=NA, width=NA)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="GAlignments-class.html">GAlignments</a> or <a href="GAlignmentsList-class.html">GAlignmentsList</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>start, end, width</code></td>
<td>

<p>Vectors of integers.
NAs and negative values are accepted and &quot;solved&quot; according to the
rules of the SEW (Start/End/Width) interface (see
<code>?<a href="../../IRanges/html/solveUserSEW.html">solveUserSEW</a></code> for more information about the
SEW interface).
</p>
<p>See <code>?`<a href="../../IRanges/html/intra-range-methods.html">intra-range-methods</a>`</code> for more information
about the <code>start</code>, <code>end</code>, and <code>width</code> arguments.
</p>
</td></tr>
</table>


<h3>Details</h3>


<ul>
<li><p>()
<code>qnarrow</code> on a <a href="GAlignments-class.html">GAlignments</a> object behaves like <code>narrow</code>
except that the <code>start</code>/<code>end</code>/<code>width</code> arguments here
specify the narrowing with respect to the query sequences instead of
the reference sequences.
</p>
<p><code>qnarrow</code> on a <a href="GAlignmentsList-class.html">GAlignmentsList</a> object
returns a <a href="GAlignmentsList-class.html">GAlignmentsList</a> object.

</p>
</li></ul>



<h3>Value</h3>

<p>An object of the same class as &ndash; and <em>parallel</em> to (i.e. same length
and names as) &ndash; the original object <code>x</code>.
</p>


<h3>Note</h3>

<p>There is no difference between <code>narrow</code> and <code>qnarrow</code> when
all the alignments have a simple CIGAR (i.e. no indels or junctions).
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès and Valerie Obenchain</p>


<h3>See Also</h3>


<ul>
<li> <p><a href="GAlignments-class.html">GAlignments</a> and <a href="GAlignmentsList-class.html">GAlignmentsList</a> objects.
</p>
</li>
<li><p> The <a href="../../IRanges/html/intra-range-methods.html">intra-range-methods</a> man page in the
<span class="pkg">IRanges</span> package.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/intra-range-methods.html">intra-range-methods</a> man page in the
<span class="pkg">GenomicRanges</span> package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. ON A GAlignments OBJECT
## ---------------------------------------------------------------------
ex1_file &lt;- system.file("extdata", "ex1.bam", package="Rsamtools")
param &lt;- ScanBamParam(what=c("seq", "qual"))
gal &lt;- readGAlignments(ex1_file, param=param)
gal

## This trims 3 nucleotides on the left and 5 nucleotides on the right
## of each alignment:
gal2 &lt;- qnarrow(gal, start=4, end=-6)
gal2

## Note that the 'start' and 'end' values are relative to the query
## sequences and specify the query substring that must be kept for each
## alignment. Negative values are relative to the right end of the query
## sequence.

## Also note that the metadata columns on 'gal' are propagated as-is so
## the "seq" and "qual" matadata columns must be adjusted "by hand" with
## narrow();
mcols(gal2)$seq &lt;- narrow(mcols(gal)$seq, start=4, end=-6)
mcols(gal2)$qual &lt;- narrow(mcols(gal)$qual, start=4, end=-6)
gal2

## Sanity checks:
stopifnot(identical(qwidth(gal2), width(mcols(gal2)$seq)))
stopifnot(identical(qwidth(gal2), width(mcols(gal2)$qual)))

## ---------------------------------------------------------------------
## B. ON A GAlignmentsList OBJECT
## ---------------------------------------------------------------------
gal1 &lt;- GAlignments(
    seqnames=Rle(factor(c("chr1", "chr2", "chr1", "chr3")),
        c(1, 3, 2, 4)),
    pos=1:10, cigar=paste0(10:1, "M"),
    strand=Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    names=head(letters, 10), score=1:10)

gal2 &lt;- GAlignments(
    seqnames=Rle(factor(c("chr2", "chr4")), c(3, 4)), pos=1:7,
    cigar=c("5M", "3M2N3M2N3M", "5M", "10M", "5M1N4M", "8M2N1M", "5M"),
    strand=Rle(strand(c("-", "+")), c(4, 3)),
    names=tail(letters, 7), score=1:7)

galist &lt;- GAlignmentsList(noGaps=gal1, Gaps=gal2)
galist

qnarrow(galist)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
