<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: OverlapEncodings objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for OverlapEncodings-class {GenomicAlignments}"><tr><td>OverlapEncodings-class {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>OverlapEncodings objects</h2>

<h3>Description</h3>

<p>The OverlapEncodings class is a container for storing the
&quot;overlap encodings&quot; returned by the <code><a href="encodeOverlaps-methods.html">encodeOverlaps</a></code>
function.
</p>


<h3>Usage</h3>

<pre>
## -=-= OverlapEncodings getters =-=-

## S4 method for signature 'OverlapEncodings'
Loffset(x)
## S4 method for signature 'OverlapEncodings'
Roffset(x)
## S4 method for signature 'OverlapEncodings'
encoding(x)
## S4 method for signature 'OverlapEncodings'
levels(x)
## S4 method for signature 'OverlapEncodings'
flippedQuery(x)

## -=-= Coercing an OverlapEncodings object =-=-

## S4 method for signature 'OverlapEncodings'
as.data.frame(x, row.names=NULL, optional=FALSE, ...)

## -=-= Low-level encoding utilities =-=-

encodingHalves(x, single.end.on.left=FALSE, single.end.on.right=FALSE,
                  as.factors=FALSE)
Lencoding(x, ...)
Rencoding(x, ...)

## S4 method for signature 'ANY'
njunc(x)

Lnjunc(x, single.end.on.left=FALSE)
Rnjunc(x, single.end.on.right=FALSE)

isCompatibleWithSplicing(x)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>An OverlapEncodings object. For the low-level encoding utilities, <code>x</code>
can also be a character vector or factor containing encodings.
</p>
</td></tr>
<tr valign="top"><td><code>row.names</code></td>
<td>

<p><code>NULL</code> or a character vector.
</p>
</td></tr>
<tr valign="top"><td><code>optional</code></td>
<td>

<p>Ignored.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Extra arguments passed to the <code>as.data.frame</code> method for
OverlapEncodings objects are ignored.
</p>
<p>Extra arguments passed to <code>Lencoding</code> or <code>Rencoding</code> are
passed down to <code>encodingHalves</code>.
</p>
</td></tr>
<tr valign="top"><td><code>single.end.on.left, single.end.on.right</code></td>
<td>

<p>By default the 2 halves of a single-end encoding are considered to be NAs.
If <code>single.end.on.left</code> (resp. <code>single.end.on.right</code>) is
<code>TRUE</code>, then the left (resp. right) half of a single-end encoding
is considered to be the unmodified encoding.
</p>
</td></tr>
<tr valign="top"><td><code>as.factors</code></td>
<td>

<p>By default <code>encodingHalves</code> returns the 2 encoding halves
as a list of 2 character vectors parallel to the input.
If <code>as.factors</code> is <code>TRUE</code>, then it returns them as a
list of 2 factors parallel to the input.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Given a <code>query</code> and a <code>subject</code> of the same length, both
list-like objects with top-level elements typically containing multiple
ranges (e.g. <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> objects), the &quot;overlap
encoding&quot; of the i-th element in <code>query</code> and i-th element in
<code>subject</code> is a character string describing how the ranges in
<code>query[[i]]</code> are <em>qualitatively</em> positioned relatively to
the ranges in <code>subject[[i]]</code>.
</p>
<p>The <code><a href="encodeOverlaps-methods.html">encodeOverlaps</a></code> function computes those overlap
encodings and returns them in an OverlapEncodings object of the same
length as <code>query</code> and <code>subject</code>.
</p>
<p>The topic of working with overlap encodings is covered in details
in the &quot;OverlapEncodings&quot; vignette located this package
(<span class="pkg">GenomicAlignments</span>) and accessible with
<code>vignette("OverlapEncodings")</code>.
</p>


<h3>OverlapEncodings getters</h3>

<p>In the following code snippets, <code>x</code> is an OverlapEncodings object
typically obtained by a call to <code><a href="encodeOverlaps-methods.html">encodeOverlaps</a>(query, subject)</code>.
</p>

<dl>
<dt></dt><dd>
<p><code>length(x)</code>:
Get the number of elements (i.e. encodings) in <code>x</code>.
This is equal to <code>length(query)</code> and <code>length(subject)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>Loffset(x)</code>, <code>Roffset(x)</code>:
Get the &quot;left offsets&quot; and &quot;right offsets&quot; of the encodings,
respectively. Both are integer vectors of the same length as <code>x</code>.
</p>
<p>Let's denote <code>Qi = query[[i]]</code>, <code>Si = subject[[i]]</code>,
and [q1,q2] the range covered by <code>Qi</code> i.e.
<code>q1 = min(start(Qi))</code> and <code>q2 = max(end(Qi))</code>,
then <code>Loffset(x)[i]</code> is the number <code>L</code> of ranges at the
<em>head</em> of <code>Si</code> that are strictly to the left of all
the ranges in <code>Qi</code> i.e. <code>L</code> is the greatest value such that
<code>end(Si)[k] &lt; q1 - 1</code> for all <code>k</code> in <code>seq_len(L)</code>.
Similarly, <code>Roffset(x)[i]</code> is the number <code>R</code> of ranges at the
<em>tail</em> of <code>Si</code> that are strictly to the right of all
the ranges in <code>Qi</code> i.e. <code>R</code> is the greatest value such that
<code>start(Si)[length(Si) + 1 - k] &gt; q2 + 1</code> for all <code>k</code>
in <code>seq_len(L)</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>encoding(x)</code>:
Factor of the same length as <code>x</code> where the i-th element is
the encoding obtained by comparing each range in <code>Qi</code> with
all the ranges in <code>tSi = Si[(1+L):(length(Si)-R)]</code> (<code>tSi</code>
stands for &quot;trimmed Si&quot;).
More precisely, here is how this encoding is obtained:
</p>

<ol>
<li><p> All the ranges in <code>Qi</code> are compared with <code>tSi[1]</code>,
then with <code>tSi[2]</code>, etc...
At each step (one step per range in <code>tSi</code>), comparing
all the ranges in <code>Qi</code> with <code>tSi[k]</code> is done with
<code>rangeComparisonCodeToLetter(compare(Qi, tSi[k]))</code>.
So at each step, we end up with a vector of <code>M</code>
single letters (where <code>M</code> is <code>length(Qi)</code>).
</p>
</li>
<li><p> Each vector obtained previously (1 vector per range in
<code>tSi</code>, all of them of length <code>M</code>) is turned
into a single string (called &quot;encoding block&quot;) by pasting
its individual letters together.
</p>
</li>
<li><p> All the encoding blocks (1 per range in <code>tSi</code>) are pasted
together into a single long string and separated by colons
(<code>":"</code>). An additional colon is prepended to the long
string and another one appended to it.
</p>
</li>
<li><p> Finally, a special block containing the value of <code>M</code> is
prepended to the long string. The final string is the encoding.
</p>
</li></ol>

</dd>
<dt></dt><dd>
<p><code>levels(x)</code>: Equivalent to <code>levels(encoding(x))</code>.
</p>
</dd>
<dt></dt><dd>
<p><code>flippedQuery(x)</code>:
Whether or not the top-level element in query used for computing the
encoding was &quot;flipped&quot; before the encoding was computed.
Note that this flipping generally affects the &quot;left offset&quot;,
&quot;right offset&quot;, in addition to the encoding itself.
</p>
</dd>
</dl>
   


<h3>Coercing an OverlapEncodings object</h3>

<p>In the following code snippets, <code>x</code> is an OverlapEncodings object.
</p>

<dl>
<dt></dt><dd>
<p><code>as.data.frame(x)</code>:
Return <code>x</code> as a data frame with columns <code>"Loffset"</code>,
<code>"Roffset"</code> and <code>"encoding"</code>.
</p>
</dd>
</dl>



<h3>Low-level encoding utilities</h3>

<p>In the following code snippets, <code>x</code> can be an OverlapEncodings object,
or a character vector or factor containing encodings.
</p>

<dl>
<dt></dt><dd>
<p><code>encodingHalves(x, single.end.on.left=FALSE, single.end.on.right=FALSE, as.factors=FALSE)</code>:
Extract the 2 halves of paired-end encodings and return them as a list
of 2 character vectors (or 2 factors) parallel to the input.
</p>
<p>Paired-end encodings are obtained by encoding paired-end overlaps
i.e. overlaps between paired-end reads and transcripts (typically).
The difference between a single-end encoding and a paired-end encoding
is that all the blocks in the latter contain a <code>"--"</code> separator
to mark the separation between the &quot;left encoding&quot; and the &quot;right
encoding&quot;.
</p>
<p>See examples below and the &quot;Overlap encodings&quot; vignette located in this
package for examples of paired-end encodings.
</p>
</dd>
<dt></dt><dd>
<p><code>Lencoding(x, ...)</code>, <code>Rencoding(x, ...)</code>:
Extract the &quot;left encodings&quot; and &quot;right encodings&quot; of paired-end
encodings.
</p>
<p>Equivalent to <code>encodingHalves(x, ...)[[1]]</code> and
<code>encodingHalves(x, ...)[[2]]</code>, respectively.
</p>
</dd>
<dt></dt><dd>
<p><code>njunc(x)</code>, <code>Lnjunc(x, single.end.on.left=FALSE)</code>,
<code>Rnjunc(x, single.end.on.right=FALSE)</code>:
Extract the number of junctions in each encoding by looking at their
first block (aka special block).
If an element <code>xi</code> in <code>x</code> is a paired-end encoding,
then <code>Lnjunc(xi)</code>, <code>Rnjunc(xi)</code>, and <code>njunc(xi)</code>,
return <code>njunc(Lencoding(xi))</code>, <code>njunc(Rencoding(xi))</code>,
and <code>Lnjunc(xi) + Rnjunc(xi)</code>, respectively.
</p>
</dd>
<dt></dt><dd>
<p><code>isCompatibleWithSplicing(x)</code>:
Returns a logical vector <em>parallel</em> to <code>x</code> indicating whether
the corresponding encoding describes a <em>splice compatible</em> overlap
i.e. an overlap that is compatible with the splicing of the transcript.
</p>
<p>WARNING: For paired-end encodings, <code>isCompatibleWithSplicing</code>
considers that the encoding is <em>splice compatible</em> if its
2 halves are <em>splice compatible</em>. This can produce false positives
if for example the right end of the alignment is located upstream of the
left end in transcript space. The paired-end read could not come from
this transcript. To eliminate these false positives, one would need to
have access and look at the position of the left and right ends in
transcript space. This can be done with
<code><a href="encodeOverlaps-methods.html">extractQueryStartInTranscript</a></code>. 
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li><p> The &quot;OverlapEncodings&quot; vignette in this package.
</p>
</li>
<li><p> The <code><a href="encodeOverlaps-methods.html">encodeOverlaps</a></code> function for computing &quot;overlap
encodings&quot;.
</p>
</li>
<li><p> The <code><a href="../../IRanges/html/pcompare.html">pcompare</a></code> function in the <span class="pkg">IRanges</span>
package for the interpretation of the strings returned by
<code>encoding</code>.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> class defined and documented
in the <span class="pkg">GenomicRanges</span> package.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. BASIC MANIPULATION OF AN OverlapEncodings OBJECT
## ---------------------------------------------------------------------

example(encodeOverlaps)  # to generate the 'ovenc' object

length(ovenc)
Loffset(ovenc)
Roffset(ovenc)
encoding(ovenc)
levels(ovenc)
nlevels(ovenc)
flippedQuery(ovenc)
njunc(ovenc)

as.data.frame(ovenc)
njunc(levels(ovenc))

## ---------------------------------------------------------------------
## B. WORKING WITH PAIRED-END ENCODINGS (POSSIBLY MIXED WITH SINGLE-END
##    ENCODINGS)
## ---------------------------------------------------------------------

encodings &lt;- c("4:jmmm:agmm:aagm:aaaf:", "3--1:jmm--b:agm--i:")

encodingHalves(encodings)
encodingHalves(encodings, single.end.on.left=TRUE)
encodingHalves(encodings, single.end.on.right=TRUE)
encodingHalves(encodings, single.end.on.left=TRUE,
                          single.end.on.right=TRUE)

Lencoding(encodings)
Lencoding(encodings, single.end.on.left=TRUE)
Rencoding(encodings)
Rencoding(encodings, single.end.on.right=TRUE)

njunc(encodings)
Lnjunc(encodings)
Lnjunc(encodings, single.end.on.left=TRUE)
Rnjunc(encodings)
Rnjunc(encodings, single.end.on.right=TRUE)

## ---------------------------------------------------------------------
## C. DETECTION OF "SPLICE COMPATIBLE" OVERLAPS
## ---------------------------------------------------------------------

## Reads that are compatible with the splicing of the transcript can
## be detected with a regular expression (the regular expression below
## assumes that reads have at most 2 junctions):
regex0 &lt;- "(:[fgij]:|:[jg].:.[gf]:|:[jg]..:.g.:..[gf]:)"
grepl(regex0, encoding(ovenc))  # read4 is NOT "compatible"

## This was for illustration purpose only. In practise you don't need
## (and should not) use this regular expression, but use instead the
## isCompatibleWithSplicing() utility function:
isCompatibleWithSplicing(ovenc)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
