<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Pile the letters of a set of aligned reads on top of a set of...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for pileLettersAt {GenomicAlignments}"><tr><td>pileLettersAt {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Pile the letters of a set of aligned reads on top of a set
of genomic positions</h2>

<h3>Description</h3>

<p><code>pileLettersAt</code> extracts the letters/nucleotides of a set of
reads that align to a set of genomic positions of interest.
The extracted letters are returned as &quot;piles of letters&quot; (one per
genomic position of interest) stored in an <a href="../../Biostrings/html/XStringSet.html">XStringSet</a>
(typically <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a>) object.
</p>


<h3>Usage</h3>

<pre>
pileLettersAt(x, seqnames, pos, cigar, at)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>An <a href="../../Biostrings/html/XStringSet.html">XStringSet</a> (typically
<a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a>) object containing N <em>unaligned</em>
read sequences (a.k.a. the query sequences) reported with respect to
the + strand. 
</p>
</td></tr>
<tr valign="top"><td><code>seqnames</code></td>
<td>

<p>A factor-<a href="../../S4Vectors/html/Rle.html">Rle</a> <em>parallel</em> to <code>x</code>.
For each <code>i</code>, <code>seqnames[i]</code> must be the name of the reference
sequence of the i-th alignment.
</p>
</td></tr>
<tr valign="top"><td><code>pos</code></td>
<td>

<p>An integer vector <em>parallel</em> to <code>x</code>.
For each <code>i</code>, <code>pos[i]</code> must be the 1-based position
on the reference sequence of the first aligned letter in <code>x[[i]]</code>.
</p>
</td></tr>
<tr valign="top"><td><code>cigar</code></td>
<td>

<p>A character vector <em>parallel</em> to <code>x</code>. Contains the extended
CIGAR strings of the alignments.
</p>
</td></tr>
<tr valign="top"><td><code>at</code></td>
<td>

<p>A <a href="../../GenomicRanges/html/GPos.html">GPos</a> object containing the genomic positions
of interest. <code>seqlevels(at)</code> must be identical to
<code>levels(seqnames)</code>.
</p>
<p>If <code>at</code> is not a <a href="../../GenomicRanges/html/GPos.html">GPos</a> object,
<code>pileLettersAt</code> will first try to turn it into one by calling
the <code><a href="../../GenomicRanges/html/GPos.html">GPos</a>()</code> constructor function on it.
So for example <code>at</code> can be a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object
(or any other <a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> derivative), and, in
that case, each range in it will be interpreted as a run of adjacent
genomic positions. See <code>?<a href="../../GenomicRanges/html/GPos.html">GPos</a></code> in the
<span class="pkg">GenomicRanges</span> package for more information.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>x</code>, <code>seqnames</code>, <code>pos</code>, <code>cigar</code> must be 4 <em>parallel</em>
vectors describing N aligned reads.
</p>


<h3>Value</h3>

<p>An <a href="../../Biostrings/html/XStringSet.html">XStringSet</a> (typically <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a>)
object <em>parallel</em> to <code>at</code> (i.e. with 1 string per genomic
position).
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li><p> The <code>pileup</code> and <code>applyPileups</code> functions defined in the
<span class="pkg">Rsamtools</span> package, as well as the SAMtools mpileup command
(available at <a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a> as part of the
SAMtools project), for more powerful flexible alternatives.
</p>
</li>
<li><p> The <code><a href="stackStringsFromGAlignments.html">stackStringsFromGAlignments</a></code> function
for stacking the read sequences (or their quality strings)
stored in a <a href="GAlignments-class.html">GAlignments</a> object or a BAM file.
</p>
</li>
<li> <p><a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> objects in the <span class="pkg">Biostrings</span>
package.
</p>
</li>
<li> <p><a href="../../GenomicRanges/html/GPos.html">GPos</a> objects in the <span class="pkg">GenomicRanges</span>
package.
</p>
</li>
<li> <p><a href="GAlignments-class.html">GAlignments</a> objects.
</p>
</li>
<li> <p><a href="cigar-utils.html">cigar-utils</a> for the CIGAR utility functions used internally
by <code>pileLettersAt</code>.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## Input

##   - A BAM file:
bamfile &lt;- BamFile(system.file("extdata", "ex1.bam", package="Rsamtools"))
seqinfo(bamfile)  # to see the seqlevels and seqlengths
stackStringsFromBam(bamfile, param="seq1:1-21")  # a quick look at
                                                 # the reads

##   - A GPos object containing Genomic Positions Of Interest:
my_GPOI &lt;- GPos(c("seq1:1-5", "seq1:21-21", "seq1:1575-1575",
                  "seq2:1513-1514"))

## Some preliminary massage on 'my_GPOI'

seqinfo(my_GPOI) &lt;- merge(seqinfo(my_GPOI), seqinfo(bamfile))
seqlevels(my_GPOI) &lt;- seqlevelsInUse(my_GPOI)

## Load the BAM file in a GAlignments object. Note that we load only
## the reads aligned to the sequences in 'seqlevels(my_GPOI)'. Also,
## in order to be consistent with applyPileups() and SAMtools (m)pileup,
## we filter out the following BAM records:
##   - secondary alignments (flag bit 0x100);
##   - reads not passing quality controls (flag bit 0x200);
##   - PCR or optical duplicates (flag bit 0x400).
## See ?ScanBamParam and the SAM Spec for more information. 

which &lt;- as(seqinfo(my_GPOI), "GRanges")
flag &lt;- scanBamFlag(isSecondaryAlignment=FALSE,
                    isNotPassingQualityControls=FALSE,
                    isDuplicate=FALSE)
what &lt;- c("seq", "qual")
param &lt;- ScanBamParam(flag=flag, what=c("seq", "qual"), which=which)
gal &lt;- readGAlignments(bamfile, param=param)
seqlevels(gal) &lt;- seqlevels(my_GPOI) 

## Extract the read sequences (a.k.a. query sequences) and quality
## strings. Both are reported with respect to the + strand.

qseq &lt;- mcols(gal)$seq
qual &lt;- mcols(gal)$qual

nucl_piles &lt;- pileLettersAt(qseq, seqnames(gal), start(gal), cigar(gal),
                            my_GPOI)
qual_piles &lt;- pileLettersAt(qual, seqnames(gal), start(gal), cigar(gal),
                            my_GPOI)
mcols(my_GPOI)$nucl_piles &lt;- nucl_piles
mcols(my_GPOI)$qual_piles &lt;- qual_piles
my_GPOI 

## Finally, to summarize A/C/G/T frequencies at each position:
alphabetFrequency(nucl_piles, baseOnly=TRUE)

## Note that the pileup() function defined in the Rsamtools package
## can be used to obtain a similar result:
scanbam_param &lt;- ScanBamParam(flag=flag, which=my_GPOI)
pileup_param &lt;- PileupParam(max_depth=5000,
                            min_base_quality=0,
                            distinguish_strands=FALSE)
pileup(bamfile, scanBamParam=scanbam_param, pileupParam=pileup_param)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
