<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Stack the read sequences stored in a GAlignments object or a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for stackStringsFromBam {GenomicAlignments}"><tr><td>stackStringsFromBam {GenomicAlignments}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Stack the read sequences stored in a GAlignments object or a BAM file</h2>

<h3>Description</h3>

<p><code>stackStringsFromGAlignments</code> stacks the read sequences (or
their quality strings) stored in a <a href="GAlignments-class.html">GAlignments</a> object over
a user-specified region.
</p>
<p><code>stackStringsFromBam</code> stacks the read sequences (or their quality
strings) stored in a BAM file over a user-specified region.
</p>
<p><code>alphabetFrequencyFromBam</code> computes the alphabet frequency of the
reads over a user-specified region.
</p>
<p>All these functions take into account the CIGAR of each read to <em>lay</em>
the read sequence (or its quality string) alongside the reference space.
This step ensures that each nucleotide in a read is associated with the
correct position on the reference sequence.
</p>


<h3>Usage</h3>

<pre>
stackStringsFromGAlignments(x, region, what="seq",
                    D.letter="-", N.letter=".",
                    Lpadding.letter="+",
                    Rpadding.letter="+")

stackStringsFromBam(file, index=file, param,
                    what="seq", use.names=FALSE,
                    D.letter="-", N.letter=".",
                    Lpadding.letter="+", Rpadding.letter="+")

alphabetFrequencyFromBam(file, index=file, param, what="seq", ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="GAlignments-class.html">GAlignments</a> object with the read sequences in the <code>"seq"</code>
metadata column (if <code>what</code> is set to <code>"seq"</code>), or with the
the read quality strings in the <code>"qual"</code> metadata column (if
<code>what</code> is set to <code>"qual"</code>).
Such an object is typically obtained by specifying
<code>param=<a href="../../Rsamtools/html/ScanBamParam.html">ScanBamParam</a>(what=c("seq", "qual"))</code>
when reading a BAM file with calling <code><a href="readGAlignments.html">readGAlignments</a>()</code>.
</p>
</td></tr>
<tr valign="top"><td><code>region</code></td>
<td>

<p>A <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object with exactly 1 genomic range.
The read sequences (or read quality strings) will be stacked over
that region.
</p>
</td></tr>
<tr valign="top"><td><code>what</code></td>
<td>

<p>A single string. Either <code>"seq"</code> or <code>"qual"</code>. If <code>"seq"</code>
(the default), the read sequences will be stacked. If <code>"qual"</code>,
the read quality strings will be stacked.
</p>
</td></tr>
<tr valign="top"><td><code>D.letter, N.letter</code></td>
<td>

<p>A single letter used as a filler for injections. The 2 arguments are
passed down to the <code><a href="sequenceLayer.html">sequenceLayer</a></code> function.
See <code>?<a href="sequenceLayer.html">sequenceLayer</a></code> for more details.
</p>
</td></tr>
<tr valign="top"><td><code>Lpadding.letter, Rpadding.letter</code></td>
<td>

<p>A single letter to use for padding the sequences on the left, and another
one to use for padding on the right. The 2 arguments are passed down to
the <code><a href="../../Biostrings/html/stackStrings.html">stackStrings</a></code> function defined in the
<span class="pkg">Biostrings</span> package.
See <code>?<a href="../../Biostrings/html/stackStrings.html">stackStrings</a></code> in the <span class="pkg">Biostrings</span>
package for more details.
</p>
</td></tr>
<tr valign="top"><td><code>file, index</code></td>
<td>

<p>The path to the BAM file containing the reads, and to its index file,
respectively. The latter is given <em>without</em> the '.bai'
extension. See <code><a href="../../Rsamtools/html/scanBam.html">scanBam</a></code> for more information.
</p>
</td></tr>
<tr valign="top"><td><code>param</code></td>
<td>

<p>A <a href="../../Rsamtools/html/ScanBamParam-class.html">ScanBamParam</a> object containing exactly 1 genomic region
(i.e. <code>unlist(bamWhich(param))</code> must have length 1).
Alternatively, <code>param</code> can be a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or
<a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object containing exactly 1 genomic
region (the strand will be ignored in case of a
<a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object), or a character string specifying
a single genomic region (in the <code>"chr14:5201-5300"</code> format).
</p>
</td></tr>
<tr valign="top"><td><code>use.names</code></td>
<td>

<p>Use the query template names (QNAME field) as the names of the returned
object? If not (the default), then the returned object has no names.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Further arguments to be passed to <a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a>.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>stackStringsFromGAlignments</code> performs the 3 following steps:
</p>

<ol>
<li><p> Subset <a href="GAlignments-class.html">GAlignments</a> object <code>x</code> to keep only the
alignments that overlap with the specified region.
</p>
</li>
<li><p> Lay the sequences in <code>x</code> alongside the reference space,
using their CIGARs. This is done with the
<code><a href="sequenceLayer.html">sequenceLayer</a></code> function.
</p>
</li>
<li><p> Stack them on the specified region. This is done with the
<code><a href="../../Biostrings/html/stackStrings.html">stackStrings</a></code> function defined in the
<span class="pkg">Biostrings</span> package.
</p>
</li></ol>

<p><code>stackStringsFromBam</code> performs the 3 following steps:
</p>

<ol>
<li><p> Load the read sequences (or their quality strings) from the BAM
file. Only the read sequences that overlap with the specified region
are loaded. This is done with the <code><a href="readGAlignments.html">readGAlignments</a></code>
function. Note that if the file contains paired-end reads, the
pairing is ignored.
</p>
</li>
<li><p> Same as <code>stackStringsFromGAlignments</code>.
</p>
</li>
<li><p> Same as <code>stackStringsFromGAlignments</code>.
</p>
</li></ol>

<p><code>alphabetFrequencyFromBam</code> also performs steps 1. and 2. but,
instead of stacking the sequences at step 3., it computes the nucleotide
frequencies for each genomic position in the specified region.
</p>


<h3>Value</h3>

<p>For <code>stackStringsFromBam</code>: A rectangular (i.e. constant-width)
<a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object (if <code>what</code> is <code>"seq"</code>)
or <a href="../../Biostrings/html/BStringSet.html">BStringSet</a> object (if <code>what</code> is <code>"qual"</code>).
</p>
<p>For <code>alphabetFrequencyFromBam</code>: By default a matrix like one returned
by <code><a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a></code>. The matrix has 1 row per
nucleotide position in the specified region.
</p>


<h3>Note</h3>

<p>TWO IMPORTANT CAVEATS ABOUT <code>stackStringsFromGAlignments</code> AND
<code>stackStringsFromBam</code>:
</p>
<p>Specifying a big genomic region, say &gt;= 100000 bp, can require a lot of
memory (especially with high coverage reads) so is not recommended.
See the <code><a href="pileLettersAt.html">pileLettersAt</a></code> function for piling the read letters
on top of a set of genomic positions, which is more flexible and more
memory efficient.
</p>
<p>Paired-end reads are treated as single-end reads (i.e. they're not paired).
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li><p> The <code><a href="pileLettersAt.html">pileLettersAt</a></code> function for piling the letters
of a set of aligned reads on top of a set of genomic positions.
</p>
</li>
<li><p> The <code><a href="readGAlignments.html">readGAlignments</a></code> function for loading read
sequences (or their quality strings) from a BAM file (via a
<a href="GAlignments-class.html">GAlignments</a> object).
</p>
</li>
<li><p> The <code><a href="sequenceLayer.html">sequenceLayer</a></code> function for laying read sequences
alongside the reference space, using their CIGARs.
</p>
</li>
<li><p> The <code><a href="../../Biostrings/html/stackStrings.html">stackStrings</a></code> function in the
<span class="pkg">Biostrings</span> package for stacking an arbitrary
<a href="../../Biostrings/html/XStringSet.html">XStringSet</a> object.
</p>
</li>
<li><p> The <code><a href="../../Biostrings/html/alphabetFrequency.html">alphabetFrequency</a></code> function in the
<span class="pkg">Biostrings</span> package.
</p>
</li>
<li><p> The SAMtools mpileup command available at
<a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a> as part of the
SAMtools project.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. EXAMPLE WITH TOY DATA
## ---------------------------------------------------------------------

bamfile1 &lt;- BamFile(system.file("extdata", "ex1.bam", package="Rsamtools"))

region1 &lt;- GRanges("seq1", IRanges(1, 60))  # region of interest

## Stack the read sequences directly from the BAM file:
stackStringsFromBam(bamfile1, param=region1, use.names=TRUE)

## or, alternatively, from a GAlignments object:
gal1 &lt;- readGAlignments(bamfile1, param=ScanBamParam(what="seq"),
                        use.names=TRUE)
stackStringsFromGAlignments(gal1, region1)

## Compute the "consensus matrix" (1 column per nucleotide position
## in the region of interest):
af &lt;- alphabetFrequencyFromBam(bamfile1, param=region1, baseOnly=TRUE)
cm1a &lt;- t(af[ , DNA_BASES])
cm1a

## Stack their quality strings:
stackStringsFromBam(bamfile1, param=region1, what="qual")

## Control the number of reads to display:
options(showHeadLines=18)
options(showTailLines=6)
stackStringsFromBam(bamfile1, param=GRanges("seq1", IRanges(61, 120)))

stacked_qseq &lt;- stackStringsFromBam(bamfile1, param="seq2:1509-1519")
stacked_qseq  # deletion in read 13
af &lt;- alphabetFrequencyFromBam(bamfile1, param="seq2:1509-1519",
                                baseOnly=TRUE)
cm1b &lt;- t(af[ , DNA_BASES])  # consensus matrix
cm1b

## Sanity check:
stopifnot(identical(consensusMatrix(stacked_qseq)[DNA_BASES, ], cm1b))

stackStringsFromBam(bamfile1, param="seq2:1509-1519", what="qual")

## ---------------------------------------------------------------------
## B. EXAMPLE WITH REAL DATA
## ---------------------------------------------------------------------

library(RNAseqData.HNRNPC.bam.chr14)
bamfile2 &lt;- BamFile(RNAseqData.HNRNPC.bam.chr14_BAMFILES[1])

## Region of interest:
region2 &lt;- GRanges("chr14", IRanges(19650095, 19650159))

## Stack the read sequences directly from the BAM file:
stackStringsFromBam(bamfile2, param=region2)

## or, alternatively, from a GAlignments object:
gal2 &lt;- readGAlignments(bamfile2, param=ScanBamParam(what="seq"))
stackStringsFromGAlignments(gal2, region2)

af &lt;- alphabetFrequencyFromBam(bamfile2, param=region2, baseOnly=TRUE)
cm2 &lt;- t(af[ , DNA_BASES])  # consensus matrix
cm2

## ---------------------------------------------------------------------
## C. COMPUTE READ CONSENSUS SEQUENCE FOR REGION OF INTEREST
## ---------------------------------------------------------------------

## Let's write our own little naive function to go from consensus matrix
## to consensus sequence. For each nucleotide position in the region of
## interest (i.e. each column in the matrix), we select the letter with
## highest frequency. We also use special letter "*" at positions where
## there is a tie, and special letter "." at positions where all the
## frequencies are 0 (a particular type of tie):
cm_to_cs &lt;- function(cm)
{
    stopifnot(is.matrix(cm))
    nr &lt;- nrow(cm)
    rnames &lt;- rownames(cm)
    stopifnot(!is.null(rnames) &amp;&amp; all(nchar(rnames) == 1L))
    selection &lt;- apply(cm, 2,
                       function(x) {
                         i &lt;- which.max(x)
                         if (x[i] == 0L)
                           return(nr + 1L)
                         if (sum(x == x[i]) != 1L)
                           return(nr + 2L)
                         i
                       })
    paste0(c(rnames, ".", "*")[selection], collapse="")
}

cm_to_cs(cm1a)
cm_to_cs(cm1b)
cm_to_cs(cm2)

## Note that the consensus sequences we obtain are relative to the
## plus strand of the reference sequence.
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicAlignments</em> version 1.26.0 <a href="00Index.html">Index</a>]</div>
</body></html>
