<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Basic usage of an Ensembl based annotation database</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for EnsDb-class {ensembldb}"><tr><td>EnsDb-class {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Basic usage of an Ensembl based annotation database</h2>

<h3>Description</h3>

<p>The <code>EnsDb</code> class provides access to an Ensembl-based annotation
package. This help page describes functions to get some basic
informations from such an object.
</p>


<h3>Usage</h3>

<pre>

## S4 method for signature 'EnsDb'
dbconn(x)

## S4 method for signature 'EnsDb'
ensemblVersion(x)

## S4 method for signature 'EnsDb'
listColumns(x, table, skip.keys=TRUE, ...)

## S4 method for signature 'EnsDb'
listGenebiotypes(x, ...)

## S4 method for signature 'EnsDb'
listTxbiotypes(x, ...)

## S4 method for signature 'EnsDb'
listTables(x, ...)

## S4 method for signature 'EnsDb'
metadata(x, ...)

## S4 method for signature 'EnsDb'
organism(object)

## S4 method for signature 'EnsDb'
returnFilterColumns(x)

## S4 method for signature 'EnsDb'
returnFilterColumns(x)

## S4 replacement method for signature 'EnsDb'
returnFilterColumns(x) &lt;- value

## S4 method for signature 'EnsDb'
seqinfo(x)

## S4 method for signature 'EnsDb'
seqlevels(x)

## S4 method for signature 'EnsDb'
updateEnsDb(x, ...)

</pre>


<h3>Arguments</h3>

<p>(in alphabetic order)
</p>
<table summary="R argblock">
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments.
Not used.
</p>
</td></tr>
<tr valign="top"><td><code>object</code></td>
<td>

<p>For <code>organism</code>: an <code>EnsDb</code> instance.
</p>
</td></tr>
<tr valign="top"><td><code>skip.keys</code></td>
<td>

<p>for <code>listColumns</code>: whether primary and foreign keys (not
being e.g. <code>"gene_id"</code> or alike) should be returned or not. By
default these will not be returned.
</p>
</td></tr>
<tr valign="top"><td><code>table</code></td>
<td>

<p>For <code>listColumns</code>: optionally specify the table name(s) for
which the columns should be returned.
</p>
</td></tr>
<tr valign="top"><td><code>value</code></td>
<td>

<p>For <code>returnFilterColumns</code>: a logical of length one specifying
whether columns that are used for eventual filters should also be
returned.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>An <code>EnsDb</code> instance.
</p>
</td></tr>
</table>


<h3>Value</h3>


<dl>
<dt>For <code>connection</code></dt><dd>
<p>The SQL connection to the RSQLite database.
</p>
</dd>
<dt>For <code>EnsDb</code></dt><dd>
<p>An <code>EnsDb</code> instance.
</p>
</dd>
<dt>For <code>lengthOf</code></dt><dd>
<p>A named integer vector with the length of the genes or transcripts.
</p>
</dd>
<dt>For <code>listColumns</code></dt><dd>
<p>A character vector with the column names.
</p>
</dd>
<dt>For <code>listGenebiotypes</code></dt><dd>
<p>A character vector with the biotypes of the genes in the database.
</p>
</dd>
<dt>For <code>listTxbiotypes</code></dt><dd>
<p>A character vector with the biotypes of the transcripts in the database.
</p>
</dd>
<dt>For <code>listTables</code></dt><dd>
<p>A list with the names corresponding to the database table names
and the elements being the attribute (column) names of the table.
</p>
</dd>
<dt>For <code>metadata</code></dt><dd>
<p>A <code>data.frame</code>.
</p>
</dd>
<dt>For <code>organism</code></dt><dd>
<p>A character string.
</p>
</dd>
<dt>For <code>returnFilterColumns</code></dt><dd>
<p>A logical of length 1.
</p>
</dd>
<dt>For <code>seqinfo</code></dt><dd>
<p>A <code>Seqinfo</code> class.
</p>
</dd>
<dt>For <code>updateEnsDb</code></dt><dd>
<p>A <code>EnsDb</code> object.
</p>
</dd>
</dl>



<h3>Objects from the Class</h3>

<p>A connection to the respective annotation database is created upon
loading of an annotation package created with the
<code><a href="makeEnsemblDbPackage.html">makeEnsembldbPackage</a></code> function. In addition, the
<code><a href="EnsDb.html">EnsDb</a></code> constructor specifying the SQLite database file can be
called to generate an instance of the object (see
<code><a href="makeEnsemblDbPackage.html">makeEnsemblSQLiteFromTables</a></code> for an example).
</p>


<h3>Slots</h3>


<dl>
<dt>ensdb</dt><dd>
<p>Object of class <code>"DBIConnection"</code>: the
connection to the database.
</p>
</dd>
<dt>tables</dt><dd>
<p>Named list of database table columns with the names being the
database table names. The tables are ordered by their degree,
i.e. the number of other tables they can be joined with.
</p>
</dd>
<dt>.properties</dt><dd>
<p>Internal list storing user-defined properties. Should not be
directly accessed.
</p>
</dd>
</dl>



<h3>Methods and Functions</h3>


<dl>
<dt>dbconn</dt><dd>
<p>Returns the connection to the internal SQL database.
</p>
</dd>
<dt>ensemblVersion</dt><dd>
<p>Returns the Ensembl version on which the package was built.
</p>
</dd>
<dt>listColumns</dt><dd>
<p>Lists all columns of all tables in the database, or, if
<code>table</code> is specified, of the respective table.
</p>
</dd>
<dt>listGenebiotypes</dt><dd>
<p>Lists all gene biotypes defined in the database.
</p>
</dd>
<dt>listTxbiotypes</dt><dd>
<p>Lists all transcript biotypes defined in the database.
</p>
</dd>
<dt>listTables</dt><dd>
<p>Returns a named list of database table columns (names of the
list being the database table names).
</p>
</dd>
<dt>metadata</dt><dd>
<p>Returns a <code>data.frame</code> with the metadata information from the
database, i.e. informations about the Ensembl version or Genome
build the database was build upon.
</p>
</dd>
<dt>organism</dt><dd>
<p>Returns the organism name (e.g. <code>"homo_sapiens"</code>).
</p>
</dd>
<dt>returnFilterColumns, returnFilterColumns&lt;-</dt><dd>
<p>Get or set the option which results in columns that are used for
eventually specified filters to be added as result columns. The
default value is <code>TRUE</code> (i.e. filter columns are returned).
</p>
</dd>
<dt>seqinfo</dt><dd>
<p>Returns the sequence/chromosome information from the database.
</p>
</dd>
<dt>seqlevels</dt><dd>
<p>Returns the chromosome/sequence names that are available in the
database.
</p>
</dd>
<dt>show</dt><dd>
<p>Displays some informations from the database.
</p>
</dd>
<dt>updateEnsDb</dt><dd>
<p>Updates the <code>EnsDb</code> object to the most recent implementation.
</p>
</dd>
</dl>



<h3>Note</h3>

<p>While a column named <code>"tx_name"</code> is listed by the
<code>listTables</code> and <code>listColumns</code> method, no such column is
present in the database. Transcript names returned by the methods are
actually the transcript IDs. This <em>virtual</em> column was only
introduced to be compliant with <code>TxDb</code> objects (which provide
transcript names).
</p>


<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="EnsDb.html">EnsDb</a></code>,
<code><a href="makeEnsemblDbPackage.html">makeEnsembldbPackage</a></code>,
<code><a href="EnsDb-exonsBy.html">exonsBy</a></code>, <code><a href="EnsDb-exonsBy.html">genes</a></code>,
<code><a href="EnsDb-exonsBy.html">transcripts</a></code>,
<code><a href="makeEnsemblDbPackage.html">makeEnsemblSQLiteFromTables</a></code>
</p>
<p><code><a href="global-filters.html">addFilter</a></code> for globally adding filters to an <code>EnsDb</code>
object.
</p>


<h3>Examples</h3>

<pre>

library(EnsDb.Hsapiens.v86)

## Display some information:
EnsDb.Hsapiens.v86

## Show the tables along with its columns
listTables(EnsDb.Hsapiens.v86)

## For what species is this database?
organism(EnsDb.Hsapiens.v86)

## What Ensembl version if the database based on?
ensemblVersion(EnsDb.Hsapiens.v86)

## Get some more information from the database
metadata(EnsDb.Hsapiens.v86)

## Get all the sequence names.
seqlevels(EnsDb.Hsapiens.v86)

## List all available gene biotypes from the database:
listGenebiotypes(EnsDb.Hsapiens.v86)

## List all available transcript biotypes:
listTxbiotypes(EnsDb.Hsapiens.v86)

## Update the EnsDb; this is in most instances not necessary at all.
updateEnsDb(EnsDb.Hsapiens.v86)

######    returnFilterColumns
returnFilterColumns(EnsDb.Hsapiens.v86)

## Get protein coding genes on chromosome X, specifying to return
## only columns gene_name as additional column.
genes(EnsDb.Hsapiens.v86, filter=list(SeqNameFilter("X"),
                                      GeneBiotypeFilter("protein_coding")),
      columns=c("gene_name"))
## By default we get also the gene_biotype column as the data was filtered
## on this column.

## This can be changed using the returnFilterColumns option
returnFilterColumns(EnsDb.Hsapiens.v86) &lt;- FALSE
genes(EnsDb.Hsapiens.v86, filter=list(SeqNameFilter("X"),
                                      GeneBiotypeFilter("protein_coding")),
      columns=c("gene_name"))


</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
