<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Calculating lengths of features</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for lengthOf {ensembldb}"><tr><td>lengthOf {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Calculating lengths of features</h2>

<h3>Description</h3>

<p>These methods allow to calculate the lengths of features (transcripts, genes,
CDS, 3' or 5' UTRs) defined in an <code>EnsDb</code> object or database.
</p>


<h3>Usage</h3>

<pre>

## S4 method for signature 'EnsDb'
lengthOf(x, of="gene", filter = AnnotationFilterList())

</pre>


<h3>Arguments</h3>

<p>(In alphabetic order)
</p>
<table summary="R argblock">
<tr valign="top"><td><code>filter</code></td>
<td>

<p>A filter describing which results to retrieve from the database. Can
be a single object extending
<code><a href="../../AnnotationFilter/html/AnnotationFilter.html">AnnotationFilter</a></code>, an
<code><a href="../../AnnotationFilter/html/AnnotationFilterList.html">AnnotationFilterList</a></code> object
combining several such objects or a <code>formula</code> representing a
filter expression (see examples below or
<code><a href="../../AnnotationFilter/html/AnnotationFilter.html">AnnotationFilter</a></code> for more details).
</p>
</td></tr>
<tr valign="top"><td><code>of</code></td>
<td>

<p>for <code>lengthOf</code>: whether the length of genes or
transcripts should be retrieved from the database.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>For <code>lengthOf</code>: either an <code>EnsDb</code> or a
<code>GRangesList</code> object. For all other methods an <code>EnsDb</code>
instance.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>For <code>lengthOf</code>: see method description above.
</p>


<h3>Methods and Functions</h3>


<dl>
<dt>lengthOf</dt><dd>
<p>Retrieve the length of genes or transcripts from the
database. The length is the sum of the lengths of all exons of a
transcript or a gene. In the latter case the exons are first reduced
so that the length corresponds to the part of the genomic sequence covered by
the exons.
</p>
<p>Note: in addition to this method, also the
<code><a href="../../GenomicFeatures/html/transcriptLengths.html">transcriptLengths</a></code> function in the
<code>GenomicFeatures</code> package can be used.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="EnsDb-exonsBy.html">exonsBy</a></code>
<code><a href="EnsDb-exonsBy.html">transcripts</a></code>
<code><a href="../../GenomicFeatures/html/transcriptLengths.html">transcriptLengths</a></code>
</p>


<h3>Examples</h3>

<pre>

library(EnsDb.Hsapiens.v86)
edb &lt;- EnsDb.Hsapiens.v86

#####    lengthOf
##
## length of a specific gene.
lengthOf(edb, filter = GeneIdFilter("ENSG00000000003"))

## length of a transcript
lengthOf(edb, of = "tx", filter = TxIdFilter("ENST00000494424"))

## Average length of all protein coding genes encoded on chromosomes X
mean(lengthOf(edb, of = "gene",
              filter = ~ gene_biotype == "protein_coding" &amp;
                  seq_name == "X"))

## Average length of all snoRNAs
mean(lengthOf(edb, of = "gene",
              filter = ~ gene_biotype == "snoRNA" &amp;
                  seq_name == "X"))

##### transcriptLengths
##
## Calculate the length of transcripts encoded on chromosome Y, including
## length of the CDS, 5' and 3' UTR.
len &lt;- transcriptLengths(edb, with.cds_len = TRUE, with.utr5_len = TRUE,
                         with.utr3_len = TRUE, filter = SeqNameFilter("Y"))
head(len)

</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
