<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Functionality related to DNA/RNA sequences</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for getGenomeFaFile {ensembldb}"><tr><td>getGenomeFaFile {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Functionality related to DNA/RNA sequences</h2>

<h3>Description</h3>

<p>Utility functions related to RNA/DNA sequences, such as extracting
RNA/DNA sequences for features defined in <code>Ensb</code>.
</p>


<h3>Usage</h3>

<pre>

## S4 method for signature 'EnsDb'
getGenomeFaFile(x, pattern="dna.toplevel.fa")

## S4 method for signature 'EnsDb'
getGenomeTwoBitFile(x)

</pre>


<h3>Arguments</h3>

<p>(In alphabetic order)
</p>
<table summary="R argblock">
<tr valign="top"><td><code>pattern</code></td>
<td>

<p>For method <code>getGenomeFaFile</code>: the pattern to be used to
identify the fasta file representing genomic DNA sequence.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>An <code>EnsDb</code> instance.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>For <code>getGenomeFaFile</code>: a <code><a href="../../Rsamtools/html/FaFile-class.html">FaFile-class</a></code>
object with the genomic DNA sequence.
</p>
<p>For <code>getGenomeTwoBitFile</code>: a <code><a href="../../rtracklayer/html/TwoBitFile-class.html">TwoBitFile-class</a></code>
object with the genome sequence.
</p>


<h3>Methods and Functions</h3>


<dl>
<dt>getGenomeFaFile</dt><dd>
<p>Returns a <code><a href="../../Rsamtools/html/FaFile-class.html">FaFile-class</a></code> (defined in
<code>Rsamtools</code>) with the genomic sequence of the genome build
matching the Ensembl version of the <code>EnsDb</code> object.
The file is retrieved using the <code>AnnotationHub</code> package,
thus, at least for the first invocation, an internet connection is
required to locate and download the file; subsequent calls will
load the cached file instead.
If no fasta file for the actual Ensembl version is available the
function tries to identify a file matching the species and genome
build version of the closest Ensembl release and returns that
instead.
See the vignette for an example to work with such files.
</p>
</dd>
<dt>getGenomeTwoBitFile</dt><dd>
<p>Returns a <code><a href="../../rtracklayer/html/TwoBitFile-class.html">TwoBitFile-class</a></code> (defined in the
<code>rtracklayer</code> package) with the genomeic sequence of the
genome build matching the Ensembl version of the <code>EnsDb</code>
object. The file is retrieved from <code>AnnotationHub</code> and hence
requires (at least for the first query) an active internet
connection to download the respective resource. If no DNA sequence
matching the Ensembl version of <code>x</code> is available, the
function tries to find the genomic sequence of the best matching
genome build (closest Ensembl release) and returns that.
</p>
<p>See the <code>ensembldb</code> vignette for details.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="EnsDb-exonsBy.html">transcripts</a></code>
<code><a href="EnsDb-exonsBy.html">exonsBy</a></code>
</p>


<h3>Examples</h3>

<pre>

## Loading an EnsDb for Ensembl version 86 (genome GRCh38):
library(EnsDb.Hsapiens.v86)
edb &lt;- EnsDb.Hsapiens.v86

## Not run: 
    ## Retrieve a TwoBitFile with the gneomic DNA sequence matching the organism,
    ## genome release version and, if possible, the Ensembl version of the
    ## EnsDb object.
    Dna &lt;- getGenomeTwoBitFile(edb)
    ## Extract the transcript sequence for all transcripts encoded on chromosome
    ## Y.
    ##extractTranscriptSeqs(Dna, edb, filter=SeqNameFilter("Y"))


## End(Not run)

</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
