<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Integration into the AnnotationDbi framework</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for select {ensembldb}"><tr><td>select {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Integration into the AnnotationDbi framework</h2>

<h3>Description</h3>

<p>Several of the methods available for <code>AnnotationDbi</code> objects are
also implemented for <code>EnsDb</code> objects. This enables to extract
data from <code>EnsDb</code> objects in a similar fashion than from objects
inheriting from the base annotation package class
<code>AnnotationDbi</code>.
In addition to the <em>standard</em> usage, the <code>select</code> and
<code>mapIds</code> for <code>EnsDb</code> objects support also the filter
framework of the ensembdb package and thus allow to perform more
fine-grained queries to retrieve data.
</p>


<h3>Usage</h3>

<pre>

## S4 method for signature 'EnsDb'
columns(x)
## S4 method for signature 'EnsDb'
keys(x, keytype, filter,...)
## S4 method for signature 'EnsDb'
keytypes(x)
## S4 method for signature 'EnsDb'
mapIds(x, keys, column, keytype, ..., multiVals)
## S4 method for signature 'EnsDb'
select(x, keys, columns, keytype, ...)

</pre>


<h3>Arguments</h3>

<p>(In alphabetic order)
</p>
<table summary="R argblock">
<tr valign="top"><td><code>column</code></td>
<td>

<p>For <code>mapIds</code>: the column to search on, i.e. from which values
should be retrieved.
</p>
</td></tr>
<tr valign="top"><td><code>columns</code></td>
<td>

<p>For <code>select</code>: the columns from which values should be
retrieved. Use the <code>columns</code> method to list all possible
columns.
</p>
</td></tr>
<tr valign="top"><td><code>keys</code></td>
<td>

<p>The keys/ids for which data should be retrieved from the
database. This can be either a character vector of keys/IDs, a
single filter object extending
<code><a href="../../AnnotationFilter/html/AnnotationFilter.html">AnnotationFilter</a></code>, an combination of
filters <code><a href="../../AnnotationFilter/html/AnnotationFilterList.html">AnnotationFilterList</a></code> or a
<code>formula</code> representing a filter expression (see
<code><a href="../../AnnotationFilter/html/AnnotationFilter.html">AnnotationFilter</a></code> for more details).
</p>
</td></tr>
<tr valign="top"><td><code>keytype</code></td>
<td>

<p>For <code>mapIds</code> and <code>select</code>: the type (column) that matches
the provided keys. This argument does not have to be specified if
argument <code>keys</code> is a filter object extending
<code>AnnotationFilter</code> or a <code>list</code> of such objects.
</p>
<p>For <code>keys</code>: which keys should be returned from the database.
</p>
</td></tr>
<tr valign="top"><td><code>filter</code></td>
<td>

<p>For <code>keys</code>: either a single object extending
<code>AnnotationFilter</code> or a list of such object to
retrieve only specific keys from the database.
</p>
</td></tr>
<tr valign="top"><td><code>multiVals</code></td>
<td>

<p>What should <code>mapIds</code> do when there are multiple values that
could be returned? Options are: <code>"first"</code> (default), <code>"list"</code>,
<code>"filter"</code>, <code>"asNA"</code>. See
<code>mapIds</code> in the <code>AnnotationDbi</code> package  
for a detailed description.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>The <code>EnsDb</code> object.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Not used.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>See method description above.
</p>


<h3>Methods and Functions</h3>


<dl>
<dt>columns</dt><dd>
<p>List all the columns that can be retrieved by the <code>mapIds</code>
and <code>select</code> methods. Note that these column names are
different from the ones supported by the <code><a href="EnsDb-exonsBy.html">genes</a></code>,
<code><a href="EnsDb-exonsBy.html">transcripts</a></code> etc. methods that can be listed by the
<code><a href="EnsDb-class.html">listColumns</a></code> method.
</p>
<p>Returns a character vector of supported column names.
</p>
</dd>
<dt>keys</dt><dd>
<p>Retrieves all keys from the column name specified with
<code>keytype</code>. By default (if <code>keytype</code> is not provided) it
returns all gene IDs. Note that <code>keytype="TXNAME"</code> will
return transcript ids, since no transcript names are available in
the database.
</p>
<p>Returns a character vector of IDs.
</p>
</dd>
<dt>keytypes</dt><dd>
<p>List all supported key types (column names).
</p>
<p>Returns a character vector of key types.
</p>
</dd>
<dt>mapIds</dt><dd>
<p>Retrieve the mapped ids for a set of keys that are of a particular
keytype. Argument <code>keys</code> can be either a character vector of
keys/IDs, a single filter object extending
<code>AnnotationFilter</code> or a list of such objects. For
the latter, the argument <code>keytype</code> does not have to be
specified. Importantly however, if the filtering system is used,
the ordering of the results might not represent the ordering of
the keys.
</p>
<p>The method usually returns a named character vector or, depending
on the argument <code>multiVals</code> a named list, with names
corresponding to the keys (same ordering is only guaranteed if
<code>keys</code> is a character vector).
</p>
</dd>
<dt>select</dt><dd>
<p>Retrieve the data as a <code>data.frame</code> based on parameters for
selected <code>keys</code>, <code>columns</code> and <code>keytype</code>
arguments. Multiple matches of the keys are returned in one row
for each possible match. Argument <code>keys</code> can be either a
character vector of keys/IDs, a single filter object extending
<code>AnnotationFilter</code> or a list of such objects. For
the latter, the argument <code>keytype</code> does not have to be
specified.
</p>
<p>Note that values from a column <code>"TXNAME"</code> will be the same
than for a column <code>"TXID"</code>, since internally no database
column <code>"tx_name"</code> is present and the column is thus mapped
to <code>"tx_id"</code>.
</p>
<p>Returns a <code>data.frame</code> with the column names corresponding to
the argument <code>columns</code> and rows with all data matching the
criteria specified with <code>keys</code>.
</p>
<p>The use of <code>select</code> without filters or keys and without
restricting to specicic columns is strongly discouraged, as the
SQL query to join all of the tables, especially if protein
annotation data is available is very expensive.
</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="EnsDb-class.html">listColumns</a></code>
<code><a href="EnsDb-exonsBy.html">transcripts</a></code>
</p>


<h3>Examples</h3>

<pre>

library(EnsDb.Hsapiens.v86)
edb &lt;- EnsDb.Hsapiens.v86

## List all supported keytypes.
keytypes(edb)

## List all supported columns for the select and mapIds methods.
columns(edb)

## List /real/ database column names.
listColumns(edb)

## Retrieve all keys corresponding to transcript ids.
txids &lt;- keys(edb, keytype = "TXID")
length(txids)
head(txids)

## Retrieve all keys corresponding to gene names of genes encoded on chromosome X
gids &lt;- keys(edb, keytype = "GENENAME", filter = SeqNameFilter("X"))
length(gids)
head(gids)

## Get a mapping of the genes BCL2 and BCL2L11 to all of their
## transcript ids and return the result as list
maps &lt;- mapIds(edb, keys = c("BCL2", "BCL2L11"), column = "TXID",
               keytype = "GENENAME", multiVals = "list")
maps

## Perform the same query using a combination of a GeneNameFilter and a
## TxBiotypeFilter to just retrieve protein coding transcripts for these
## two genes.
mapIds(edb, keys = list(GeneNameFilter(c("BCL2", "BCL2L11")),
                        TxBiotypeFilter("protein_coding")), column = "TXID",
       multiVals = "list")

## select:
## Retrieve all transcript and gene related information for the above example.
select(edb, keys = list(GeneNameFilter(c("BCL2", "BCL2L11")),
                        TxBiotypeFilter("protein_coding")),
       columns = c("GENEID", "GENENAME", "TXID", "TXBIOTYPE", "TXSEQSTART",
                   "TXSEQEND", "SEQNAME", "SEQSTRAND"))

## Get all data for genes encoded on chromosome Y
Y &lt;- select(edb, keys = "Y", keytype = "SEQNAME")
head(Y)
nrow(Y)

## Get selected columns for all lincRNAs encoded on chromosome Y. Here we use
## a filter expression to define what data to retrieve.
Y &lt;- select(edb, keys = ~ seq_name == "Y" &amp; gene_biotype == "lincRNA",
            columns = c("GENEID", "GENEBIOTYPE", "TXID", "GENENAME"))
head(Y)
nrow(Y)

</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
