<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Filters supported by ensembldb</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for Filter-classes {ensembldb}"><tr><td>Filter-classes {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Filters supported by ensembldb</h2>

<h3>Description</h3>

<p><code>ensembldb</code> supports most of the filters from the <a href="../../AnnotationFilter/html/AnnotationFilter.html">AnnotationFilter</a>
package to retrieve specific content from <a href="EnsDb.html">EnsDb</a> databases. These filters
can be passed to the methods such as <code><a href="EnsDb-exonsBy.html">genes()</a></code> with the <code>filter</code> parameter
or can be added as a <em>global</em> filter to an <code>EnsDb</code> object (see
<code><a href="global-filters.html">addFilter()</a></code> for more details). Use <code><a href="../../AnnotationFilter/html/AnnotationFilter.html">supportedFilters()</a></code> to get an
overview of all filters supported by <code>EnsDb</code> object.
</p>
<p><code>seqnames</code>: accessor for the sequence names of the <code>GRanges</code>
object within a <code>GRangesFilter</code>.
</p>
<p><code>seqnames</code>: accessor for the <code>seqlevels</code> of the <code>GRanges</code>
object within a <code>GRangesFilter</code>.
</p>
<p><code>supportedFilters</code> returns a <code>data.frame</code> with the
names of all filters and the corresponding field supported by the
<code>EnsDb</code> object.
</p>


<h3>Usage</h3>

<pre>
OnlyCodingTxFilter()

ProtDomIdFilter(value, condition = "==")

ProteinDomainIdFilter(value, condition = "==")

ProteinDomainSourceFilter(value, condition = "==")

UniprotDbFilter(value, condition = "==")

UniprotMappingTypeFilter(value, condition = "==")

TxSupportLevelFilter(value, condition = "==")

## S4 method for signature 'GRangesFilter'
seqnames(x)

## S4 method for signature 'GRangesFilter'
seqlevels(x)

## S4 method for signature 'EnsDb'
supportedFilters(object, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>value</code></td>
<td>
<p>The value(s) for the filter. For <code>GRangesFilter</code> it has to be a
<code>GRanges</code> object.</p>
</td></tr>
<tr valign="top"><td><code>condition</code></td>
<td>
<p><code>character(1)</code> specifying the <em>condition</em> of the
filter. For <code>character</code>-based filters (such as
<code>GeneIdFilter</code>) <code>"=="</code>, <code>"!="</code>, <code>"startsWith"</code> and <code>"endsWith"</code> are
supported. Allowed values for <code>integer</code>-based filters (such as
<code>GeneStartFilter</code>) are <code>"=="</code>, <code>"!="</code>, <code>"&lt;"</code>. <code>"&lt;="</code>, <code>"&gt;"</code> and <code>"&gt;="</code>.</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>
<p>For <code>seqnames</code>, <code>seqlevels</code>: a <code>GRangesFilter</code> object.</p>
</td></tr>
<tr valign="top"><td><code>object</code></td>
<td>
<p>For <code>supportedFilters</code>: an <code>EnsDb</code> object.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>For <code>supportedFilters</code>: currently not used.</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>ensembldb</code> supports the following filters from the <code>AnnotationFilter</code>
package:
</p>

<ul>
<li> <p><code>GeneIdFilter</code>: filter based on the Ensembl gene ID.
</p>
</li>
<li> <p><code>GeneNameFilter</code>: filter based on the name of the gene as provided
Ensembl. In most cases this will correspond to the official gene symbol.
</p>
</li>
<li> <p><code>SymbolFilter</code> filter based on the gene names. <code>EnsDb</code> objects don't
have a dedicated <em>symbol</em> column, the filtering is hence based on the
gene names.
</p>
</li>
<li> <p><code>GeneBiotype</code>: filter based on the biotype of genes (e.g.
<code>"protein_coding"</code>).
</p>
</li>
<li> <p><code>GeneStartFilter</code>: filter based on the genomic start coordinate of genes.
</p>
</li>
<li> <p><code>GeneEndFilter</code>: filter based on the genomic end coordinate of genes.
</p>
</li>
<li> <p><code>EntrezidFilter</code>: filter based on the genes' NCBI Entrezgene ID.
</p>
</li>
<li> <p><code>TxIdFilter</code>: filter based on the Ensembld transcript ID.
</p>
</li>
<li> <p><code>TxNameFilter</code>: filter based on the Ensembld transcript ID; no transcript
names are provided in <code>EnsDb</code> databases.
</p>
</li>
<li> <p><code>TxBiotypeFilter</code>: filter based on the transcripts' biotype.
</p>
</li>
<li> <p><code>TxStartFilter</code>: filter based on the genomic start coordinate of the
transcripts.
</p>
</li>
<li> <p><code>TxEndFilter</code>: filter based on the genonic end coordinates of the
transcripts.
</p>
</li>
<li> <p><code>ExonIdFilter</code>: filter based on Ensembl exon IDs.
</p>
</li>
<li> <p><code>ExonRankFilter</code>: filter based on the index/rank of the exon within the
transcrips.
</p>
</li>
<li> <p><code>ExonStartFilter</code>: filter based on the genomic start coordinates of the
exons.
</p>
</li>
<li> <p><code>ExonEndFilter</code>: filter based on the genomic end coordinates of the exons.
</p>
</li>
<li> <p><code>GRangesFilter</code>: Allows to fetch features within or overlapping specified
genomic region(s)/range(s). This filter takes a <code>GRanges</code> object
as input and, if <code>type = "any"</code> (the default) will restrict results to
features (genes, transcripts or exons) that are partially overlapping the
region. Alternatively, by specifying <code>condition = "within"</code> it will
return features located within the range. In addition, the <code>GRangesFilter</code>
<code>condition = "start"</code>, <code>condition = "end"</code> and <code>condition = "equal"</code>
filtering for features with the same start or end coordinate or that are
equal to the <code>GRanges</code>.
</p>
<p>Note that the type of feature on which the filter is applied depends on
the method that is called, i.e. <code><a href="EnsDb-exonsBy.html">genes()</a></code> will filter on the
genomic coordinates of genes, <code><a href="EnsDb-exonsBy.html">transcripts()</a></code> on those of
transcripts and <code><a href="EnsDb-exonsBy.html">exons()</a></code> on exon coordinates.
</p>
<p>Calls to the methods <code><a href="EnsDb-exonsBy.html">exonsBy()</a></code>, <code><a href="EnsDb-exonsBy.html">cdsBy()</a></code> and
<code><a href="EnsDb-exonsBy.html">transcriptsBy()</a></code> use the start and end coordinates of the
feature type specified with argument <code>by</code> (i.e. <code>"gene"</code>,
<code>"transcript"</code> or <code>"exon"</code>) for the filtering.
</p>
<p>If the specified <code>GRanges</code> object defines multiple regions, all
features within (or overlapping) any of these regions are returned.
</p>
<p>Chromosome names/seqnames can be provided in UCSC format (e.g.
<code>"chrX"</code>) or Ensembl format (e.g. <code>"X"</code>); see <code><a href="EnsDb-seqlevels.html">seqlevelsStyle()</a></code> for
more information.
</p>
</li>
<li> <p><code>SeqNameFilter</code>: filter based on chromosome names.
</p>
</li>
<li> <p><code>SeqStrandFilter</code>: filter based on the chromosome strand. The strand can
be specified with <code>value = "+"</code>, <code>value = "-"</code>, <code>value = -1</code> or
<code>value = 1</code>.
</p>
</li>
<li> <p><code>ProteinIdFilter</code>: filter based on Ensembl protein IDs. This filter is
only supported if the <code>EnsDb</code> provides protein annotations; use the
<code><a href="hasProteinData-EnsDb-method.html">hasProteinData()</a></code> method to check.
</p>
</li>
<li> <p><code>UniprotFilter</code>: filter based on Uniprot IDs. This filter is only
supported if the <code>EnsDb</code> provides protein annotations; use the
<code><a href="hasProteinData-EnsDb-method.html">hasProteinData()</a></code> method to check.
</p>
</li></ul>

<p>In addition, the following filters are defined by <code>ensembldb</code>:
</p>

<ul>
<li> <p><code>TxSupportLevel</code>: allows to filter results using the provided transcript
support level. Support levels for transcripts are defined by Ensembl
based on the available evidences for a transcript with 1 being the
highest evidence grade and 5 the lowest level. This filter is only
supported on <code>EnsDb</code> databases with a db schema version higher 2.1.
</p>
</li>
<li> <p><code>UniprotDbFilter</code>: allows to filter results based on the specified Uniprot
database name(s).
</p>
</li>
<li> <p><code>UniprotMappingTypeFilter</code>: allows to filter results based on the mapping
method/type that was used to assign Uniprot IDs to Ensembl protein IDs.
</p>
</li>
<li> <p><code>ProtDomIdFilter</code>, <code>ProteinDomainIdFilter</code>: allows to retrieve entries
from the database matching the provided filter criteria based on their
protein domain ID (<em>protein_domain_id</em>).
</p>
</li>
<li> <p><code>ProteinDomainSourceFilter</code>: filter results based on the source
(database/method) defining the protein domain (e.g. <code>"pfam"</code>).
</p>
</li>
<li> <p><code>OnlyCodingTxFilter</code>: allows to retrieve entries only for protein coding
transcripts, i.e. transcripts with a CDS. This filter does not take any
input arguments.
</p>
</li></ul>



<h3>Value</h3>

<p>For <code>ProtDomIdFilter</code>: A <code>ProtDomIdFilter</code> object.
</p>
<p>For <code>ProteinDomainIdFilter</code>: A <code>ProteinDomainIdFilter</code> object.
</p>
<p>For <code>ProteinDomainSourceFilter</code>: A <code>ProteinDomainSourceFilter</code>
object.
</p>
<p>For <code>UniprotDbFilter</code>: A <code>UniprotDbFilter</code> object.
</p>
<p>For <code>UniprotMappingTypeFilter</code>: A <code>UniprotMappingTypeFilter</code> object.
</p>
<p>For <code>TxSupportLevel</code>: A <code>TxSupportLevel</code> object.
</p>
<p>For <code>supportedFilters</code>: a <code>data.frame</code> with the names and
the corresponding field of the supported filter classes.
</p>


<h3>Note</h3>

<p>For users of <code>ensembldb</code> version &lt; 2.0: in the <code>GRangesFilter</code> from the
<code>AnnotationFilter</code> package the <code>condition</code> parameter was renamed to <code>type</code>
(to be consistent with the <code>IRanges</code> package). In addition,
<code>condition = "overlapping"</code> is no longer recognized. To retrieve all
features overlapping the range <code>type = "any"</code> has to be used.
</p>
<p>Protein annotation based filters can only be used if the
<code>EnsDb</code> database contains protein annotations, i.e. if <code>hasProteinData</code>
is <code>TRUE</code>. Also, only protein coding transcripts will have protein
annotations available, thus, non-coding transcripts/genes will not be
returned by the queries using protein annotation filters.
</p>


<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="../../AnnotationFilter/html/AnnotationFilter.html">supportedFilters()</a></code> to list all filters supported for <code>EnsDb</code> objects.
</p>
<p><code><a href="ProteinFunctionality.html">listUniprotDbs()</a></code> and <code><a href="ProteinFunctionality.html">listUniprotMappingTypes()</a></code> to list all Uniprot
database names respectively mapping method types from the database.
</p>
<p><code><a href="../../AnnotationFilter/html/AnnotationFilter.html">GeneIdFilter()</a></code> in the <code>AnnotationFilter</code> package for more details on the
filter objects.
</p>
<p><code><a href="EnsDb-exonsBy.html">genes()</a></code>, <code><a href="EnsDb-exonsBy.html">transcripts()</a></code>, <code><a href="EnsDb-exonsBy.html">exons()</a></code>, <code><a href="EnsDb-class.html">listGenebiotypes()</a></code>,
<code><a href="EnsDb-class.html">listTxbiotypes()</a></code>.
</p>
<p><code><a href="global-filters.html">addFilter()</a></code> and <code><a href="global-filters.html">filter()</a></code> for globally adding filters to an <code>EnsDb</code>.
</p>


<h3>Examples</h3>

<pre>

## Create a filter that could be used to retrieve all informations for
## the respective gene.
gif &lt;- GeneIdFilter("ENSG00000012817")
gif

## Create a filter for a chromosomal end position of a gene
sef &lt;- GeneEndFilter(10000, condition = "&gt;")
sef

## For additional examples see the help page of "genes".


## Example for GRangesFilter:
## retrieve all genes overlapping the specified region
grf &lt;- GRangesFilter(GRanges("11", ranges = IRanges(114129278, 114129328),
                             strand = "+"), type = "any")
library(EnsDb.Hsapiens.v86)
edb &lt;- EnsDb.Hsapiens.v86
genes(edb, filter = grf)

## Get also all transcripts overlapping that region.
transcripts(edb, filter = grf)

## Retrieve all transcripts for the above gene
gn &lt;- genes(edb, filter = grf)
txs &lt;- transcripts(edb, filter = GeneNameFilter(gn$gene_name))
## Next we simply plot their start and end coordinates.
plot(3, 3, pch=NA, xlim=c(start(gn), end(gn)), ylim=c(0, length(txs)),
yaxt="n", ylab="")
## Highlight the GRangesFilter region
rect(xleft=start(grf), xright=end(grf), ybottom=0, ytop=length(txs),
col="red", border="red")
for(i in 1:length(txs)){
    current &lt;- txs[i]
    rect(xleft=start(current), xright=end(current), ybottom=i-0.975, ytop=i-0.125, border="grey")
    text(start(current), y=i-0.5,pos=4, cex=0.75, labels=current$tx_id)
}
## Thus, we can see that only 4 transcripts of that gene are indeed
## overlapping the region.


## No exon is overlapping that region, thus we're not getting anything
exons(edb, filter = grf)


## Example for ExonRankFilter
## Extract all exons 1 and (if present) 2 for all genes encoded on the
## Y chromosome
exons(edb, columns = c("tx_id", "exon_idx"),
      filter=list(SeqNameFilter("Y"),
                  ExonRankFilter(3, condition = "&lt;")))


## Get all transcripts for the gene SKA2
transcripts(edb, filter = GeneNameFilter("SKA2"))

## Which is the same as using a SymbolFilter
transcripts(edb, filter = SymbolFilter("SKA2"))


## Create a ProteinIdFilter:
pf &lt;- ProteinIdFilter("ENSP00000362111")
pf
## Using this filter would retrieve all database entries that are associated
## with a protein with the ID "ENSP00000362111"
if (hasProteinData(edb)) {
    res &lt;- genes(edb, filter = pf)
    res
}

## UniprotFilter:
uf &lt;- UniprotFilter("O60762")
## Get the transcripts encoding that protein:
if (hasProteinData(edb)) {
    transcripts(edb, filter = uf)
    ## The mapping Ensembl protein ID to Uniprot ID can however be 1:n:
    transcripts(edb, filter = TxIdFilter("ENST00000371588"),
        columns = c("protein_id", "uniprot_id"))
}

## ProtDomIdFilter:
pdf &lt;- ProtDomIdFilter("PF00335")
## Also here we could get all transcripts related to that protein domain
if (hasProteinData(edb)) {
    transcripts(edb, filter = pdf, columns = "protein_id")
}

</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
