<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Generating a Ensembl annotation package from Ensembl</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeEnsembldbPackage {ensembldb}"><tr><td>makeEnsembldbPackage {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Generating a Ensembl annotation package from Ensembl
</h2>

<h3>Description</h3>

<p>The functions described on this page allow to build <code>EnsDb</code>
annotation objects/databases from Ensembl annotations. The most
complete set of annotations, which include also the NCBI Entrezgene
identifiers for each gene, can be retrieved by the functions using
the Ensembl Perl API (i.e. functions <code>fetchTablesFromEnsembl</code>,
<code>makeEnsemblSQLiteFromTables</code>). Alternatively the functions
<code>ensDbFromAH</code>, <code>ensDbFromGRanges</code>, <code>ensDbFromGff</code> and
<code>ensDbFromGtf</code> can be used to build <code>EnsDb</code> objects using
GFF or GTF files from Ensembl, which can be either manually downloaded
from the Ensembl ftp server, or directly form within R using
<code>AnnotationHub</code>.
The generated SQLite database can be packaged into an R package using
the <code>makeEnsembldbPackage</code>.
</p>


<h3>Usage</h3>

<pre>

ensDbFromAH(ah, outfile, path, organism, genomeVersion, version)

ensDbFromGRanges(x, outfile, path, organism, genomeVersion,
                 version, ...)

ensDbFromGff(gff, outfile, path, organism, genomeVersion,
             version, ...)

ensDbFromGtf(gtf, outfile, path, organism, genomeVersion,
             version, ...)

fetchTablesFromEnsembl(version, ensemblapi, user="anonymous",
                       host="ensembldb.ensembl.org", pass="",
                       port=5306, species="human")

makeEnsemblSQLiteFromTables(path=".", dbname)

makeEnsembldbPackage(ensdb, version, maintainer, author,
                     destDir=".", license="Artistic-2.0")

</pre>


<h3>Arguments</h3>

<p>(in alphabetical order)
</p>
<table summary="R argblock">
<tr valign="top"><td><code>ah</code></td>
<td>

<p>For <code>ensDbFromAH</code>: an <code>AnnotationHub</code> object representing
a single resource (i.e. GTF file from Ensembl) from
<code>AnnotationHub</code>.
</p>
</td></tr>
<tr valign="top"><td><code>author</code></td>
<td>

<p>The author of the package.
</p>
</td></tr>
<tr valign="top"><td><code>dbname</code></td>
<td>

<p>The name for the database (optional). By default a name based on the
species and Ensembl version will be automatically generated (and
returned by the function).
</p>
</td></tr>
<tr valign="top"><td><code>destDir</code></td>
<td>

<p>Where the package should be saved to.
</p>
</td></tr>
<tr valign="top"><td><code>ensdb</code></td>
<td>

<p>The file name of the SQLite database generated by <code>makeEnsemblSQLiteFromTables</code>.
</p>
</td></tr>
<tr valign="top"><td><code> ensemblapi </code></td>
<td>

<p>The path to the Ensembl perl API installed locally on the
system. The Ensembl perl API version has to fit the version.
</p>
</td></tr>
<tr valign="top"><td><code>genomeVersion</code></td>
<td>

<p>For <code>ensDbFromAH</code>, <code>ensDbFromGtf</code> and <code>ensDbFromGff</code>:
the version of the genome (e.g. <code>"GRCh37"</code>). If not provided
the function will try to guess it from the file name (assuming file
name convention of Ensembl GTF files).
</p>
</td></tr>
<tr valign="top"><td><code>gff</code></td>
<td>

<p>The GFF file to import.
</p>
</td></tr>
<tr valign="top"><td><code>gtf</code></td>
<td>

<p>The GTF file name.
</p>
</td></tr>
<tr valign="top"><td><code>host</code></td>
<td>

<p>The hostname to access the Ensembl database.
</p>
</td></tr>
<tr valign="top"><td><code>license</code></td>
<td>

<p>The license of the package.
</p>
</td></tr>
<tr valign="top"><td><code>maintainer</code></td>
<td>

<p>The maintainer of the package.
</p>
</td></tr>
<tr valign="top"><td><code>organism</code></td>
<td>

<p>For <code>ensDbFromAH</code>, <code>ensDbFromGff</code> and <code>ensDbFromGtf</code>:
the organism name (e.g. <code>"Homo_sapiens"</code>). If not provided the
function will try to guess it from the file name (assuming file name
convention of Ensembl GTF files).
</p>
</td></tr>
<tr valign="top"><td><code>outfile</code></td>
<td>

<p>The desired file name of the SQLite file. If not provided the name
of the GTF file will be used.
</p>
</td></tr>
<tr valign="top"><td><code>pass</code></td>
<td>

<p>The password for the Ensembl database.
</p>
</td></tr>
<tr valign="top"><td><code>path</code></td>
<td>

<p>The directory in which the tables retrieved by
<code>fetchTablesFromEnsembl</code> or the SQLite database file generated
by <code>ensDbFromGtf</code> are stored.
</p>
</td></tr>
<tr valign="top"><td><code>port</code></td>
<td>

<p>The port to be used to connect to the Ensembl database.
</p>
</td></tr>
<tr valign="top"><td><code>species</code></td>
<td>

<p>The species for which the annotations should be retrieved.
</p>
</td></tr>
<tr valign="top"><td><code>user</code></td>
<td>

<p>The username for the Ensembl database.
</p>
</td></tr>
<tr valign="top"><td><code>version</code></td>
<td>

<p>For <code>fetchTablesFromEnsembl</code>, <code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code>: the
Ensembl version for which the annotation should be retrieved
(e.g. 75). The <code>ensDbFromGtf</code> function will try to guess the
Ensembl version from the GTF file name if not provided.
</p>
<p>For <code>makeEnsemblDbPackage</code>: the version for the package.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>For <code>ensDbFromGRanges</code>: the <code>GRanges</code> object.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Currently not used.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The <code>fetchTablesFromEnsembl</code> function internally calls the perl
script <code>get_gene_transcript_exon_tables.pl</code> to retrieve all
required information from the Ensembl database using the Ensembl perl
API.
</p>
<p>As an alternative way, a EnsDb database file can be generated by the
<code>ensDbFromGtf</code> or <code>ensDbFromGff</code> from a GTF or GFF file
downloaded from the Ensembl ftp server or using the <code>ensDbFromAH</code>
to build a database directly from corresponding resources from the
AnnotationHub. The returned database file name can then
be used as an input to the <code>makeEnsembldbPackage</code> or it can be
directly loaded and used by the <code>EnsDb</code> constructor.
</p>


<h3>Value</h3>

<p><code>makeEnsemblSQLiteFromTables</code>, <code>ensDbFromAH</code>,
<code>ensDbFromGRanges</code> and <code>ensDbFromGtf</code>: the name of the
SQLite file.
</p>


<h3>Functions</h3>


<dl>
<dt>ensDbFromAH</dt><dd>
<p>Create an <code>EnsDb</code> (SQLite) database from a GTF file provided
by <code>AnnotationHub</code>. The function returns the file name of the
generated database file. For usage see the examples below.
</p>
</dd>
<dt>ensDbFromGff</dt><dd>
<p>Create an <code>EnsDb</code> (SQLite) database from a GFF file from
Ensembl. The function returns the file name of the
generated database file. For usage see the examples below.
</p>
</dd>
<dt>ensDbFromGtf</dt><dd>
<p>Create an <code>EnsDb</code> (SQLite) database from a GTF file from
Ensembl. The function returns the file name of the generated
database file. For usage see the examplesbelow.
</p>
</dd>
<dt>ensDbFromGRanges</dt><dd>
<p>Create an <code>EnsDb</code> (SQLite) database from a GRanges object
(e.g. from <code>AnnotationHub</code>). The function returns the file
name of the generated database file. For usage see the examples
below.
</p>
</dd>
<dt>fetchTablesFromEnsembl</dt><dd>
<p>Uses the Ensembl Perl API to fetch all required data from an
Ensembl database server and stores them locally to text files
(that can be used as input for the
<code>makeEnsembldbSQLiteFromTables</code> function).
</p>
</dd>
<dt>makeEnsemblSQLiteFromTables</dt><dd>
<p>Creates the SQLite <code>EnsDb</code> database from the tables generated
by the <code>fetchTablesFromEnsembl</code>.
</p>
</dd>
<dt>makeEnsembldbPackage</dt><dd>
<p>Creates an R package containing the <code>EnsDb</code> database from a
<code>EnsDb</code> SQLite database created by any of the above
functions <code>ensDbFromAH</code>, <code>ensDbFromGff</code>,
<code>ensDbFromGtf</code> or <code>makeEnsemblSQLiteFromTables</code>.
</p>
</dd>
</dl>



<h3>Note</h3>

<p>A local installation of the Ensembl perl API is required for the
<code>fetchTablesFromEnsembl</code>. See
<a href="http://www.ensembl.org/info/docs/api/api_installation.html">http://www.ensembl.org/info/docs/api/api_installation.html</a> for
installation inscructions.
</p>
<p>A database generated from a GTF/GFF files lacks some features as they are
not available in the GTF files from Ensembl. These are: NCBI
Entrezgene IDs.
</p>


<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="EnsDb.html">EnsDb</a></code>, <code><a href="EnsDb-exonsBy.html">genes</a></code>
</p>


<h3>Examples</h3>

<pre>

## Not run: 

    ## get all human gene/transcript/exon annotations from Ensembl (75)
    ## the resulting tables will be stored by default to the current working
    ## directory; if the correct Ensembl api (version 75) is defined in the
    ## PERL5LIB environment variable, the ensemblapi parameter can also be omitted.
    fetchTablesFromEnsembl(75,
                           ensemblapi="/home/bioinfo/ensembl/75/API/ensembl/modules",
                           species="human")

    ## These tables can then be processed to generate a SQLite database
    ## containing the annotations
    DBFile &lt;- makeEnsemblSQLiteFromTables()

    ## and finally we can generate the package
    makeEnsembldbPackage(ensdb=DBFile, version="0.0.1",
                         maintainer="Johannes Rainer &lt;johannes.rainer@eurac.edu&gt;",
                         author="J Rainer")

    ## Build an annotation database form a GFF file from Ensembl.
    ## ftp://ftp.ensembl.org/pub/release-83/gff3/rattus_norvegicus
    gff &lt;- "Rattus_norvegicus.Rnor_6.0.83.gff3.gz"
    DB &lt;- ensDbFromGff(gff=gff)
    edb &lt;- EnsDb(DB)
    edb

    ## Build an annotation file from a GTF file.
    ## the GTF file can be downloaded from
    ## ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/
    gtffile &lt;- "Homo_sapiens.GRCh37.75.gtf.gz"
    ## generate the SQLite database file
    DB &lt;- ensDbFromGtf(gtf=paste0(ensemblhost, gtffile))

    ## load the DB file directly
    EDB &lt;- EnsDb(DB)

    ## Alternatively, we could fetch a GTF file directly from AnnotationHub
    ## and build the database from that:
    library(AnnotationHub)
    ah &lt;- AnnotationHub()
    ## Query for all GTF files from Ensembl for Ensembl version 81
    query(ah, c("Ensembl", "release-81", "GTF"))
    ## We could get the one from e.g. Bos taurus:
    DB &lt;- ensDbFromAH(ah["AH47941"])
    edb &lt;- EnsDb(DB)
    edb

## End(Not run)

## Generate a sqlite database for genes encoded on chromosome Y
chrY &lt;- system.file("chrY", package="ensembldb")
DBFile &lt;- makeEnsemblSQLiteFromTables(path=chrY ,dbname=tempfile())
## load this database:
edb &lt;- EnsDb(DBFile)

edb

## Generate a sqlite database from a GRanges object specifying
## genes encoded on chromosome Y
load(system.file("YGRanges.RData", package="ensembldb"))

Y

DB &lt;- ensDbFromGRanges(Y, path=tempdir(), version=75,
                       organism="Homo_sapiens")
edb &lt;- EnsDb(DB)


</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
