<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Map transcript-relative coordinates to genomic coordinates</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for transcriptToGenome {ensembldb}"><tr><td>transcriptToGenome {ensembldb}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Map transcript-relative coordinates to genomic coordinates</h2>

<h3>Description</h3>

<p><code>transcriptToGenome</code> maps transcript-relative coordinates to genomic
coordinates. Provided coordinates are expected to be relative to the first
nucleotide of the <strong>transcript</strong>, not the <strong>CDS</strong>. CDS-relative coordinates
have to be converted to transcript-relative positions first with the
<code><a href="cdsToTranscript.html">cdsToTranscript()</a></code> function.
</p>


<h3>Usage</h3>

<pre>
transcriptToGenome(x, db, id = "name")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p><code>IRanges</code> with the coordinates within the transcript. Coordinates
are counted from the start of the transcript (including the 5' UTR). The
Ensembl IDs of the corresponding transcripts have to be provided either
as <code>names</code> of the <code>IRanges</code>, or in one of its metadata columns.</p>
</td></tr>
<tr valign="top"><td><code>db</code></td>
<td>
<p><code>EnsDb</code> object.</p>
</td></tr>
<tr valign="top"><td><code>id</code></td>
<td>
<p><code>character(1)</code> specifying where the transcript identifier can be
found. Has to be either <code>"name"</code> or one of <code>colnames(mcols(prng))</code>.</p>
</td></tr>
</table>


<h3>Value</h3>

<p><code>GRangesList</code> with the same length (and order) than the input <code>IRanges</code>
<code>x</code>. Each <code>GRanges</code> in the <code>GRangesList</code> provides the genomic coordinates
corresponding to the provided within-transcript coordinates. The
original transcript ID and the transcript-relative coordinates are provided
as metadata columns as well as the ID of the individual exon(s). An empty
<code>GRanges</code> is returned for transcripts that can not be found in the database.
</p>


<h3>Author(s)</h3>

<p>Johannes Rainer
</p>


<h3>See Also</h3>

<p><code><a href="cdsToTranscript.html">cdsToTranscript()</a></code> and <code><a href="transcriptToCds.html">transcriptToCds()</a></code> for the mapping between
CDS- and transcript-relative coordinates.
</p>
<p>Other coordinate mapping functions: 
<code><a href="cdsToTranscript.html">cdsToTranscript</a>()</code>,
<code><a href="genomeToProtein.html">genomeToProtein</a>()</code>,
<code><a href="genomeToTranscript.html">genomeToTranscript</a>()</code>,
<code><a href="proteinToGenome.html">proteinToGenome</a>()</code>,
<code><a href="proteinToTranscript.html">proteinToTranscript</a>()</code>,
<code><a href="transcriptToCds.html">transcriptToCds</a>()</code>,
<code><a href="transcriptToProtein.html">transcriptToProtein</a>()</code>
</p>


<h3>Examples</h3>

<pre>

library(EnsDb.Hsapiens.v86)
## Restrict all further queries to chromosome x to speed up the examples
edbx &lt;- filter(EnsDb.Hsapiens.v86, filter = ~ seq_name == "X")

## Below we map positions 1 to 5 within the transcript ENST00000381578 to
## the genome. The ID of the transcript has to be provided either as names
## or in one of the IRanges' metadata columns
txpos &lt;- IRanges(start = 1, end = 5, names = "ENST00000381578")

transcriptToGenome(txpos, edbx)
## The object returns a GRangesList with the genomic coordinates, in this
## example the coordinates are within the same exon and map to a single
## genomic region.

## Next we map nucleotides 501 to 505 of ENST00000486554 to the genome
txpos &lt;- IRanges(start = 501, end = 505, names = "ENST00000486554")

transcriptToGenome(txpos, edbx)
## The positions within the transcript are located within two of the
## transcripts exons and thus a `GRanges` of length 2 is returned.

## Next we map multiple regions, two within the same transcript and one
## in a transcript that does not exist.
txpos &lt;- IRanges(start = c(501, 1, 5), end = c(505, 10, 6),
    names = c("ENST00000486554", "ENST00000486554", "some"))

res &lt;- transcriptToGenome(txpos, edbx)

## The length of the result GRangesList has the same length than the
## input IRanges
length(res)

## The result for the last region is an empty GRanges, because the
## transcript could not be found in the database
res[[3]]

res
</pre>

<hr /><div style="text-align: center;">[Package <em>ensembldb</em> version 2.14.1 <a href="00Index.html">Index</a>]</div>
</body></html>
