<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Views into a set of BAM files</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for BamViews {Rsamtools}"><tr><td>BamViews {Rsamtools}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Views into a set of BAM files</h2>

<h3>Description</h3>

<p>Use <code>BamViews()</code> to reference a set of disk-based BAM files to be
processed (e.g., queried using <code><a href="scanBam.html">scanBam</a></code>) as a single
&lsquo;experiment&rsquo;.
</p>


<h3>Usage</h3>

<pre>

## Constructor
BamViews(bamPaths=character(0),
     bamIndicies=bamPaths,
     bamSamples=DataFrame(row.names=make.unique(basename(bamPaths))),
     bamRanges, bamExperiment = list(), ...)
## S4 method for signature 'missing'
BamViews(bamPaths=character(0),
     bamIndicies=bamPaths,
     bamSamples=DataFrame(row.names=make.unique(basename(bamPaths))),
     bamRanges, bamExperiment = list(), ..., auto.range=FALSE)
## Accessors
bamPaths(x)
bamSamples(x)
bamSamples(x) &lt;- value
bamRanges(x)
bamRanges(x) &lt;- value
bamExperiment(x)

## S4 method for signature 'BamViews'
names(x)
## S4 replacement method for signature 'BamViews'
names(x) &lt;- value
## S4 method for signature 'BamViews'
dimnames(x)
## S4 replacement method for signature 'BamViews,ANY'
dimnames(x) &lt;- value

bamDirname(x, ...) &lt;- value

## Subset
## S4 method for signature 'BamViews,ANY,ANY'
x[i, j, ..., drop=TRUE]
## S4 method for signature 'BamViews,ANY,missing'
x[i, j, ..., drop=TRUE]
## S4 method for signature 'BamViews,missing,ANY'
x[i, j, ..., drop=TRUE]

## Input
## S4 method for signature 'BamViews'
scanBam(file, index = file, ..., param = ScanBamParam(what=scanBamWhat()))
## S4 method for signature 'BamViews'
countBam(file, index = file, ..., param = ScanBamParam())

## Show
## S4 method for signature 'BamViews'
show(object)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>bamPaths</code></td>
<td>
<p>A character() vector of BAM path names.</p>
</td></tr>
<tr valign="top"><td><code>bamIndicies</code></td>
<td>
<p>A character() vector of BAM index file path names,
<em>without</em> the &lsquo;.bai&rsquo; extension.</p>
</td></tr>
<tr valign="top"><td><code>bamSamples</code></td>
<td>
<p>A <code><a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a></code> instance with as
many rows as <code>length(bamPaths)</code>, containing sample information
associated with each path.</p>
</td></tr>
<tr valign="top"><td><code>bamRanges</code></td>
<td>
<p>Missing or a <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> instance with
ranges defined on the reference chromosomes of the BAM files. Ranges
are <em>not</em> validated against the BAM files.</p>
</td></tr>
<tr valign="top"><td><code>bamExperiment</code></td>
<td>
<p>A list() containing additional information about
the experiment.</p>
</td></tr>
<tr valign="top"><td><code>auto.range</code></td>
<td>
<p>If <code>TRUE</code> and all <code>bamPaths</code> exist,
populate the ranges with the union of ranges returned in the
<code>target</code> element of <code>scanBamHeader</code>.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments.</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>
<p>An instance of <code>BamViews</code>.</p>
</td></tr>
<tr valign="top"><td><code>object</code></td>
<td>
<p>An instance of <code>BamViews</code>.</p>
</td></tr>
<tr valign="top"><td><code>value</code></td>
<td>
<p>An object of appropriate type to replace content.</p>
</td></tr>
<tr valign="top"><td><code>i</code></td>
<td>
<p>During subsetting, a logical or numeric index into
<code>bamRanges</code>.</p>
</td></tr>
<tr valign="top"><td><code>j</code></td>
<td>
<p>During subsetting, a logical or numeric index into
<code>bamSamples</code> and <code>bamPaths</code>.</p>
</td></tr>
<tr valign="top"><td><code>drop</code></td>
<td>
<p>A logical(1), <em>ignored</em> by all <code>BamViews</code>
subsetting methods.</p>
</td></tr>
</table>

<table summary="R argblock">
<tr valign="top"><td><code>file</code></td>
<td>
<p>An instance of <code>BamViews</code>.</p>
</td></tr>
<tr valign="top"><td><code>index</code></td>
<td>
<p>A character vector of indices, corresponding to the
<code>bamPaths(file)</code>. </p>
</td></tr>
<tr valign="top"><td><code>param</code></td>
<td>
<p>An optional <code><a href="ScanBamParam-class.html">ScanBamParam</a></code> instance to
further influence scanning or counting.</p>
</td></tr>
</table>


<h3>Objects from the Class</h3>

<p>Objects are created by calls of the form <code>BamViews()</code>.
</p>


<h3>Slots</h3>


<dl>
<dt>bamPaths</dt><dd><p>A character() vector of BAM path names.</p>
</dd>
<dt>bamIndicies</dt><dd><p>A character() vector of BAM index path names.</p>
</dd>
<dt>bamSamples</dt><dd><p>A <code><a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a></code> instance with as
many rows as <code>length(bamPaths)</code>, containing sample information
associated with each path.</p>
</dd>
<dt>bamRanges</dt><dd><p>A <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> instance with
ranges defined on the spaces of the BAM files. Ranges are <em>not</em>
validated against the BAM files.</p>
</dd>
<dt>bamExperiment</dt><dd><p>A list() containing additional information about
the experiment.</p>
</dd>
</dl>



<h3>Functions and methods</h3>

<p>See 'Usage' for details on invocation.
</p>
<p>Constructor:
</p>

<dl>
<dt>BamViews:</dt><dd><p>Returns a <code>BamViews</code> object.</p>
</dd>
</dl>

<p>Accessors:
</p>

<dl>
<dt>bamPaths</dt><dd><p>Returns a character() vector of BAM path names.</p>
</dd>
<dt>bamIndicies</dt><dd><p>Returns a character() vector of BAM index path
names.</p>
</dd>
<dt>bamSamples</dt><dd><p>Returns a <code><a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a></code> instance
with as many rows as <code>length(bamPaths)</code>, containing sample
information associated with each path.</p>
</dd>
<dt>bamSamples&lt;-</dt><dd><p>Assign a <code><a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a></code> instance
with as many rows as <code>length(bamPaths)</code>, containing sample
information associated with each path.</p>
</dd>
<dt>bamRanges</dt><dd><p>Returns a <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> instance
with ranges defined on the spaces of the BAM files. Ranges are
<em>not</em> validated against the BAM files.</p>
</dd>
<dt>bamRanges&lt;-</dt><dd><p>Assign a <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> instance
with ranges defined on the spaces of the BAM files. Ranges are
<em>not</em> validated against the BAM files.</p>
</dd>
<dt>bamExperiment</dt><dd><p>Returns a list() containing additional
information about the experiment.</p>
</dd>
<dt>names</dt><dd><p>Return the column names of the <code>BamViews</code>
instance; same as <code>names(bamSamples(x))</code>.</p>
</dd>
<dt>names&lt;-</dt><dd><p>Assign the column names of the <code>BamViews</code>
instance.</p>
</dd>
<dt>dimnames</dt><dd><p>Return the row and column names of the
<code>BamViews</code> instance.</p>
</dd>
<dt>dimnames&lt;-</dt><dd><p>Assign the row and column names of the
<code>BamViews</code> instance.</p>
</dd>
</dl>

<p>Methods:
</p>

<dl>
<dt>&quot;[&quot;</dt><dd><p>Subset the object by <code>bamRanges</code> or <code>bamSamples</code>.</p>
</dd>
<dt>scanBam</dt><dd><p>Visit each path in <code>bamPaths(file)</code>, returning
the result of <code>scanBam</code> applied to the specified
path. <code>bamRanges(file)</code> takes precedence over
<code>bamWhich(param)</code>.</p>
</dd>
<dt>countBam</dt><dd><p>Visit each path in <code>bamPaths(file)</code>, returning
the result of <code>countBam</code> applied to the specified
path. <code>bamRanges(file)</code> takes precedence over
<code>bamWhich(param)</code>.</p>
</dd>
<dt>show</dt><dd><p>Compactly display the object.</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Martin Morgan</p>


<h3>Examples</h3>

<pre>
fls &lt;- system.file("extdata", "ex1.bam", package="Rsamtools",
                   mustWork=TRUE)
rngs &lt;- GRanges(seqnames = Rle(c("chr1", "chr2"), c(9, 9)),
                ranges = c(IRanges(seq(10000, 90000, 10000), width=500),
                           IRanges(seq(100000, 900000, 100000), width=5000)),
                Count = seq_len(18L))
v &lt;- BamViews(fls, bamRanges=rngs)
v
v[1:5,]
bamRanges(v[c(1:5, 11:15),])
bamDirname(v) &lt;- getwd()
v
</pre>

<hr /><div style="text-align: center;">[Package <em>Rsamtools</em> version 2.6.0 <a href="00Index.html">Index</a>]</div>
</body></html>
