<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Manipulate indexed fasta files.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for FaFile {Rsamtools}"><tr><td>FaFile {Rsamtools}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Manipulate indexed fasta files.</h2>

<h3>Description</h3>

<p>Use <code>FaFile()</code> to create a reference to an indexed fasta
file. The reference remains open across calls to methods, avoiding
costly index re-loading.
</p>
<p><code>FaFileList()</code> provides a convenient way of managing a list of
<code>FaFile</code> instances.
</p>


<h3>Usage</h3>

<pre>

## Constructors

FaFile(file, index=sprintf("%s.fai", file),
             gzindex=sprintf("%s.gzi", file))
FaFileList(...)

## Opening / closing

## S3 method for class 'FaFile'
open(con, ...)
## S3 method for class 'FaFile'
close(con, ...)

## accessors; also path(), index()

## S4 method for signature 'FaFile'
gzindex(object, asNA=TRUE)
## S4 method for signature 'FaFileList'
gzindex(object, asNA=TRUE)
## S4 method for signature 'FaFile'
isOpen(con, rw="")

## actions

## S4 method for signature 'FaFile'
indexFa(file, ...)

## S4 method for signature 'FaFile'
scanFaIndex(file, ...)
## S4 method for signature 'FaFileList'
scanFaIndex(file, ..., as=c("GRangesList", "GRanges"))

## S4 method for signature 'FaFile'
seqinfo(x)

## S4 method for signature 'FaFile'
countFa(file, ...)

## S4 method for signature 'FaFile,GRanges'
scanFa(file, param, ...,
    as=c("DNAStringSet", "RNAStringSet", "AAStringSet"))
## S4 method for signature 'FaFile,IntegerRangesList'
scanFa(file, param, ...,
    as=c("DNAStringSet", "RNAStringSet", "AAStringSet"))
## S4 method for signature 'FaFile,missing'
scanFa(file, param, ...,
    as=c("DNAStringSet", "RNAStringSet", "AAStringSet"))

## S4 method for signature 'FaFile'
getSeq(x, param, ...)
## S4 method for signature 'FaFileList'
getSeq(x, param, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>con, object, x</code></td>
<td>
<p>An instance of <code>FaFile</code> or (for <code>gzindex</code>
and <code>getSeq</code>) <code>FaFileList</code>.</p>
</td></tr>
<tr valign="top"><td><code>file, index, gzindex</code></td>
<td>
<p>A character(1) vector of the fasta or fasta index
or bgzip index file path (for <code>FaFile</code>), or an instance of class
<code>FaFile</code> or <code>FaFileList</code> (for <code>scanFaIndex</code>, <code>getSeq</code>).</p>
</td></tr>
<tr valign="top"><td><code>asNA</code></td>
<td>
<p>logical indicating if missing output should be NA or
character()</p>
</td></tr>
<tr valign="top"><td><code>param</code></td>
<td>
<p>An optional <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> or
<code><a href="../../IRanges/html/IntegerRangesList-class.html">IntegerRangesList</a></code> instance to select reads (and
sub-sequences) for input. See Methods, below.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments. </p>

<ul>
<li><p> For <code>FaFileList</code>, this can either be a single character
vector of paths to BAM files, or several instances of
<code>FaFile</code> objects.
</p>
</li>
<li><p> For <code>scanFa,FaFile,missing-method</code> this can include
arguemnts to <code>readDNAStringSet</code> / <code>readRNAStringSet</code> /
<code>readAAStringSet</code> when <code>param</code> is &lsquo;missing&rsquo;.
</p>
</li></ul>

</td></tr>
<tr valign="top"><td><code>rw</code></td>
<td>
<p>Mode of file; ignored.</p>
</td></tr>
<tr valign="top"><td><code>as</code></td>
<td>
<p>A character(1) vector indicating the type of object to
return. </p>

<ul>
<li><p> For <code>scanFaIndex</code>, default <code>GRangesList</code>, with
index information from each file is returned as an element of the
list.
</p>
</li>
<li><p> For <code>scanFa</code>, default <code>DNAStringSet</code>.
</p>
</li></ul>

<p><code>GRangesList</code>, index information is collapsed across files into
the unique index elements.</p>
</td></tr>
</table>


<h3>Objects from the Class</h3>

<p>Objects are created by calls of the form <code>FaFile()</code>.
</p>


<h3>Fields</h3>

<p>The <code>FaFile</code> class inherits fields from the
<code><a href="RsamtoolsFile-class.html">RsamtoolsFile</a></code> class.
</p>


<h3>Functions and methods</h3>

<p><code>FaFileList</code> inherits methods from
<code><a href="RsamtoolsFileList-class.html">RsamtoolsFileList</a></code> and <code><a href="../../S4Vectors/html/SimpleList-class.html">SimpleList</a></code>.
</p>
<p>Opening / closing:
</p>

<dl>
<dt>open.FaFile</dt><dd><p>Opens the (local or remote) <code>path</code> and
<code>index</code> files.  Returns a <code>FaFile</code> instance.</p>
</dd>
<dt>close.FaFile</dt><dd><p>Closes the <code>FaFile</code> <code>con</code>; returning
(invisibly) the updated <code>FaFile</code>. The instance may be
re-opened with <code>open.FaFile</code>.</p>
</dd>
</dl>

<p>Accessors: 
</p>

<dl>
<dt>path</dt><dd><p>Returns a character(1) vector of the fasta path name.</p>
</dd>
<dt>index</dt><dd><p>Returns a character(1) vector of fasta index name
(minus the '.fai' extension).</p>
</dd>
</dl>

<p>Methods:
</p>

<dl>
<dt>indexFa</dt><dd><p>Visit the path in <code>path(file)</code> and create an
index file (with the extension &lsquo;.fai&rsquo;).</p>
</dd>
<dt>scanFaIndex</dt><dd><p>Read the sequence names and and widths of
recorded in an indexed fasta file, returning the information as a
<code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code> object.</p>
</dd>
<dt>seqinfo</dt><dd><p>Consult the index file for defined sequences
(<code>seqlevels()</code>) and lengths (<code>seqlengths()</code>).</p>
</dd>
<dt>countFa</dt><dd><p>Return the number of records in the fasta file.</p>
</dd>
<dt>scanFa</dt><dd><p>Return the sequences indicated by <code>param</code> as a
<code><a href="../../Biostrings/html/XStringSet-class.html">DNAStringSet</a></code> instance. <code>seqnames(param)</code>
selects the sequences to return; <code>start(param)</code> and
<code>end{param}</code> define the (1-based) region of the sequence to
return. Values of <code>end(param)</code> greater than the width of the
sequence cause an error; use <code>seqlengths(FaFile(file))</code> to
discover sequence widths. When <code>param</code> is missing, all
records are selected. When <code>length(param)==0</code> no records are
selected.</p>
</dd>
<dt>getSeq</dt><dd><p>Returns the sequences indicated by <code>param</code> from
the indexed fasta file(s) of <code>file</code>.
</p>
<p>For the <code>FaFile</code> method, the return type is a
<code>DNAStringSet</code>. The <code>getSeq,FaFile</code> and
<code>scanFa,FaFile,GRanges</code> methods differ in that <code>getSeq</code>
will reverse complement sequences selected from the minus strand.
</p>
<p>For the <code>FaFileList</code> method, the <code>param</code> argument must
be a <code>GRangesList</code> of the same length as <code>file</code>,
creating a one-to-one mapping between the ith element of
<code>file</code> and the ith element of <code>param</code>; the return type
is a <code>SimpleList</code> of <code>DNAStringSet</code> instances, with
elements of the list in the same order as the input elements.</p>
</dd>
<dt>show</dt><dd><p>Compactly display the object.</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Martin Morgan</p>


<h3>Examples</h3>

<pre>

fl &lt;- system.file("extdata", "ce2dict1.fa", package="Rsamtools",
                  mustWork=TRUE)
fa &lt;- open(FaFile(fl))                   # open
countFa(fa)
(idx &lt;- scanFaIndex(fa))
(dna &lt;- scanFa(fa, param=idx[1:2]))
ranges(idx) &lt;- narrow(ranges(idx), -10)  # last 10 nucleotides
(dna &lt;- scanFa(fa, param=idx[1:2]))

</pre>

<hr /><div style="text-align: center;">[Package <em>Rsamtools</em> version 2.6.0 <a href="00Index.html">Index</a>]</div>
</body></html>
