<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Import samtools 'pileup' files.</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for readPileup {Rsamtools}"><tr><td>readPileup {Rsamtools}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Import samtools 'pileup' files.
</h2>

<h3>Description</h3>

<p>Import files created by evaluation of samtools' <code>pileup -cv</code>
command.
</p>


<h3>Usage</h3>

<pre>
readPileup(file, ...)
## S4 method for signature 'connection'
readPileup(file, ..., variant=c("SNP", "indel", "all"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>file</code></td>
<td>
<p>The file name, or
<code><a href="../../base/html/connections.html">connection</a></code>, of the pileup output file
to be parsed.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments, passed to methods. For instance,
specify <code>variant</code> for the readPileup,character-method.</p>
</td></tr>
<tr valign="top"><td><code>variant</code></td>
<td>
<p>Type of variant to parse; select one.</p>
</td></tr>
</table>


<h3>Value</h3>

<p><code>readPileup</code> returns a <code><a href="../../GenomicRanges/html/GRanges.html">GRanges</a></code> object.
</p>
<p>The value returned by <code>variant="SNP"</code> or <code>variant="all"</code>
contains:
</p>

<dl>
<dt>space:</dt><dd><p>The chromosome names (fastq ids) of the reference
sequence</p>
</dd>
<dt>position:</dt><dd><p>The nucleotide position (base 1) of the variant.</p>
</dd>
<dt>referenceBase:</dt><dd><p>The nucleotide in the reference sequence.</p>
</dd>
<dt>consensusBase;</dt><dd><p>The consensus nucleotide, as determined by
samtools pileup.</p>
</dd>
<dt>consensusQuality:</dt><dd><p>The phred-scaled consensus quality.</p>
</dd>
<dt>snpQuality:</dt><dd><p>The phred-scaled SNP quality (probability of the
consensus being identical to the reference).</p>
</dd>
<dt>maxMappingQuality:</dt><dd><p>The root mean square mapping quality of reads
overlapping the site.</p>
</dd>
<dt>coverage:</dt><dd><p>The number of reads covering the site.</p>
</dd>
</dl>

<p>The value returned by <code>variant="indel"</code> contains space, position,
reference, consensus, consensusQuality, snpQuality, maxMappingQuality,
and coverage fields, and:
</p>

<dl>
<dt>alleleOne, alleleTwo</dt><dd><p>The first (typically, in the reference
sequence) and second allelic variants.</p>
</dd>
<dt>alleleOneSupport, alleleTwoSupport</dt><dd><p>The number of reads
supporting each allele.</p>
</dd>
<dt>additionalIndels</dt><dd><p>The number of additional indels present.</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Sean Davis
</p>


<h3>References</h3>

<p><a href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a>
</p>


<h3>Examples</h3>

<pre>

fl &lt;- system.file("extdata", "pileup.txt", package="Rsamtools",
                  mustWork=TRUE)
(res &lt;- readPileup(fl))
xtabs(~referenceBase + consensusBase, mcols(res))[DNA_BASES,]

## Not run: ## uses a pipe, and arguments passed to read.table
## three successive piles of 100 records each
cmd &lt;- "samtools pileup -cvf human_b36_female.fa.gz na19240_3M.bam"
p &lt;- pipe(cmd, "r")
snp &lt;- readPileup(p, nrow=100)  # variant="SNP"
indel &lt;- readPileup(p, nrow=100, variant="indel")
all &lt;- readPileup(p, nrow=100, variant="all")

## End(Not run)

</pre>

<hr /><div style="text-align: center;">[Package <em>Rsamtools</em> version 2.6.0 <a href="00Index.html">Index</a>]</div>
</body></html>
