<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Parameters for scanning BAM files</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for ScanBamParam {Rsamtools}"><tr><td>ScanBamParam {Rsamtools}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Parameters for scanning BAM files</h2>

<h3>Description</h3>

<p>Use <code>ScanBamParam()</code> to create a parameter object influencing
what fields and which records are imported from a (binary) BAM file.
Use of <code>which</code> requires that a BAM index file
(<code>&lt;filename&gt;.bai</code>) exists.
</p>


<h3>Usage</h3>

<pre>

# Constructor
ScanBamParam(flag = scanBamFlag(), simpleCigar = FALSE,
    reverseComplement = FALSE, tag = character(0), tagFilter = list(),
    what = character(0), which, mapqFilter=NA_integer_)

# Constructor helpers
scanBamFlag(isPaired = NA, isProperPair = NA, isUnmappedQuery = NA, 
    hasUnmappedMate = NA, isMinusStrand = NA, isMateMinusStrand = NA,
    isFirstMateRead = NA, isSecondMateRead = NA, isNotPrimaryRead = NA,
    isSecondaryAlignment = NA, isNotPassingQualityControls = NA,
    isDuplicate = NA, isSupplementaryAlignment = NA)

scanBamWhat()

# Accessors
bamFlag(object, asInteger=FALSE)
bamFlag(object) &lt;- value
bamReverseComplement(object)
bamReverseComplement(object) &lt;- value
bamSimpleCigar(object)
bamSimpleCigar(object) &lt;- value
bamTag(object)
bamTag(object) &lt;- value
bamTagFilter(object)
bamTagFilter(object) &lt;- value
bamWhat(object)
bamWhat(object) &lt;- value
bamWhich(object)
bamWhich(object) &lt;- value
bamMapqFilter(object)
bamMapqFilter(object) &lt;- value

## S4 method for signature 'ScanBamParam'
show(object)

# Flag utils
bamFlagAsBitMatrix(flag, bitnames=FLAG_BITNAMES)
bamFlagAND(flag1, flag2)
bamFlagTest(flag, value)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>flag</code></td>
<td>
<p>For <code>ScanBamParam</code>, an integer(2) vector used to filter
reads based on their 'flag' entry. This is most easily created with the
<code>scanBamFlag()</code> helper function.
</p>
<p>For <code>bamFlagAsBitMatrix</code>, <code>bamFlagTest</code> an integer vector
where each element represents a 'flag' entry.  </p>
</td></tr>
<tr valign="top"><td><code>simpleCigar</code></td>
<td>
<p>A logical(1) vector which, when TRUE, returns only
those reads for which the cigar (run-length encoded representation
of the alignment) is missing or contains only matches / mismatches
(<code>'M'</code>).</p>
</td></tr>
<tr valign="top"><td><code>reverseComplement</code></td>
<td>
<p>A logical(1) vectors. BAM files store reads
mapping to the minus strand as though they are on the plus
strand. Rsamtools obeys this convention by default
(<code>reverseComplement=FALSE</code>), but when this value is set to TRUE
returns the sequence and quality scores of reads mapped to the minus
strand in the reverse complement (sequence) and reverse (quality) of
the read as stored in the BAM file. This might be useful if wishing
to recover read and quality scores as represented in fastq files,
but is NOT appropriate for variant calling or other alignment-based
operations.</p>
</td></tr>
<tr valign="top"><td><code>tag</code></td>
<td>
<p>A character vector naming tags to be extracted. A tag is an
optional field, with arbitrary information, stored with each
record. Tags are identified by two-letter codes, so all elements of
<code>tag</code> must have exactly 2 characters.</p>
</td></tr>
<tr valign="top"><td><code>tagFilter</code></td>
<td>
<p>A named <code>list</code> of atomic vectors. The name of
each <code>list</code> element is the tag name (two-letter code), and the
corresponding atomic vector is the set of acceptable values for the
tag. Only reads with specified tags are included. <code>NULL</code>s,
<code>NA</code>s, and empty strings are not allowed in the atomic
vectors.</p>
</td></tr>
<tr valign="top"><td><code>what</code></td>
<td>
<p>A character vector naming the fields to return
<code>scanBamWhat()</code> returns a vector of available fields.
Fields are described on the <code><a href="scanBam.html">scanBam</a></code> help page.</p>
</td></tr>
<tr valign="top"><td><code>mapqFilter</code></td>
<td>
<p>A non-negative integer(1) specifying the minimum
mapping quality to include. BAM records with mapping qualities less
than <code>mapqFilter</code> are discarded.</p>
</td></tr>
<tr valign="top"><td><code>which</code></td>
<td>
<p>A <code><a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a></code>,
<code><a href="../../IRanges/html/IntegerRangesList-class.html">IntegerRangesList</a></code>, or any object that can be
coerced to a <code>IntegerRangesList</code>, or missing object, from which
a <code>IRangesList</code> instance will be constructed. Names of the
<code>IRangesList</code> correspond to reference sequences, and ranges to
the regions on that reference sequence for which matches are
desired. Because data types are coerced to <code>IRangesList</code>,
<code>which</code> does <em>not</em> include strand information (use the
<code>flag</code> argument instead). Only records with a read overlapping
the specified ranges are returned. All ranges must have ends less
than or equal to 536870912. When one record overlaps two ranges in
<code>which</code>, the record is returned <em>twice</em>.</p>
</td></tr>
<tr valign="top"><td><code>isPaired</code></td>
<td>
<p>A logical(1) indicating whether unpaired (FALSE),
paired (TRUE), or any (NA) read should be returned.</p>
</td></tr>
<tr valign="top"><td><code>isProperPair</code></td>
<td>
<p>A logical(1) indicating whether improperly paired
(FALSE), properly paired (TRUE), or any (NA) read should be
returned. A properly paired read is defined by the alignment
algorithm and might, e.g., represent reads aligning to identical
reference sequences and with a specified distance.</p>
</td></tr>
<tr valign="top"><td><code>isUnmappedQuery</code></td>
<td>
<p>A logical(1) indicating whether unmapped
(TRUE), mapped (FALSE), or any (NA) read should be returned.</p>
</td></tr>
<tr valign="top"><td><code>hasUnmappedMate</code></td>
<td>
<p>A logical(1) indicating whether reads with
mapped (FALSE), unmapped (TRUE), or any (NA) mate should be
returned.</p>
</td></tr>
<tr valign="top"><td><code>isMinusStrand</code></td>
<td>
<p>A logical(1) indicating whether reads aligned to
the plus (FALSE), minus (TRUE), or any (NA) strand should be
returned.</p>
</td></tr>
<tr valign="top"><td><code>isMateMinusStrand</code></td>
<td>
<p>A logical(1) indicating whether mate reads
aligned to the plus (FALSE), minus (TRUE), or any (NA) strand should
be returned.</p>
</td></tr>
<tr valign="top"><td><code>isFirstMateRead</code></td>
<td>
<p>A logical(1) indicating whether the first mate
read should be returned (TRUE) or not (FALSE), or whether mate read
number should be ignored (NA).</p>
</td></tr>
<tr valign="top"><td><code>isSecondMateRead</code></td>
<td>
<p>A logical(1) indicating whether the second
mate read should be returned (TRUE) or not (FALSE), or whether mate
read number should be ignored (NA).</p>
</td></tr>
<tr valign="top"><td><code>isNotPrimaryRead</code></td>
<td>
<p>Deprecated; use <code>isSecondaryAlignment</code>.</p>
</td></tr>
<tr valign="top"><td><code>isSecondaryAlignment</code></td>
<td>
<p>A logical(1) indicating whether alignments
that are secondary (TRUE), are not (FALSE) or whose secondary status
does not matter (NA) should be returned. A non-primary alignment
(&ldquo;secondary alignment&rdquo; in the SAM specification) might result
when a read aligns to multiple locations. One alignment is
designated as primary and has this flag set to FALSE; the remainder,
for which this flag is TRUE, are designated by the aligner as
secondary.</p>
</td></tr>
<tr valign="top"><td><code>isNotPassingQualityControls</code></td>
<td>
<p>A logical(1) indicating whether
reads passing quality controls (FALSE), reads not passing quality
controls (TRUE), or any (NA) read should be returned.</p>
</td></tr>
<tr valign="top"><td><code>isDuplicate</code></td>
<td>
<p>A logical(1) indicating that un-duplicated (FALSE),
duplicated (TRUE), or any (NA) reads should be
returned. 'Duplicated' reads may represent PCR or optical
duplicates.</p>
</td></tr>
<tr valign="top"><td><code>isSupplementaryAlignment</code></td>
<td>
<p>A logical(1) indicating that
non-supplementary (FALSE), supplementary (TRUE), or any (NA) reads
should be returned. The SAM specification indicates that
'supplementary' reads are part of a chimeric alignment.</p>
</td></tr>
<tr valign="top"><td><code>object</code></td>
<td>
<p>An instance of class <code>ScanBamParam</code>.</p>
</td></tr>
<tr valign="top"><td><code>value</code></td>
<td>
<p>An instance of the corresponding slot, to be assigned to
<code>object</code> or, for <code>bamFlagTest</code>, a <code>character(1)</code> name
of the flag to test, e.g., &ldquo;isUnmappedQuery&rdquo;, from the
arguments to <code>scanBamFlag</code>.</p>
</td></tr>
<tr valign="top"><td><code>asInteger</code></td>
<td>
<p>logical(1) indicating whether &lsquo;flag&rsquo; should be
returned as an encoded integer vector (<code>TRUE</code>) or
human-readable form (<code>FALSE</code>).</p>
</td></tr>
<tr valign="top"><td><code>bitnames</code></td>
<td>
<p>Names of the flag bits to extract. Will be the colnames
of the returned matrix.</p>
</td></tr>
<tr valign="top"><td><code>flag1, flag2</code></td>
<td>
<p>Integer vectors containing &lsquo;flag&rsquo; entries.</p>
</td></tr>
</table>


<h3>Objects from the Class</h3>

<p>Objects are created by calls of the form <code>ScanBamParam()</code>.
</p>


<h3>Slots</h3>


<dl>
<dt><code>flag</code></dt><dd><p>Object of class <code>integer</code> encoding flags
to be kept when they have their '0' (<code>keep0</code>) or '1'
(<code>keep1</code>) bit set.</p>
</dd>
<dt><code>simpleCigar</code></dt><dd><p>Object of class <code>logical</code>
indicating, when TRUE, that only 'simple' cigars (empty or 'M') are
returned.</p>
</dd>
<dt><code>reverseComplement</code></dt><dd><p>Object of class <code>logical</code>
indicating, when TRUE, that reads on the minus strand are to be
reverse complemented (sequence) and reversed (quality).</p>
</dd>
<dt><code>tag</code></dt><dd><p>Object of class <code>character</code> indicating what
tags are to be returned.</p>
</dd>
<dt><code>tagFilter</code></dt><dd><p>Object of class <code>list</code> (named)
indicating tags to filter by, and the set of acceptable values for
each tag.</p>
</dd>
<dt><code>what</code></dt><dd><p>Object of class <code>character</code> indicating
what fields are to be returned.</p>
</dd>
<dt><code>which</code></dt><dd><p>Object of class <code>IntegerRangesList</code> indicating
which reference sequence and coordinate reads must overlap.</p>
</dd>
<dt><code>mapqFilter</code></dt><dd><p>Object of class <code>integer</code> indicating
the minimum mapping quality required for input, or NA to indicate
no filtering.</p>
</dd>
</dl>



<h3>Functions and methods</h3>

<p>See 'Usage' for details on invocation.
</p>
<p>Constructor:
</p>

<dl>
<dt>ScanBamParam:</dt><dd><p>Returns a <code>ScanBamParam</code> object. The
<code>which</code> argument to the constructor can be one of several
different types, as documented above.</p>
</dd>
</dl>

<p>Accessors:
</p>

<dl>
<dt>bamTag, bamTag&lt;-</dt><dd><p>Returns or sets a <code>character</code> vector of
tags to be extracted.</p>
</dd>
<dt>bamTagFilter, bamTagFilter&lt;-</dt><dd><p>Returns or sets a named
<code>list</code> of tags to filter by, and the set of their acceptable
values.</p>
</dd>
<dt>bamWhat, bamWhat&lt;-</dt><dd><p>Returns or sets a <code>character</code> vector
of fields to be extracted.</p>
</dd>
<dt>bamWhich, bamWhich&lt;-</dt><dd><p>Returns or sets a <code>IntegerRangesList</code> of
bounds on reads to be extracted. A length 0 <code>IntegerRangesList</code>
represents all reads.</p>
</dd>
<dt>bamFlag, bamFlag&lt;-</dt><dd><p>Returns or sets an <code>integer(2)</code>
representation of reads flagged to be kept or excluded.</p>
</dd>
<dt>bamSimpleCigar, bamSimpleCigar&lt;-</dt><dd><p>Returns or sets a
<code>logical(1)</code> vector indicating whether reads without indels
or clipping be kept.</p>
</dd>
<dt>bamReverseComplement, bamReverseComplement&lt;-</dt><dd><p>Returns or sets
a <code>logical(1)</code> vector indicating whether reads on the minus
strand will be returned with sequence reverse complemented and
quality reversed.</p>
</dd>
</dl>

<p>Methods:
</p>

<dl>
<dt>show</dt><dd><p>Compactly display the object.</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p>Martin Morgan</p>


<h3>See Also</h3>

<p><code><a href="scanBam.html">scanBam</a></code>
</p>


<h3>Examples</h3>

<pre>
## defaults
p0 &lt;- ScanBamParam()

## subset of reads based on genomic coordinates
which &lt;- IRangesList(seq1=IRanges(1000, 2000),
                     seq2=IRanges(c(100, 1000), c(1000, 2000)))
p1 &lt;- ScanBamParam(what=scanBamWhat(), which=which)

## subset of reads based on 'flag' value
p2 &lt;- ScanBamParam(what=scanBamWhat(),
                   flag=scanBamFlag(isMinusStrand=FALSE))

## subset of fields
p3 &lt;- ScanBamParam(what=c("rname", "strand", "pos", "qwidth"))
                
## use
fl &lt;- system.file("extdata", "ex1.bam", package="Rsamtools",
                  mustWork=TRUE)
res &lt;- scanBam(fl, param=p2)[[1]]
lapply(res, head)

## tags; NM: edit distance; H1: 1-difference hits
p4 &lt;- ScanBamParam(tag=c("NM", "H1"), what="flag")
bam4 &lt;- scanBam(fl, param=p4)
str(bam4[[1]][["tag"]])

## tagFilter
p5 &lt;- ScanBamParam(tag=c("NM", "H1"), tagFilter=list(NM=c(2, 3, 4)))
bam5 &lt;- scanBam(fl, param=p5)
table(bam5[[1]][["tag"]][["NM"]])

## flag utils
flag &lt;- scanBamFlag(isUnmappedQuery=FALSE, isMinusStrand=TRUE)

p6 &lt;- ScanBamParam(what="flag")
bam6 &lt;- scanBam(fl, param=p6)
flag6 &lt;- bam6[[1]][["flag"]]
head(bamFlagAsBitMatrix(flag6[1:9]))
colSums(bamFlagAsBitMatrix(flag6))
flag
bamFlagAsBitMatrix(flag)
</pre>

<hr /><div style="text-align: center;">[Package <em>Rsamtools</em> version 2.6.0 <a href="00Index.html">Index</a>]</div>
</body></html>
