<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Import and export</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for io {rtracklayer}"><tr><td>io {rtracklayer}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2> Import and export  </h2>

<h3>Description</h3>

<p>The functions <code>import</code> and <code>export</code> load and save
objects from and to particular file formats. The rtracklayer package
implements support for a number of annotation and sequence formats.
</p>


<h3>Usage</h3>

<pre>
export(object, con, format, ...)
import(con, format, text, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>object</code></td>
<td>
<p> The object to export. </p>
</td></tr>
<tr valign="top"><td><code>con</code></td>
<td>
<p> The connection from which data is loaded or to which data
is saved. If this is a character vector, it is assumed to be a
filename and a corresponding file connection is created and then
closed after exporting the object. If a <code><a href="RTLFile-class.html">RTLFile</a></code>
derivative, the data is loaded from or saved to the underlying
resource. If missing, the function will return the output as a
character vector, rather than writing to a connection.
</p>
</td></tr>
<tr valign="top"><td><code>format</code></td>
<td>
<p> The format of the output. If missing and <code>con</code> is
a filename, the format is derived from the file extension. This
argument is unnecessary when <code>con</code> is a derivative of
<code><a href="RTLFile-class.html">RTLFile</a></code>.
</p>
</td></tr>
<tr valign="top"><td><code>text</code></td>
<td>
<p> If <code>con</code> is missing, this can be a character vector
directly providing the string data to import. </p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p> Parameters to pass to the format-specific method.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The rtracklayer package supports a number of file formats for
representing annotated genomic intervals. These are
each represented as a subclass of <code><a href="RTLFile-class.html">RTLFile</a></code>. Below,
we list the major supported formats, with some advice for when a
particular file format is appropriate:
</p>

<dl>
<dt><acronym><span class="acronym"><a href="GFFFile-class.html">GFF</a></span></acronym></dt><dd><p>The General Feature Format is
meant to represent any set of genomic features, with
application-specific columns represented as
&ldquo;attributes&rdquo;. There are three principal versions (1, 2, and
3). This is a good format for interoperating with other genomic
tools and is the most flexible format, in that a feature may have
any number of attributes (in version 2 and above). Version 3
(GFF3) is the preferred version. Its specification lays out
conventions for representing various types of data, including gene
models, for which it is the format of choice. For variants,
rtracklayer has rudimentary support for an extention of GFF3
called GVF. UCSC supports GFF1, but it needs to be encapsulated in
the UCSC metaformat, i.e. <code>export.ucsc(subformat =
      "gff1")</code>. The BED format is typically preferred over GFF for
interaction with UCSC. GFF files can be indexed with the tabix
utility for fast range-based queries via rtracklayer and
Rsamtools.
</p>
</dd>
<dt><acronym><span class="acronym"><a href="BEDFile-class.html">BED</a></span></acronym></dt><dd><p>The Browser Extended Display
format is for displaying qualitative tracks in a genome browser,
in particular UCSC. It finds a good balance between simplicity and
expressiveness. It is much simpler than GFF and yet can still
represent multi-exon gene structures. It is somewhat limited by
its lack of the attribute support of GFF. To circumvent this, many
tools and organizations have extended BED with additional
columns. Use the <code>extraCols</code> argument on <code>import</code> to
read those columns into R. The rtracklayer package supports two official
extensions of BED: Bed15 and bedGraph, and the unofficial BEDPE
format, see below.  BED files can be indexed with the tabix utility
for fast range-based queries via rtracklayer and Rsamtools.
</p>
</dd>
<dt><acronym><span class="acronym"><a href="BEDFile-class.html">Bed15</a></span></acronym></dt><dd><p>An extension of BED with 15
columns, Bed15 is meant to represent data from microarray
experiments. Multiple samples/columns are supported, and the data
is displayed in UCSC as a compact heatmap. Few other tools support
this format. With 15 columns per feature, this format is probably
too verbose for e.g. ChIP-seq coverage (use multiple BigWig tracks
instead).</p>
</dd>
<dt><acronym><span class="acronym"><a href="BEDFile-class.html">bedGraph</a></span></acronym></dt><dd><p>A variant of BED that
represents a score column more compactly than <acronym><span class="acronym">BED</span></acronym> and
especially <acronym><span class="acronym">Bed15</span></acronym>, although only one sample is
supported. The data is displayed in UCSC as a bar or line
graph. For large data (the typical case), <code>BigWig</code> is
preferred.
</p>
</dd>
<dt><acronym><span class="acronym"><a href="BEDFile-class.html">BEDPE</a></span></acronym></dt><dd><p>A variant of BED that
represents pairs of genomic regions, such as interaction data or
chromosomal rearrangements. The data cannot be displayed in UCSC
directly but can be represented using the BED12 format.
</p>
</dd>
<dt><acronym><span class="acronym"><a href="WIGFile-class.html">WIG</a></span></acronym></dt><dd><p>The Wiggle format is meant for
storing dense numerical data, such as window-based GC and
conservation scores. The data is displayed in UCSC as a bar or
line graph. The WIG format only works for intervals with a uniform
width. For non-uniform widths, consider <code>bedGraph</code>. For large
data, consider <code>BigWig</code>.
</p>
</dd>
<dt><acronym><span class="acronym"><a href="BigWigFile.html">BigWig</a></span></acronym></dt><dd><p>The BigWig format is a
binary version of both <code>bedGraph</code> and <code>WIG</code> (which are
now somewhat obsolete). A BigWig file contains a spatial index for
fast range-based queries and also embeds summary statistics of the
scores at several zoom levels. Thus, it is ideal for visualization
of and parallel computing on genome-scale vectors, like the
coverage from a high-throughput sequencing experiment.
</p>
</dd>  
</dl>

<p>In summary, for the typical use case of combining gene models with
experimental data, <acronym><span class="acronym">GFF</span></acronym> is preferred for gene models and
<code>BigWig</code> is preferred for quantitative score vectors. Note that
the Rsamtools package provides support for the
<code><a href="../../Rsamtools/html/BamFile.html">BAM</a></code> file format (for representing
read alignments), among others. Based on this, the rtracklayer package
provides an <code>export</code> method for writing <code>GAlignments</code>
and <code>GappedReads</code> objects as <code>BAM</code>. For variants, consider
VCF, supported by the VariantAnnotation package.
</p>
<p>There is also support for reading and writing biological sequences,
including the UCSC <code><a href="TwoBitFile-class.html">TwoBit</a></code> format for
compactly storing a genome sequence along with a mask. The files are
binary, so they are efficiently queried for particular ranges. A
similar format is <code><a href="../../Rsamtools/html/FaFile.html">FA</a></code>, supported by
Rsamtools.
</p>


<h3>Value</h3>

<p>If <code>con</code> is missing, a character vector containing the string
output. Otherwise, nothing is returned.
</p>


<h3>Author(s)</h3>

<p> Michael Lawrence </p>


<h3>See Also</h3>

<p>Format-specific options for the popular formats:
<acronym><span class="acronym"><a href="GFFFile-class.html">GFF</a></span></acronym>, <acronym><span class="acronym"><a href="BEDFile-class.html">BED</a></span></acronym>,
<acronym><span class="acronym"><a href="BEDFile-class.html">Bed15</a></span></acronym>,
<acronym><span class="acronym"><a href="BEDFile-class.html">bedGraph</a></span></acronym>,
<acronym><span class="acronym"><a href="WIGFile-class.html">WIG</a></span></acronym>,
<acronym><span class="acronym"><a href="BigWigFile.html">BigWig</a></span></acronym>
</p>


<h3>Examples</h3>

<pre>
  track &lt;- import(system.file("tests", "v1.gff", package = "rtracklayer"))
  ## Not run: export(track, "my.gff", version = "3")
  ## equivalently,
  ## Not run: export(track, "my.gff3")
  ## or
  ## Not run: 
  con &lt;- file("my.gff3")
  export(track, con, "gff3")
  
## End(Not run)
  ## or as a string
  export(track, format = "gff3")
</pre>

<hr /><div style="text-align: center;">[Package <em>rtracklayer</em> version 1.50.0 <a href="00Index.html">Index</a>]</div>
</body></html>
