<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: TabixFile Import/Export</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for TabixFile-methods {rtracklayer}"><tr><td>TabixFile-methods {rtracklayer}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>TabixFile Import/Export</h2>

<h3>Description</h3>

<p>These methods support the import and export of
<code>Rsamtools:TabixFile</code>TabixFile objects. These are generally
useful when working with tabix-indexed files that have a non-standard
format (i.e., not BED nor GFF), as well as exporting an object with
arbitrary columns (like a GRanges) to an indexed, tab-separated
file. This relies on the tabix header, which indicates the columns in
the file that correspond to the chromosome, start and end. The BED and
GFF parsers handle tabix transparently.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'TabixFile,ANY,ANY'
import(con, format, text,
                   which = if (is.na(genome)) NULL
                           else as(seqinfoForGenome(genome), "GenomicRanges"),
                   genome = NA, header = TRUE, ...)
exportToTabix(object, con, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>con</code></td>
<td>
<p>For <code>import</code>, a <code>TabixFile</code> object;
for <code>exportToTabix</code>, a string naming the destination file.
</p>
</td></tr>
<tr valign="top"><td><code>object</code></td>
<td>
<p>The object to export. It is coerced to a
<code>data.frame</code>, written to a tab-separated file, and indexed with
tabix for efficient range-based retrieval of the data using
<code>import</code>.
</p>
</td></tr>
<tr valign="top"><td><code>format</code></td>
<td>
<p>If any known format, like &ldquo;bed&rdquo; or &ldquo;gff&rdquo;
(or one of their variants), then the appropriate parser is
applied. If any other value, then the tabix header is consulted for
the format. By default, this is taken from the file extension.
</p>
</td></tr>
<tr valign="top"><td><code>text</code></td>
<td>
<p>Ignored.</p>
</td></tr>
<tr valign="top"><td><code>which</code></td>
<td>
<p>A range data structure coercible to <code>IntegerRangesList</code>,
like a <code>GRanges</code>. Only the intervals in the file overlapping the
given ranges are returned. The default is to use the range over the
entire genome given by <code>genome</code>, if specified.
</p>
</td></tr>
<tr valign="top"><td><code>genome</code></td>
<td>
<p>The identifier of a genome, or <code>NA</code> if
unknown. Typically, this is a UCSC identifier like &ldquo;hg19&rdquo;. An
attempt will be made to derive the <code>seqinfo</code> on the return
value using either an installed BSgenome package or UCSC, if network
access is available.
</p>
</td></tr>
<tr valign="top"><td><code>header</code></td>
<td>
<p>If <code>TRUE</code>, then the header in the indexed file,
which might include a track line, is sent to the parser. Otherwise,
the initial lines are skipped, according to the <code>skip</code> field in
the tabix index header.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Extra arguments to pass to the underlying import routine,
which for non-standard formats is <code><a href="../../utils/html/read.table.html">read.table</a></code> or
<code><a href="../../utils/html/write.table.html">write.table</a></code>.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>For <code>import</code>, a <code>GRanges</code> object.
</p>
<p>For <code>exportToTabix</code>, a <code>TabixFile</code> object that is directly
passable to <code>import</code>.
</p>


<h3>Author(s)</h3>

<p>Michael Lawrence</p>


<h3>References</h3>

<p><a href="http://samtools.sourceforge.net/tabix.shtml">http://samtools.sourceforge.net/tabix.shtml</a>
</p>


<h3>See Also</h3>

<p><code><a href="../../Rsamtools/html/scanTabix.html">scanTabix</a></code> and friends
</p>

<hr /><div style="text-align: center;">[Package <em>rtracklayer</em> version 1.50.0 <a href="00Index.html">Index</a>]</div>
</body></html>
