<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Querying UCSC Tables</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for UCSCTableQuery-class {rtracklayer}"><tr><td>UCSCTableQuery-class {rtracklayer}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Querying UCSC Tables</h2>

<h3>Description</h3>

<p>The UCSC genome browser is backed by a large database,
which is exposed by the Table Browser web interface. Tracks are
stored as tables, so this is also the mechanism for retrieving tracks. The
<code>UCSCTableQuery</code> class represents a query against the Table
Browser. Storing the query fields in a formal class facilitates
incremental construction and adjustment of a query.</p>


<h3>Details</h3>

<p>There are five supported fields for a table query:
</p>

<dl>
<dt>session</dt><dd><p>The <code><a href="ucscSession-class.html">UCSCSession</a></code> instance from
the tables are retrieved. Although all sessions are based on the
same database, the set of user-uploaded tracks, which are represented
as tables, is not the same, in general.
</p>
</dd>
<dt>trackName</dt><dd><p>The name of a track from which to retrieve a
table. Each track can have multiple tables. Many times there is a
primary table that is used to display the track, while the other
tables are supplemental. Sometimes, tracks are displayed by
aggregating multiple tables. If <code>NULL</code>, search for a primary
table across all of the tracks (will not find secondary tables).
</p>
</dd>
<dt>tableName</dt><dd><p>The name of the specific table to retrieve. May be
<code>NULL</code>, in which case the behavior depends on how the query
is executed, see below.
</p>
</dd>
<dt>range</dt><dd><p>A genome identifier, a
<code><a href="../../GenomicRanges/html/GRanges.html">GRanges</a></code> or 
a <code><a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a></code> indicating
the portion of the table to retrieve, in genome coordinates.
Simply specifying the genome string is the easiest way to download
data for the entire genome, and <code><a href="GRangesForUCSCGenome.html">GRangesForUCSCGenome</a></code>
facilitates downloading data for e.g. an entire chromosome.
</p>
</dd>
<dt>names</dt><dd><p>Names/accessions of the desired features</p>
</dd>
</dl>
 
<p>A common workflow for querying the UCSC database is to create an
instance of <code>UCSCTableQuery</code> using the <code>ucscTableQuery</code>
constructor, invoke <code>tableNames</code> to list the available tables for
a track, and finally to retrieve the desired table either as a
<code>data.frame</code> via <code>getTable</code> or as a track
via <code>track</code>. See the examples.
</p>
<p>The reason for a formal query class is to facilitate multiple queries
when the differences between the queries are small. For example, one
might want to query multiple tables within the track and/or same
genomic region, or query the same table for multiple regions. The
<code>UCSCTableQuery</code> instance can be incrementally adjusted for each
new query. Some caching is also performed, which enhances performance. 
</p>


<h3>Constructor</h3>


<dl>
<dt></dt><dd>
<p><code>ucscTableQuery(x, track, range = seqinfo(x), table = NULL,
        names = NULL)</code>: Creates a <code>UCSCTableQuery</code> with the
<code>UCSCSession</code> or genome identifier given as <code>x</code> and
the track name given by
the single string <code>track</code>. <code>range</code> should be a genome
string identifier, a <code>GRanges</code> instance or
<code>IntegerRangesList</code> instance, and it effectively defaults to
<code>genome(x)</code>. If the genome is missing, it is taken from the
session. The table name is given by
<code>table</code>, which may be a single string or
<code>NULL</code>. Feature names, such as gene identifiers, may be
passed via <code>names</code> as a character vector.
</p>
</dd>
</dl>



<h3>Executing Queries</h3>

<p>Below, <code>object</code> is a <code>UCSCTableQuery</code> instance.
</p>

<dl>
<dt></dt><dd>
<p><code>track(object)</code>:
Retrieves the indicated table as a track, i.e. a <code>GRanges</code>
object. Note that not all tables are available as tracks.
</p>
</dd>
<dt></dt><dd>
<p><code>getTable(object)</code>: Retrieves the indicated table as a
<code>data.frame</code>. Note that not all tables are output in
parseable form, and that UCSC will truncate responses if they
exceed certain limits (usually around 100,000 records). The safest
(and most efficient) bet for large queries is to download the file
via FTP and query it locally.
</p>
</dd>
<dt></dt><dd>
<p><code>tableNames(object)</code>: Gets the names of the tables available
for the session, track and range specified by the query.
</p>
</dd>
</dl>



<h3>Accessor methods</h3>

<p>In the code snippets below, <code>x</code>/<code>object</code> is a
<code>UCSCTableQuery</code> object.
</p>

<dl>
<dt></dt><dd><p><code>browserSession(object)</code>,
<code>browserSession(object) &lt;- value</code>:
Get or set the <code>UCSCSession</code> to query.
</p>
</dd>
<dt></dt><dd><p><code>trackName(x)</code>, <code>trackName(x) &lt;- value</code>: Get or
set the single string indicating the track containing the table of
interest.
</p>
</dd>
<dt></dt><dd><p><code>trackNames(x)</code></p>
</dd></dl>
<p>List the names of the tracks available
for retrieval for the assigned genome.
</p>
<dl>
<dt></dt><dd><p><code>tableName(x)</code>, <code>tableName(x) &lt;- value</code>: Get or
set the single string indicating the name of the table to
retrieve. May be <code>NULL</code>, in which case the table is
automatically determined.
</p>
</dd>
<dt></dt><dd><p><code>range(x)</code>, <code>range(x) &lt;- value</code>: Get or set the
<code>GRanges</code> indicating the portion of the table to retrieve in
genomic coordinates. Any missing information, such as the genome
identifier, is filled in using <code>range(browserSession(x))</code>. It
is also possible to set the genome identifier string or
a <code>IntegerRangesList</code>.
</p>
</dd>
<dt></dt><dd><p><code>names(x)</code>, <code>names(x) &lt;- value</code>: Get or set the
names of the features to retrieve. If <code>NULL</code>, this filter is
disabled.
</p>
</dd>
<dt></dt><dd><p><code>ucscSchema(x)</code>: Get
the <code><a href="UCSCSchema-class.html">UCSCSchema</a></code> object describing the selected table.</p>
</dd>
</dl>



<h3>Author(s)</h3>

<p> Michael Lawrence </p>


<h3>Examples</h3>

<pre>
## Not run: 
session &lt;- browserSession()
genome(session) &lt;- "mm9"
trackNames(session) ## list the track names
## choose the Conservation track for a portion of mm9 chr1
query &lt;- ucscTableQuery(session, "Conservation",
                        GRangesForUCSCGenome("mm9", "chr12",
                                             IRanges(57795963, 57815592)))
## list the table names
tableNames(query)
## get the phastCons30way track
tableName(query) &lt;- "phastCons30way"
## retrieve the track data
track(query)  # a GRanges object
## get a data.frame summarizing the multiple alignment
tableName(query) &lt;- "multiz30waySummary"
getTable(query)

genome(session) &lt;- "hg18"
query &lt;- ucscTableQuery(session, "snp129",
                        names = c("rs10003974", "rs10087355", "rs10075230"))
ucscSchema(query)
getTable(query)

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>rtracklayer</em> version 1.50.0 <a href="00Index.html">Index</a>]</div>
</body></html>
