<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: SummarizedExperiment objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for SummarizedExperiment-class {SummarizedExperiment}"><tr><td>SummarizedExperiment-class {SummarizedExperiment}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>SummarizedExperiment objects</h2>

<h3>Description</h3>

<p>The SummarizedExperiment class is a matrix-like container where rows
represent features of interest (e.g. genes, transcripts, exons, etc...)
and columns represent samples (with sample data summarized as a
<a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a>). A SummarizedExperiment object contains one or more
assays, each represented by a matrix-like object of numeric or other mode.
</p>
<p>Note that SummarizedExperiment is the parent of the
<a href="RangedSummarizedExperiment-class.html">RangedSummarizedExperiment</a> class which means that all the methods
documented below also work on a <a href="RangedSummarizedExperiment-class.html">RangedSummarizedExperiment</a> object.
</p>


<h3>Usage</h3>

<pre>

## Constructor

# See ?RangedSummarizedExperiment for the constructor function.

## Accessors

assayNames(x, ...)
assayNames(x, ...) &lt;- value
assays(x, withDimnames=TRUE, ...)
assays(x, withDimnames=TRUE, ...) &lt;- value
assay(x, i, withDimnames=TRUE, ...)
assay(x, i, withDimnames=TRUE, ...) &lt;- value
rowData(x, use.names=TRUE, ...)
rowData(x, ...) &lt;- value
colData(x, ...)
colData(x, ...) &lt;- value
#dim(x)
#dimnames(x)
#dimnames(x) &lt;- value

## Quick colData access

## S4 method for signature 'SummarizedExperiment'
x$name
## S4 replacement method for signature 'SummarizedExperiment'
x$name &lt;- value
## S4 method for signature 'SummarizedExperiment,ANY,missing'
x[[i, j, ...]]
## S4 replacement method for signature 'SummarizedExperiment,ANY,missing'
x[[i, j, ...]] &lt;- value

## Subsetting

## S4 method for signature 'SummarizedExperiment'
x[i, j, ..., drop=TRUE]
## S4 replacement method for signature 'SummarizedExperiment,ANY,ANY,SummarizedExperiment'
x[i, j] &lt;- value
## S4 method for signature 'SummarizedExperiment'
subset(x, subset, select, ...)

## Combining

## S4 method for signature 'SummarizedExperiment'
cbind(..., deparse.level=1)
## S4 method for signature 'SummarizedExperiment'
rbind(..., deparse.level=1)

## On-disk realization
## S4 method for signature 'SummarizedExperiment'
realize(x, BACKEND=getAutoRealizationBackend())
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>A SummarizedExperiment object.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>For <code>assay</code>, arguments in <code>...</code> are forwarded to <code>assays</code>.
</p>
<p>For <code>cbind</code>, <code>rbind</code>, <code>...</code> contains SummarizedExperiment
objects to be combined.
</p>
<p>For other accessors, ignored.
</p>
</td></tr>
<tr valign="top"><td><code>value</code></td>
<td>
<p>An object of a class specified in the S4 method
signature or as outlined in &lsquo;Details&rsquo;.</p>
</td></tr>
<tr valign="top"><td><code>i, j</code></td>
<td>

<p>For <code>assay</code>, <code>assay&lt;-</code>, <code>i</code> is an integer or
numeric scalar; see &lsquo;Details&rsquo; for additional constraints.
</p>
<p>For <code>[,SummarizedExperiment</code>,
<code>[,SummarizedExperiment&lt;-</code>, <code>i</code>, <code>j</code> are subscripts
that can act to subset the rows and columns of <code>x</code>, that is the
<code>matrix</code> elements of <code>assays</code>.
</p>
<p>For <code>[[,SummarizedExperiment</code>,
<code>[[&lt;-,SummarizedExperiment</code>, <code>i</code> is a scalar index (e.g.,
<code>character(1)</code> or <code>integer(1)</code>) into a column of
<code>colData</code>.
</p>
</td></tr>
<tr valign="top"><td><code>name</code></td>
<td>
<p>A symbol representing the name of a column of
<code>colData</code>.</p>
</td></tr>
<tr valign="top"><td><code>withDimnames</code></td>
<td>
<p>A <code>logical(1)</code>, indicating whether the
dimnames of the SummarizedExperiment object should be applied (i.e.
copied) to the extracted assays. More precisely, setting
<code>withDimnames=FALSE</code> in the <em>getter</em> returns the
assays <em>as-is</em> whereas setting <code>withDimnames=FALSE</code>
return them with possibly modified dimnames.
</p>
<p>Setting <code>withDimnames=FALSE</code> in the <em>setter</em> (<code>assays&lt;-</code>)
is required when the dimnames on the supplied assays are not identical
to the dimnames on the SummarizedExperiment object; it does not
influence actual assignment of dimnames to assays (they're always
stored as-is).
</p>
<p>Note that </p>
<pre>  assays(x, withDimnames=FALSE) &lt;- assays(x, withDimnames=FALSE)</pre>
<p>is guaranteed to always work and be a no-op. This is not the case
if <code>withDimnames=TRUE</code> is used or if <code>withDimnames</code> is
not specified.
</p>
</td></tr>
<tr valign="top"><td><code>use.names</code></td>
<td>
<p>Like <code><a href="../../S4Vectors/html/mcols.html">mcols</a>(x)</code>, by default
<code>rowData(x)</code> propagates the rownames of <code>x</code> to the returned
<a href="../../S4Vectors/html/DataFrame.html">DataFrame</a> object (note that for a SummarizedExperiment
object, the rownames are also the names i.e. <code>rownames(x)</code> is
always the same as <code>names(x)</code>). Setting <code>use.names=FALSE</code>
suppresses this propagation i.e. it returns a <a href="../../S4Vectors/html/DataFrame.html">DataFrame</a>
object with no rownames. Use this when <code>rowData(x)</code> fails,
which can happen when the rownames contain NAs (because the rownames
of a SummarizedExperiment object can contain NAs, but the rownames of
a <a href="../../S4Vectors/html/DataFrame.html">DataFrame</a> object cannot).</p>
</td></tr>
<tr valign="top"><td><code>drop</code></td>
<td>
<p>A <code>logical(1)</code>, ignored by these methods.</p>
</td></tr>
<tr valign="top"><td><code>deparse.level</code></td>
<td>
<p>See <code>?base::<a href="../../base/html/cbind.html">cbind</a></code> for a description of
this argument.</p>
</td></tr>
<tr valign="top"><td><code>subset</code></td>
<td>
<p>An expression which, when evaluated in the
context of <code>rowData(x)</code>, is a logical vector indicating
elements or rows to keep: missing values are taken as false.</p>
</td></tr>
<tr valign="top"><td><code>select</code></td>
<td>
<p>An expression which, when evaluated in the
context of <code>colData(x)</code>, is a logical vector indicating
elements or rows to keep: missing values are taken as false.</p>
</td></tr>
<tr valign="top"><td><code>BACKEND</code></td>
<td>
<p><code>NULL</code> (the default), or a single string specifying the
name of the backend. When the backend is set to <code>NULL</code>,
each element of <code>assays(x)</code> is realized in memory as an ordinary array
by just calling <code>as.array</code> on it.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The SummarizedExperiment class is meant for numeric and other
data types derived from a sequencing experiment. The structure is
rectangular like a <code>matrix</code>, but with additional annotations on
the rows and columns, and with the possibility to manage several
assays simultaneously.
</p>
<p>The rows of a SummarizedExperiment object represent features
of interest. Information about these features is stored in a
<a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> object, accessible using the function
<code>rowData</code>. The <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> must have as many rows
as there are rows in the SummarizedExperiment object, with each row
of the <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> providing information on the feature in the
corresponding row of the SummarizedExperiment object. Columns of the
<a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> represent different attributes of the features
of interest, e.g., gene or transcript IDs, etc.
</p>
<p>Each column of a SummarizedExperiment object represents a sample.
Information about the samples are stored in a <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a>,
accessible using the function <code>colData</code>, described below.
The <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> must have as many rows as there are
columns in the SummarizedExperiment object, with each row of the
<a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> providing information on the sample in the
corresponding column of the SummarizedExperiment object.
Columns of the <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> represent different sample
attributes, e.g., tissue of origin, etc. Columns of the
<a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> can themselves be annotated (via the
<code><a href="../../S4Vectors/html/mcols.html">mcols</a></code> function). Column names typically
provide a short identifier unique to each sample.
</p>
<p>A SummarizedExperiment object can also contain information about
the overall experiment, for instance the lab in which it was conducted,
the publications with which it is associated, etc. This information is
stored as a <code>list</code> object, accessible using the <code>metadata</code>
function. The form of the data associated with the experiment is left to
the discretion of the user.
</p>
<p>The SummarizedExperiment container is appropriate for matrix-like
data. The data are accessed using the <code>assays</code> function,
described below. This returns a <a href="../../S4Vectors/html/SimpleList-class.html">SimpleList</a> object. Each
element of the list must itself be a matrix (of any mode) and must
have dimensions that are the same as the dimensions of the
SummarizedExperiment in which they are stored. Row and column
names of each matrix must either be <code>NULL</code> or match those of the
SummarizedExperiment during construction. It is convenient for
the elements of <a href="../../S4Vectors/html/SimpleList-class.html">SimpleList</a> of assays to be named.
</p>


<h3>Constructor</h3>

<p>SummarizedExperiment instances are constructed using the
<code>SummarizedExperiment</code> function documented in
<code>?<a href="RangedSummarizedExperiment-class.html">RangedSummarizedExperiment</a></code>.
</p>


<h3>Accessors</h3>

<p>In the following code snippets, <code>x</code> is a SummarizedExperiment
object.
</p>

<dl>
<dt><code>assays(x)</code>, <code>assays(x) &lt;- value</code>:</dt><dd><p>Get or set the
assays. <code>value</code> is a <code>list</code> or <code>SimpleList</code>, each
element of which is a matrix with the same dimensions as
<code>x</code>.</p>
</dd>
<dt><code>assay(x, i)</code>, <code>assay(x, i) &lt;- value</code>:</dt><dd><p>A convenient
alternative (to <code>assays(x)[[i]]</code>, <code>assays(x)[[i]] &lt;-
      value</code>) to get or set the <code>i</code>th (default first) assay
element. <code>value</code> must be a matrix of the same dimension as
<code>x</code>, and with dimension names <code>NULL</code> or consistent with
those of <code>x</code>.</p>
</dd>
<dt><code>assayNames(x)</code>, <code>assayNames(x) &lt;- value</code>:</dt><dd><p>Get or
set the names of <code>assay()</code> elements.</p>
</dd>
<dt><code>rowData(x, use.names=TRUE)</code>, <code>rowData(x) &lt;- value</code>:</dt><dd>
<p>Get or set the row data. <code>value</code> is a <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> object.</p>
</dd>
<dt><code>colData(x)</code>, <code>colData(x) &lt;- value</code>:</dt><dd><p>Get or set the
column data. <code>value</code> is a <a href="../../S4Vectors/html/DataFrame-class.html">DataFrame</a> object. Row
names of <code>value</code> must be NULL or consistent with the existing
column names of <code>x</code>.</p>
</dd>
<dt><code>metadata(x)</code>, <code>metadata(x) &lt;- value</code>:</dt><dd><p>Get or set
the experiment data. <code>value</code> is a <code>list</code> with arbitrary
content.</p>
</dd>
<dt><code>dim(x)</code>:</dt><dd><p>Get the dimensions (features of interest x samples)
of the SummarizedExperiment.</p>
</dd>
<dt><code>dimnames(x)</code>, <code>dimnames(x) &lt;- value</code>:</dt><dd><p>Get or set
the dimension names. <code>value</code> is usually a list of length 2,
containing elements that are either <code>NULL</code> or vectors of
appropriate length for the corresponding dimension. <code>value</code>
can be <code>NULL</code>, which removes dimension names. This method
implies that <code>rownames</code>, <code>rownames&lt;-</code>, <code>colnames</code>,
and <code>colnames&lt;-</code> are all available.</p>
</dd>
</dl>



<h3>Subsetting</h3>

<p>In the code snippets below, <code>x</code> is a SummarizedExperiment object.
</p>

<dl>
<dt><code>x[i,j]</code>, <code>x[i,j] &lt;- value</code>:</dt><dd><p>Create or replace a
subset of <code>x</code>. <code>i</code>, <code>j</code> can be <code>numeric</code>,
<code>logical</code>, <code>character</code>, or <code>missing</code>. <code>value</code>
must be a SummarizedExperiment object with dimensions,
dimension names, and assay elements consistent with the subset
<code>x[i,j]</code> being replaced.</p>
</dd>
<dt><code>subset(x, subset, select)</code>:</dt><dd><p>Create a subset of <code>x</code>
using an expression <code>subset</code> referring to columns of
<code>rowData(x)</code> and / or <code>select</code> referring to column names
of <code>colData(x)</code>.</p>
</dd>
</dl>

<p>Additional subsetting accessors provide convenient access to
<code>colData</code> columns
</p>

<dl>
<dt><code>x$name</code>, <code>x$name &lt;- value</code></dt><dd><p>Access or replace
column <code>name</code> in <code>x</code>.</p>
</dd>
<dt><code>x[[i, ...]]</code>, <code>x[[i, ...]] &lt;- value</code></dt><dd><p>Access or
replace column <code>i</code> in <code>x</code>.</p>
</dd>
</dl>



<h3>Combining</h3>

<p>In the code snippets below, <code>...</code> are SummarizedExperiment objects
to be combined.
</p>

<dl>
<dt><code>cbind(...)</code>:</dt><dd>
<p><code>cbind</code> combines objects with the same features of interest
but different samples (columns in <code>assays</code>).
The colnames in <code>colData(SummarizedExperiment)</code> must match or
an error is thrown.
Duplicate columns of <code>rowData(SummarizedExperiment)</code> must
contain the same data.
</p>
<p>Data in <code>assays</code> are combined by name matching; if all assay
names are NULL matching is by position. A mixture of names and NULL
throws an error.
</p>
<p><code>metadata</code> from all objects are combined into a <code>list</code>
with no name checking.
</p>
</dd>
<dt><code>rbind(...)</code>:</dt><dd>
<p><code>rbind</code> combines objects with the same samples
but different features of interest (rows in <code>assays</code>).
The colnames in <code>rowData(SummarizedExperiment)</code> must match or
an error is thrown.
Duplicate columns of <code>colData(SummarizedExperiment)</code> must
contain the same data.
</p>
<p>Data in <code>assays</code> are combined by name matching; if all assay
names are NULL matching is by position. A mixture of names and NULL
throws an error.
</p>
<p><code>metadata</code> from all objects are combined into a <code>list</code>
with no name checking.
</p>
</dd>
</dl>



<h3>Implementation and Extension</h3>

<p>This section contains advanced material meant for package developers.
</p>
<p>SummarizedExperiment is implemented as an S4 class, and can be extended in
the usual way, using <code>contains="SummarizedExperiment"</code> in the new
class definition.
</p>
<p>In addition, the representation of the <code>assays</code> slot of
SummarizedExperiment is as a virtual class Assays. This
allows derived classes (<code>contains="Assays"</code>) to implement
alternative requirements for the assays, e.g., backed by file-based
storage like NetCDF or the <code>ff</code> package, while re-using the existing
SummarizedExperiment class without modification.
See <a href="Assays-class.html">Assays</a> for more information.
</p>


<h3>Author(s)</h3>

<p>Martin Morgan, <a href="mtmorgan@fhcrc.org">mtmorgan@fhcrc.org</a></p>


<h3>See Also</h3>


<ul>
<li> <p><a href="RangedSummarizedExperiment-class.html">RangedSummarizedExperiment</a> objects.
</p>
</li>
<li> <p><a href="../../S4Vectors/html/DataFrame.html">DataFrame</a>, <a href="../../S4Vectors/html/SimpleList.html">SimpleList</a>, and
<a href="../../S4Vectors/html/Annotated.html">Annotated</a> objects in the <span class="pkg">S4Vectors</span> package.
</p>
</li>
<li><p> The <code><a href="../../S4Vectors/html/metadata.html">metadata</a></code> and
<code><a href="../../S4Vectors/html/mcols.html">mcols</a></code> accessors in the <span class="pkg">S4Vectors</span>
package.
</p>
</li>
<li> <p><code><a href="../../HDF5Array/html/saveHDF5SummarizedExperiment.html">saveHDF5SummarizedExperiment</a></code> and
<code><a href="../../HDF5Array/html/loadHDF5SummarizedExperiment.html">loadHDF5SummarizedExperiment</a></code> in the
<span class="pkg">HDF5Array</span> package for saving/loading an HDF5-based
SummarizedExperiment object to/from disk.
</p>
</li>
<li><p> The <code><a href="../../DelayedArray/html/realize.html">realize</a></code> generic function in the
<span class="pkg">DelayedArray</span> package for more information about on-disk
realization of objects carrying delayed operations.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
nrows &lt;- 200; ncols &lt;- 6
counts &lt;- matrix(runif(nrows * ncols, 1, 1e4), nrows)
colData &lt;- DataFrame(Treatment=rep(c("ChIP", "Input"), 3),
                     row.names=LETTERS[1:6])
se0 &lt;- SummarizedExperiment(assays=SimpleList(counts=counts),
                            colData=colData)
se0
dim(se0)
dimnames(se0)
assayNames(se0)
head(assay(se0))
assays(se0) &lt;- endoapply(assays(se0), asinh)
head(assay(se0))

rowData(se0)
colData(se0)

se0[, se0$Treatment == "ChIP"]
subset(se0, select = Treatment == "ChIP")

## cbind() combines objects with the same features of interest
## but different samples:
se1 &lt;- se0
se2 &lt;- se1[,1:3]
colnames(se2) &lt;- letters[seq_len(ncol(se2))]
cmb1 &lt;- cbind(se1, se2)
dim(cmb1)
dimnames(cmb1)

## rbind() combines objects with the same samples but different
## features of interest:
se1 &lt;- se0
se2 &lt;- se1[1:50,]
rownames(se2) &lt;- letters[seq_len(nrow(se2))]
cmb2 &lt;- rbind(se1, se2)
dim(cmb2)
dimnames(cmb2)

## ---------------------------------------------------------------------
## ON-DISK REALIZATION
## ---------------------------------------------------------------------
library(DelayedArray)
setAutoRealizationBackend("HDF5Array")
cmb3 &lt;- realize(cmb2)
assay(cmb3, withDimnames=FALSE)  # an HDF5Matrix object
</pre>

<hr /><div style="text-align: center;">[Package <em>SummarizedExperiment</em> version 1.20.0 <a href="00Index.html">Index</a>]</div>
</body></html>
