<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract non-overlapping exonic or intronic parts from a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for exonicParts {GenomicFeatures}"><tr><td>exonicParts {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Extract non-overlapping exonic or intronic parts from a TxDb-like object
</h2>

<h3>Description</h3>

<p><code>exonicParts</code> and <code>intronicParts</code> extract the non-overlapping
(a.k.a. disjoint) exonic or intronic parts from a <a href="TxDb-class.html">TxDb</a>-like object.
</p>


<h3>Usage</h3>

<pre>
exonicParts(txdb, linked.to.single.gene.only=FALSE)
intronicParts(txdb, linked.to.single.gene.only=FALSE)

## 3 helper functions used internally by exonicParts() and intronicParts():
tidyTranscripts(txdb, drop.geneless=FALSE)
tidyExons(txdb, drop.geneless=FALSE)
tidyIntrons(txdb, drop.geneless=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>txdb</code></td>
<td>

<p>A <a href="TxDb-class.html">TxDb</a> object, or any <a href="TxDb-class.html">TxDb</a>-like object that supports the
<code><a href="transcripts.html">transcripts</a>()</code> and <code><a href="transcriptsBy.html">exonsBy</a>()</code> extractors
(e.g. an <a href="../../ensembldb/html/EnsDb.html">EnsDb</a> object).
</p>
</td></tr>
<tr valign="top"><td><code>linked.to.single.gene.only</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>.
</p>
<p>If <code>FALSE</code> (the default), then the disjoint parts are obtained
by calling <code><a href="../../IRanges/html/disjoin.html">disjoin</a>()</code> on all the exons (or introns)
in <code>txdb</code>, including on exons (or introns) not linked to a gene or
linked to more than one gene.
</p>
<p>If <code>TRUE</code>, then the disjoint parts are obtained in 2 steps:
</p>

<ol>
<li><p> call <code><a href="../../IRanges/html/disjoin.html">disjoin</a>()</code> on the exons (or introns)
linked to <em>at least one gene</em>,
</p>
</li>
<li><p> then drop the parts linked to more than one gene from
the set of exonic (or intronic) parts obtained previously.
</p>
</li></ol>

</td></tr>
<tr valign="top"><td><code>drop.geneless</code></td>
<td>

<p>If <code>FALSE</code> (the default), then all the transcripts (or exons, or
introns) get extracted from the <a href="TxDb-class.html">TxDb</a> object.
</p>
<p>If <code>TRUE</code>, then only the transcripts (or exons, or introns) that
are linked to a gene get extracted from the <a href="TxDb-class.html">TxDb</a> object.
</p>
<p>Note that <code>drop.geneless</code> also impacts the order in which the
features are returned:
</p>

<ul>
<li><p> Transcripts: If <code>drop.geneless</code> is <code>FALSE</code> then
transcripts are returned in the same order as with
<code><a href="transcripts.html">transcripts</a></code>, which is expected to be by
internal transcript id (<code>tx_id</code>).
Otherwise they are ordered first by gene id (<code>gene_id</code>),
then by internal transcript id.
</p>
</li>
<li><p> Exons: If <code>drop.geneless</code> is <code>FALSE</code> then exons are
ordered first by internal transcript id (<code>tx_id</code>),
then by exon rank (<code>exon_rank</code>).
Otherwise they are ordered first by gene id (<code>gene_id</code>),
then by internal transcript id, and then by exon rank.
</p>
</li>
<li><p> Introns: If <code>drop.geneless</code> is <code>FALSE</code> then introns
are ordered by internal transcript id (<code>tx_id</code>).
Otherwise they are ordered first by gene id (<code>gene_id</code>),
then by internal transcript id.
</p>
</li></ul>

</td></tr>
</table>


<h3>Value</h3>

<p><code>exonicParts</code> returns a disjoint and strictly sorted
<a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object with 1 range per exonic part
and with metadata columns <code>tx_id</code>, <code>tx_name</code>, <code>gene_id</code>,
<code>exon_id</code>, <code>exon_name</code>, and <code>exon_rank</code>.
If <code>linked.to.single.gene.only</code> was set to <code>TRUE</code>,
an additional <code>exonic_part</code> metadata column is added that
indicates the rank of each exonic part within all the exonic parts
linked to the same gene.
</p>
<p><code>intronicParts</code> returns a disjoint and strictly sorted
<a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object with 1 range per intronic part
and with metadata columns <code>tx_id</code>, <code>tx_name</code>, and <code>gene_id</code>.
If <code>linked.to.single.gene.only</code> was set to <code>TRUE</code>,
an additional <code>intronic_part</code> metadata column is added that
indicates the rank of each intronic part within all the intronic parts
linked to the same gene.
</p>
<p><code>tidyTranscripts</code> returns a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object
with 1 range per transcript and with metadata columns <code>tx_id</code>,
<code>tx_name</code>, and <code>gene_id</code>.
</p>
<p><code>tidyExons</code> returns a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object
with 1 range per exon and with metadata columns <code>tx_id</code>,
<code>tx_name</code>, <code>gene_id</code>, <code>exon_id</code>, <code>exon_name</code>,
and <code>exon_rank</code>.
</p>
<p><code>tidyIntrons</code> returns a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object
with 1 range per intron and with metadata columns <code>tx_id</code>,
<code>tx_name</code>, and <code>gene_id</code>.
</p>


<h3>Note</h3>

<p><code>exonicParts</code> is a replacement for <code><a href="disjointExons.html">disjointExons</a></code> with
the following differences/improvements:
</p>

<ul>
<li><p> Argument <code>linked.to.single.gene.only</code> in <code>exonicParts</code>
replaces argument <code>aggregateGenes</code> in <code>disjointExons</code>,
but has opposite meaning i.e.
<code>exonicParts(txdb, linked.to.single.gene.only=TRUE)</code>
returns the same exonic parts as
<code>disjointExons(txdb, aggregateGenes=FALSE)</code>.
</p>
</li>
<li><p> Unlike <code>disjointExons(txdb, aggregateGenes=TRUE)</code>,
<code>exonicParts(txdb, linked.to.single.gene.only=FALSE)</code> does
NOT discard exon parts that are not linked to a gene.
</p>
</li>
<li> <p><code>exonicParts</code> is almost 2x more efficient than
<code>disjointExons</code>.
</p>
</li>
<li> <p><code>exonicParts</code> works out-of-the-box on any <a href="TxDb-class.html">TxDb</a>-like
object that supports the <code><a href="transcripts.html">transcripts</a>()</code> and
<code><a href="transcriptsBy.html">exonsBy</a>()</code> extractors (e.g. on an
<a href="../../ensembldb/html/EnsDb.html">EnsDb</a> object).
</p>
</li></ul>



<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="../../IRanges/html/disjoin.html">disjoin</a></code> in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><code><a href="transcripts.html">transcripts</a></code>, <code><a href="transcriptsBy.html">transcriptsBy</a></code>,
and <code><a href="transcriptsByOverlaps.html">transcriptsByOverlaps</a></code>, for extracting
genomic feature locations from a <a href="TxDb-class.html">TxDb</a>-like object.
</p>
</li>
<li> <p><code><a href="transcriptLengths.html">transcriptLengths</a></code> for extracting the transcript
lengths (and other metrics) from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="extractTranscriptSeqs.html">extractTranscriptSeqs</a></code> for extracting transcript
(or CDS) sequences from chromosome sequences.
</p>
</li>
<li> <p><code><a href="coverageByTranscript.html">coverageByTranscript</a></code> for computing coverage by
transcript (or CDS) of a set of ranges.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene

## ---------------------------------------------------------------------
## exonicParts()
## ---------------------------------------------------------------------

exonic_parts1 &lt;- exonicParts(txdb)
exonic_parts1

## Mapping from exonic parts to genes is many-to-many:
gene_id1 &lt;- mcols(exonic_parts1)$gene_id
gene_id1  # CharacterList object
table(lengths(gene_id1))
## The number of known genes a Human exonic part can be linked to
## varies from 0 to 22!

exonic_parts2 &lt;- exonicParts(txdb, linked.to.single.gene.only=TRUE)
exonic_parts2

## Mapping from exonic parts to genes now is many-to-one:
gene_id2 &lt;- mcols(exonic_parts2)$gene_id
gene_id2[1:20]  # character vector

## Select exonic parts for a given gene:
exonic_parts2[gene_id2 %in% "643837"]

## Sanity checks:
stopifnot(isDisjoint(exonic_parts1), isStrictlySorted(exonic_parts1))
stopifnot(isDisjoint(exonic_parts2), isStrictlySorted(exonic_parts2))
stopifnot(all(exonic_parts2 %within% reduce(exonic_parts1)))
stopifnot(identical(
    lengths(gene_id1) == 1L,
    exonic_parts1 %within% exonic_parts2
))

## ---------------------------------------------------------------------
## intronicParts()
## ---------------------------------------------------------------------

intronic_parts1 &lt;- intronicParts(txdb)
intronic_parts1

## Mapping from intronic parts to genes is many-to-many:
mcols(intronic_parts1)$gene_id
table(lengths(mcols(intronic_parts1)$gene_id))
## A Human intronic part can be linked to 0 to 22 known genes!

intronic_parts2 &lt;- intronicParts(txdb, linked.to.single.gene.only=TRUE)
intronic_parts2

## Mapping from intronic parts to genes now is many-to-one:
class(mcols(intronic_parts2)$gene_id)  # character vector

## Sanity checks:
stopifnot(isDisjoint(intronic_parts1), isStrictlySorted(intronic_parts1))
stopifnot(isDisjoint(intronic_parts2), isStrictlySorted(intronic_parts2))
stopifnot(all(intronic_parts2 %within% reduce(intronic_parts1)))
stopifnot(identical(
    lengths(mcols(intronic_parts1)$gene_id) == 1L,
    intronic_parts1 %within% intronic_parts2
))

## ---------------------------------------------------------------------
## Helper functions
## ---------------------------------------------------------------------

tidyTranscripts(txdb)                      # Ordered by 'tx_id'.
tidyTranscripts(txdb, drop.geneless=TRUE)  # Ordered first by 'gene_id',
                                           # then by 'tx_id'.

tidyExons(txdb)                            # Ordered first by 'tx_id',
                                           # then by 'exon_rank'.
tidyExons(txdb, drop.geneless=TRUE)        # Ordered first by 'gene_id',
                                           # then by 'tx_id',
                                           # then by 'exon_rank'.

tidyIntrons(txdb)                          # Ordered by 'tx_id'.
tidyIntrons(txdb, drop.geneless=TRUE)      # Ordered first by 'gene_id',
                                           # then by 'tx_id'.
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
