<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract sequences upstream of a set of genes or transcripts</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for extractUpstreamSeqs {GenomicFeatures}"><tr><td>extractUpstreamSeqs {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Extract sequences upstream of a set of genes or transcripts</h2>

<h3>Description</h3>

<p><code>extractUpstreamSeqs</code> is a generic function for extracting
sequences upstream of a supplied set of genes or transcripts.
</p>


<h3>Usage</h3>

<pre>
extractUpstreamSeqs(x, genes, width=1000, ...)

## Dispatch is on the 2nd argument!

## S4 method for signature 'GenomicRanges'
extractUpstreamSeqs(x, genes, width=1000)

## S4 method for signature 'TxDb'
extractUpstreamSeqs(x, genes, width=1000, exclude.seqlevels=NULL)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>An object containing the chromosome sequences from which to extract the
upstream sequences. It can be a <a href="../../BSgenome/html/BSgenome.html">BSgenome</a>,
<a href="../../rtracklayer/html/TwoBitFile.html">TwoBitFile</a>, or <a href="../../Rsamtools/html/FaFile.html">FaFile</a> object,
or any <em>genome sequence container</em>.
More formally, <code>x</code> must be an object for which
<code><a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a></code> and <code><a href="../../Biostrings/html/getSeq.html">getSeq</a></code>
are defined.
</p>
</td></tr>
<tr valign="top"><td><code>genes</code></td>
<td>

<p>An object containing the locations (i.e. chromosome name, start, end, and
strand) of the genes or transcripts with respect to the reference genome.
Only <a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> and <a href="TxDb-class.html">TxDb</a> objects
are supported at the moment. If the latter, the gene locations are obtained
by calling the <code><a href="transcripts.html">genes</a></code> function on the <a href="TxDb-class.html">TxDb</a>
object internally.
</p>
</td></tr>
<tr valign="top"><td><code>width</code></td>
<td>

<p>How many bases to extract upstream of each TSS (transcription start site).
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments, for use in specific methods.
</p>
</td></tr>
<tr valign="top"><td><code>exclude.seqlevels</code></td>
<td>

<p>A character vector containing the chromosome names (a.k.a. sequence levels)
to exclude when the genes are obtained from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object containing one upstream sequence
per gene (or per transcript if <code>genes</code> is a
<a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> object containing transcript ranges).
</p>
<p>More precisely, if <code>genes</code> is a <a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a>
object, the returned object is <em>parallel</em> to it, that is, the i-th
element in the returned object is the upstream sequence corresponding to
the i-th gene (or transcript) in <code>genes</code>. Also the names on the
<a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> object are propagated to the returned
object.
</p>
<p>If <code>genes</code> is a <a href="TxDb-class.html">TxDb</a> object, the names on the returned
object are the gene IDs found in the <a href="TxDb-class.html">TxDb</a> object. To see the
type of gene IDs (i.e. Entrez gene ID or Ensembl gene ID or ...), you can
display <code>genes</code> with <code>show(genes)</code>.
</p>
<p>In addition, the returned object has the following metadata columns
(accessible with <code><a href="../../S4Vectors/html/Vector-class.html">mcols</a></code>) that provide some information about
the gene (or transcript) corresponding to each upstream sequence:
</p>

<ul>
<li> <p><code>gene_seqnames</code>: the chromosome name of the gene (or
transcript);
</p>
</li>
<li> <p><code>gene_strand</code>: the strand of the gene (or transcript);
</p>
</li>
<li> <p><code>gene_TSS</code>: the transcription start site of the gene (or
transcript).
</p>
</li></ul>



<h3>Note</h3>

<p>IMPORTANT: Always make sure to use a TxDb package (or <a href="TxDb-class.html">TxDb</a>
object) that contains a gene model compatible with the <em>genome sequence
container</em> <code>x</code>, that is, a gene model based on the exact same reference
genome as <code>x</code>.
</p>
<p>See
<a href="http://bioconductor.org/packages/release/BiocViews.html#___TxDb">http://bioconductor.org/packages/release/BiocViews.html#___TxDb</a>
for the list of TxDb packages available in the current release of
Bioconductor.
Note that you can make your own custom <a href="TxDb-class.html">TxDb</a> object from
various annotation resources by using one of the <code>makeTxDbFrom*()</code>
functions listed in the &quot;See also&quot; section below.
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="makeTxDbFromUCSC.html">makeTxDbFromUCSC</a></code>, <code><a href="makeTxDbFromBiomart.html">makeTxDbFromBiomart</a></code>,
and <code><a href="makeTxDbFromEnsembl.html">makeTxDbFromEnsembl</a></code>, for making a <a href="TxDb-class.html">TxDb</a>
object from online resources.
</p>
</li>
<li> <p><code><a href="makeTxDbFromGRanges.html">makeTxDbFromGRanges</a></code> and <code><a href="makeTxDbFromGFF.html">makeTxDbFromGFF</a></code>
for making a <a href="TxDb-class.html">TxDb</a> object from a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a>
object, or from a GFF or GTF file.
</p>
</li>
<li><p> The <code><a href="../../BSgenome/html/available.genomes.html">available.genomes</a></code> function in the
<span class="pkg">BSgenome</span> package for checking avaibility of BSgenome
data packages (and installing the desired one).
</p>
</li>
<li><p> The <a href="../../BSgenome/html/BSgenome.html">BSgenome</a>, <a href="../../rtracklayer/html/TwoBitFile.html">TwoBitFile</a>, and
<a href="../../Rsamtools/html/FaFile.html">FaFile</a> classes, defined and documented
in the <span class="pkg">BSgenome</span>, <span class="pkg">rtracklayer</span>, and <span class="pkg">Rsamtools</span>
packages, respectively.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li>
<li><p> The <code><a href="transcripts.html">genes</a></code> function for extracting gene ranges from
a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GenomicRanges.html">GenomicRanges</a> class defined and documented
in the <span class="pkg">GenomicRanges</span> package.
</p>
</li>
<li><p> The <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> class defined and documented
in the <span class="pkg">Biostrings</span> package.
</p>
</li>
<li><p> The <code><a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a></code> getter defined and documented
in the <span class="pkg">GenomeInfoDb</span> package.
</p>
</li>
<li><p> The <code><a href="../../Biostrings/html/getSeq.html">getSeq</a></code> function for extracting
subsequences from a sequence container.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## Load a genome:
library(BSgenome.Dmelanogaster.UCSC.dm3)
genome &lt;- BSgenome.Dmelanogaster.UCSC.dm3
genome

## Use a TxDb object:
library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
txdb &lt;- TxDb.Dmelanogaster.UCSC.dm3.ensGene
txdb  # contains Ensembl gene IDs

## Because the chrU and chrUextra sequences are made of concatenated
## scaffolds (see http://genome.ucsc.edu/cgi-bin/hgGateway?db=dm3),
## extracting the upstream sequences for genes located on these
## scaffolds is not reliable. So we exclude them:
exclude &lt;- c("chrU", "chrUextra")
up1000seqs &lt;- extractUpstreamSeqs(genome, txdb, width=1000,
                                  exclude.seqlevels=exclude)
up1000seqs  # the names are Ensembl gene IDs
mcols(up1000seqs)

## Upstream sequences for genes close to the chromosome bounds can be
## shorter than 1000 (note that this does not happen for circular
## chromosomes like chrM):
table(width(up1000seqs))
mcols(up1000seqs)[width(up1000seqs) != 1000, ]
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
