<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract transcript (or CDS) sequences from chromosome...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for extractTranscriptSeqs {GenomicFeatures}"><tr><td>extractTranscriptSeqs {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Extract transcript (or CDS) sequences from chromosome sequences</h2>

<h3>Description</h3>

<p><code>extractTranscriptSeqs</code> extracts transcript (or CDS) sequences from
an object representing a single chromosome or a collection of chromosomes.
</p>


<h3>Usage</h3>

<pre>
extractTranscriptSeqs(x, transcripts, ...)

## S4 method for signature 'DNAString'
extractTranscriptSeqs(x, transcripts, strand="+")

## S4 method for signature 'ANY'
extractTranscriptSeqs(x, transcripts, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>An object representing a single chromosome or a collection of chromosomes.
More precisely, <code>x</code> can be a <a href="../../Biostrings/html/DNAString.html">DNAString</a> object
(single chromosome), or a <a href="../../BSgenome/html/BSgenome.html">BSgenome</a> object (collection
of chromosomes).
</p>
<p>Other objects representing a collection of chromosomes are supported
(e.g. <a href="../../Rsamtools/html/FaFile.html">FaFile</a> objects in the <span class="pkg">Rsamtools</span> package)
as long as <code><a href="../../GenomeInfoDb/html/seqinfo.html">seqinfo</a></code> and
<code><a href="../../Biostrings/html/getSeq.html">getSeq</a></code> work on them.
</p>
</td></tr>
<tr valign="top"><td><code>transcripts</code></td>
<td>

<p>An object representing the exon ranges of each transcript to extract.
</p>
<p>More precisely:
</p>

<ul>
<li><p> If <code>x</code> is a <a href="../../Biostrings/html/DNAString.html">DNAString</a> object, then
<code>transcripts</code> must be an <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a>
object.
</p>
</li>
<li><p> If <code>x</code> is a <a href="../../BSgenome/html/BSgenome.html">BSgenome</a> object or any object
representing a collection of chromosomes, then <code>transcripts</code>
must be a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object or any object
for which <code><a href="transcriptsBy.html">exonsBy</a></code> is implemented (e.g. a <a href="TxDb-class.html">TxDb</a>
or <a href="../../ensembldb/html/EnsDb.html">EnsDb</a> object). If the latter, then it's
first turned into a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object
with <code><a href="transcriptsBy.html">exonsBy</a>(transcripts, by="tx", ...)</code>.
</p>
</li></ul>

<p>Note that, for each transcript, the exons must be ordered by ascending
<em>rank</em>, that is, by ascending position <em>in the transcript</em>
(when going in the 5' to 3' direction). This generally means (but not
always) that they are also ordered from 5' to 3' on the reference genome.
More precisely:
</p>

<ul>
<li><p> For a transcript located on the plus strand, the exons will
typically (but not necessarily) be ordered by ascending position
on the reference genome.
</p>
</li>
<li><p> For a transcript located on the minus strand, the exons will
typically (but not necessarily) be ordered by descending position
on the reference genome.
</p>
</li></ul>

<p>If <code>transcripts</code> was obtained with <code><a href="transcriptsBy.html">exonsBy</a></code> (see above),
then the exons are guaranteed to be ordered by ascending rank. See
<code>?<a href="transcriptsBy.html">exonsBy</a></code> for more information.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>Additional arguments, for use in specific methods.
</p>
<p>For the default method, additional arguments are allowed only when
<code>transcripts</code> is not a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object,
in which case they are passed to the internal call to <code><a href="transcriptsBy.html">exonsBy</a></code>
(see above).
</p>
</td></tr>
<tr valign="top"><td><code>strand</code></td>
<td>

<p>Only supported when <code>x</code> is a <a href="../../Biostrings/html/DNAString.html">DNAString</a> object.
</p>
<p>Can be an atomic vector, a factor, or an <a href="../../S4Vectors/html/Rle.html">Rle</a> object,
in which case it indicates the strand of each transcript (i.e. all the
exons in a transcript are considered to be on the same strand).
More precisely: it's turned into a factor (or factor-<a href="../../S4Vectors/html/Rle.html">Rle</a>)
that has the &quot;standard strand levels&quot; (this is done by calling the
<code><a href="../../BiocGenerics/html/strand.html">strand</a></code> function on it). Then it's recycled
to the length of <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object <code>transcripts</code>
if needed. In the resulting object, the i-th element is interpreted
as the strand of all the exons in the i-th transcript.
</p>
<p><code>strand</code> can also be a list-like object, in which case it indicates
the strand of each exon, individually. Thus it must have the same
<em>shape</em> as <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object <code>transcripts</code>
(i.e. same length plus <code>strand[[i]]</code> must have the same length
as <code>transcripts[[i]]</code> for all <code>i</code>).
</p>
<p><code>strand</code> can only contain <code>"+"</code> and/or <code>"-"</code> values.
<code>"*"</code> is not allowed.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> object <em>parallel</em> to
<code>transcripts</code>, that is, the i-th element in it is the sequence
of the i-th transcript in <code>transcripts</code>.
</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="coverageByTranscript.html">coverageByTranscript</a></code> for computing coverage by
transcript (or CDS) of a set of ranges.
</p>
</li>
<li> <p><code><a href="transcriptLengths.html">transcriptLengths</a></code> for extracting the transcript
lengths (and other metrics) from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li><p> The <code><a href="transcriptLocs2refLocs.html">transcriptLocs2refLocs</a></code> function for converting
transcript-based locations into reference-based locations.
</p>
</li>
<li><p> The <code><a href="../../BSgenome/html/available.genomes.html">available.genomes</a></code> function in the
<span class="pkg">BSgenome</span> package for checking avaibility of BSgenome
data packages (and installing the desired one).
</p>
</li>
<li><p> The <a href="../../Biostrings/html/DNAString.html">DNAString</a> and <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a>
classes defined and documented in the <span class="pkg">Biostrings</span> package.
</p>
</li>
<li><p> The <code><a href="../../Biostrings/html/translate.html">translate</a></code> function in the
<span class="pkg">Biostrings</span> package for translating DNA or RNA sequences
into amino acid sequences.
</p>
</li>
<li><p> The <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> class defined and documented
in the <span class="pkg">GenomicRanges</span> package.
</p>
</li>
<li><p> The <a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> class defined and documented
in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li><p> The <code><a href="transcriptsBy.html">exonsBy</a></code> function for extracting exon ranges
grouped by transcript.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## 1. A TOY EXAMPLE
## ---------------------------------------------------------------------

library(Biostrings)

## A chromosome of length 30:
x &lt;- DNAString("ATTTAGGACACTCCCTGAGGACAAGACCCC")

## 2 transcripts on 'x':
tx1 &lt;- IRanges(1, 8)            # 1 exon
tx2 &lt;- c(tx1, IRanges(12, 30))  # 2 exons
transcripts &lt;- IRangesList(tx1=tx1, tx2=tx2)
extractTranscriptSeqs(x, transcripts)

## By default, all the exons are considered to be on the plus strand.
## We can use the 'strand' argument to tell extractTranscriptSeqs()
## to extract them from the minus strand.

## Extract all the exons from the minus strand:
extractTranscriptSeqs(x, transcripts, strand="-")

## Note that, for a transcript located on the minus strand, the exons
## should typically be ordered by descending position on the reference
## genome in order to reflect their rank in the transcript:
extractTranscriptSeqs(x, IRangesList(tx1=tx1, tx2=rev(tx2)), strand="-")

## Extract the exon of the 1st transcript from the minus strand:
extractTranscriptSeqs(x, transcripts, strand=c("-", "+"))

## Extract the 2nd exon of the 2nd transcript from the minus strand:
extractTranscriptSeqs(x, transcripts, strand=list("-", c("+", "-")))

## ---------------------------------------------------------------------
## 2. A REAL EXAMPLE
## ---------------------------------------------------------------------

## Load a genome:
library(BSgenome.Hsapiens.UCSC.hg19)
genome &lt;- BSgenome.Hsapiens.UCSC.hg19

## Load a TxDb object:
txdb_file &lt;- system.file("extdata", "hg19_knownGene_sample.sqlite",
                         package="GenomicFeatures")
txdb &lt;- loadDb(txdb_file)

## Check that 'txdb' is based on the hg19 assembly:
txdb

## Extract the exon ranges grouped by transcript from 'txdb':
transcripts &lt;- exonsBy(txdb, by="tx", use.names=TRUE)

## Extract the transcript sequences from the genome:
tx_seqs &lt;- extractTranscriptSeqs(genome, transcripts)
tx_seqs

## A sanity check:
stopifnot(identical(width(tx_seqs), unname(sum(width(transcripts)))))

## Note that 'tx_seqs' can also be obtained with:
extractTranscriptSeqs(genome, txdb, use.names=TRUE)

## ---------------------------------------------------------------------
## 3. USING extractTranscriptSeqs() TO EXTRACT CDS SEQUENCES
## ---------------------------------------------------------------------

cds &lt;- cdsBy(txdb, by="tx", use.names=TRUE)
cds_seqs &lt;- extractTranscriptSeqs(genome, cds)
cds_seqs

## A sanity check:
stopifnot(identical(width(cds_seqs), unname(sum(width(cds)))))

## Note that, alternatively, the CDS sequences can be obtained from the
## transcript sequences by removing the 5' and 3' UTRs:
tx_lens &lt;- transcriptLengths(txdb, with.utr5_len=TRUE, with.utr3_len=TRUE)
stopifnot(identical(tx_lens$tx_name, names(tx_seqs)))  # sanity
## Keep the rows in 'tx_lens' that correspond to a sequence in 'cds_seqs'
## and put them in the same order as in 'cds_seqs':
m &lt;- match(names(cds_seqs), names(tx_seqs))
tx_lens &lt;- tx_lens[m, ]
utr5_width &lt;- tx_lens$utr5_len
utr3_width &lt;- tx_lens$utr3_len
cds_seqs2 &lt;- narrow(tx_seqs[m],
                    start=utr5_width+1L, end=-(utr3_width+1L))
stopifnot(identical(as.character(cds_seqs2), as.character(cds_seqs)))

## ---------------------------------------------------------------------
## 4. TRANSLATE THE CDS SEQUENCES
## ---------------------------------------------------------------------

prot_seqs &lt;- translate(cds_seqs, if.fuzzy.codon="solve")

## Note that, by default, translate() uses The Standard Genetic Code to
## translate codons into amino acids. However, depending on the organism,
## a different genetic code might be needed to translate CDS sequences
## located on the mitochodrial chromosome. For example, for vertebrates,
## the following code could be used to correct 'prot_seqs':
SGC1 &lt;- getGeneticCode("SGC1")
chrM_idx &lt;- which(all(seqnames(cds) == "chrM"))
prot_seqs[chrM_idx] &lt;- translate(cds_seqs[chrM_idx], genetic.code=SGC1,
                                 if.fuzzy.codon="solve")
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
