<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Get gene promoter sequences</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for getPromoterSeq {GenomicFeatures}"><tr><td>getPromoterSeq {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Get gene promoter sequences</h2>

<h3>Description</h3>

<p>Extract sequences for the genes or transcripts specified in the query
(a<a href="../../GenomicRanges/html/GRanges-class.html">GRanges</a> or <a href="../../GenomicRanges/html/GRangesList-class.html">GRangesList</a> object)
from a <a href="../../BSgenome/html/BSgenome.html">BSgenome</a> object or an <a href="../../Rsamtools/html/FaFile.html">FaFile</a>.
</p>


<h3>Usage</h3>

<pre>
  ## S4 method for signature 'GRangesList'
getPromoterSeq(query, subject, upstream=2000, downstream=200, ...)
  ## S4 method for signature 'GRangesList'
getPromoterSeq(query, subject, upstream=2000, downstream=200, ...)
  ## S4 method for signature 'GRanges'
getPromoterSeq(query, subject, upstream=2000, downstream=200, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>query</code></td>
<td>
<p>A <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> or <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object containing
genes grouped by transcript.
</p>
</td></tr>
<tr valign="top"><td><code>subject</code></td>
<td>
<p>A <a href="../../BSgenome/html/BSgenome.html">BSgenome</a> object or a <a href="../../Rsamtools/html/FaFile.html">FaFile</a> from which
the sequences will be taken.</p>
</td></tr>
<tr valign="top"><td><code>upstream</code></td>
<td>
<p>The number of DNA bases to include upstream of the TSS (transcription start site)</p>
</td></tr>
<tr valign="top"><td><code>downstream</code></td>
<td>
<p>The number of DNA bases to include downstream of the TSS (transcription start site)</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>getPromoterSeq</code> is an overloaded method dispatching on query,
which is either a GRanges or a GRangesList.   It is a wrapper for the
<code>promoters</code> and <code>getSeq</code> functions. The 
purpose is to allow sequence extraction from either a
<a href="../../BSgenome/html/BSgenome.html">BSgenome</a> or <a href="../../Rsamtools/html/FaFile.html">FaFile</a>.
</p>
<p>Default values for <code>upstream</code> and <code>downstream</code> were chosen based
on our current understanding of gene regulation. On average, promoter
regions in the mammalian genome are 5000 bp upstream and downstream of the
transcription start site.
</p>


<h3>Value</h3>

<p>A <a href="../../Biostrings/html/DNAStringSet.html">DNAStringSet</a> or
<a href="../../Biostrings/html/DNAStringSetList.html">DNAStringSetList</a> instance corresponding to the
GRanges or GRangesList supplied in the query.
</p>


<h3>Author(s)</h3>

<p>Paul Shannon</p>


<h3>See Also</h3>

<p><a href="../../IRanges/html/intra-range-methods.html">intra-range-methods</a>        ## promoters method for IntegerRanges objects
<a href="../../GenomicRanges/html/intra-range-methods.html">intra-range-methods</a>  ## promoters method for GenomicRanges objects
<a href="../../Biostrings/html/getSeq.html">getSeq</a>
</p>


<h3>Examples</h3>

<pre>
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(BSgenome.Hsapiens.UCSC.hg19)
 
  e2f3 &lt;- "1871"  # entrez geneID for a cell cycle control transcription
                  # factor, chr6 on the plus strand

  transcriptCoordsByGene.GRangesList &lt;-
     transcriptsBy (TxDb.Hsapiens.UCSC.hg19.knownGene, by = "gene") [e2f3]
    # a GrangesList of length one, describing three transcripts

  promoter.seqs &lt;- getPromoterSeq (transcriptCoordsByGene.GRangesList,
                                   Hsapiens, upstream=10, downstream=0)
    # DNAStringSetList of length 1
    # [["1871"]] GCTTCCTGGA GCTTCCTGGA CGGAGCCAGG

</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
