<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Making a FeatureDb object from annotations available at the...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeFeatureDbFromUCSC {GenomicFeatures}"><tr><td>makeFeatureDbFromUCSC {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Making a FeatureDb object from annotations available at the
UCSC Genome Browser
</h2>

<h3>Description</h3>

<p>The <code>makeFeatureDbFromUCSC</code> function allows the user
to make a <a href="FeatureDb-class.html">FeatureDb</a> object from simple annotation tracks at
UCSC.  The tracks in question must (at a minimum) have a start, end and
a chromosome affiliation in order to be made into a <a href="FeatureDb-class.html">FeatureDb</a>.
This function requires a precise declaration of its first three
arguments to indicate which genome, track and table wish to be
imported.  There are discovery functions provided to make this process
go smoothly.
</p>


<h3>Usage</h3>

<pre>
supportedUCSCFeatureDbTracks(genome)

supportedUCSCFeatureDbTables(genome, track)
 
UCSCFeatureDbTableSchema(genome,
                         track,
                         tablename)

makeFeatureDbFromUCSC(
        genome,
        track,
        tablename,
        columns = UCSCFeatureDbTableSchema(genome,track,tablename),
        url="http://genome.ucsc.edu/cgi-bin/",
        goldenPath.url=getOption("UCSC.goldenPath.url"),
        chromCol, 
        chromStartCol, 
        chromEndCol,
        taxonomyId=NA)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>genome</code></td>
<td>
<p>genome abbreviation used by UCSC and obtained by
<code><a href="../../rtracklayer/html/ucscGenomes.html">ucscGenomes</a>()[ , "db"]</code>.
For example: <code>"hg18"</code>.</p>
</td></tr>
<tr valign="top"><td><code>track</code></td>
<td>
<p>name of the UCSC track.  Use
<code>supportedUCSCFeatureDbTracks</code> to get the list of available
tracks for a particular genome</p>
</td></tr>
<tr valign="top"><td><code>tablename</code></td>
<td>
<p>name of the UCSC table containing the annotations to
retrieve. Use the <code>supportedUCSCFeatureDbTables</code> utility
function to get the list of supported tables for a track.</p>
</td></tr>
<tr valign="top"><td><code>columns</code></td>
<td>
<p>a named character vector to list out the names and
types of the other columns that the downloaded track should
have.  Use <code>UCSCFeatureDbTableSchema</code> to retrieve this
information for a particular table.</p>
</td></tr>
<tr valign="top"><td><code>url,goldenPath.url</code></td>
<td>
<p>use to specify the location of an
alternate UCSC Genome Browser.</p>
</td></tr>
<tr valign="top"><td><code>chromCol</code></td>
<td>
<p>If the schema comes back and the 'chrom' column has been 
labeled something other than 'chrom', use this argument to indicate 
what that column has been labeled as so we can properly designate it.
This could happen (for example) with the knownGene track tables, which
has no 'chromStart' or 'chromEnd' columns, but which DOES have columns
that could reasonably substitute for these columns under particular 
circumstances.  Therefore we allow these three columns to have arguments
so that their definition can be re-specified</p>
</td></tr>
<tr valign="top"><td><code>chromStartCol</code></td>
<td>
<p>Same thing as chromCol, but for renames of 'chromStart'</p>
</td></tr>
<tr valign="top"><td><code>chromEndCol</code></td>
<td>
<p>Same thing as chromCol, but for renames of 'chromEnd'</p>
</td></tr>
<tr valign="top"><td><code>taxonomyId</code></td>
<td>
<p>By default this value is NA and the organism
inferred will be used to look up the correct value for this.  But
you can use this argument to override that and supply your own
valid taxId here.</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>makeFeatureDbFromUCSC</code> is a convenience function that builds a
tiny database from one of the UCSC track tables.
<code>supportedUCSCFeatureDbTracks</code> a convenience function that
returns potential track names that could be used to make
FeatureDb objects
<code>supportedUCSCFeatureDbTables</code> a convenience function that
returns potential table names for FeatureDb objects (table names
go with a track name)
<code>UCSCFeatureDbTableSchema</code> A convenience function that creates a
named vector of types for all the fields that can potentially be
supported for a given track.  By default, this will be called on
your specified tablename to include all of the fields in a track.
</p>






<h3>Value</h3>

<p>A <a href="FeatureDb-class.html">FeatureDb</a> object for <code>makeFeatureDbFromUCSC</code>.  Or in
the case of <code>supportedUCSCFeatureDbTracks</code> and
<code>UCSCFeatureDbTableSchema</code> a named character vector</p>


<h3>Author(s)</h3>

<p>M. Carlson</p>


<h3>See Also</h3>

<p><code><a href="../../rtracklayer/html/ucscGenomes.html">ucscGenomes</a></code>,


</p>


<h3>Examples</h3>

<pre>
## Display the list of genomes available at UCSC:
library(GenomicFeatures)
library(rtracklayer)
ucscGenomes()[ , "db"]

## Display the list of Tracks supported by makeFeatureDbFromUCSC():
# supportedUCSCFeatureDbTracks("mm10")

## Display the list of tables supported by your track:
supportedUCSCFeatureDbTables(genome="mm10",
                             track="qPCR Primers")

## Display fields that could be passed in to colnames:
UCSCFeatureDbTableSchema(genome="mm10",
                         track="qPCR Primers",
                         tablename="qPcrPrimers")

## Retrieving a full transcript dataset for Mouse from UCSC:
fdb &lt;- makeFeatureDbFromUCSC(genome="mm10",
                              track="qPCR Primers",
                              tablename="qPcrPrimers")
fdb
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
