<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Make a TxDb object from an Ensembl database</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeTxDbFromEnsembl {GenomicFeatures}"><tr><td>makeTxDbFromEnsembl {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Make a TxDb object from an Ensembl database
</h2>

<h3>Description</h3>

<p>The <code>makeTxDbFromEnsembl</code> function creates a <a href="TxDb-class.html">TxDb</a> object for
a given organism by importing the genomic locations of its transcripts,
exons, CDS, and genes from an Ensembl database.
</p>
<p>Note that it uses the <span class="pkg">RMariaDB</span> package internally so make sure
that this package is installed.
</p>


<h3>Usage</h3>

<pre>
makeTxDbFromEnsembl(organism="Homo sapiens",
                    release=NA,
                    circ_seqs=NULL,
                    server="ensembldb.ensembl.org",
                    username="anonymous", password=NULL, port=0L,
                    tx_attrib=NULL)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>organism</code></td>
<td>

<p>The <em>scientific name</em> (i.e. genus and species, or genus and species
and subspecies) of the organism for which to import the data.
Case is not sensitive. Underscores can be used instead of white spaces
e.g. <code>"homo_sapiens"</code> is accepted.
</p>
</td></tr>
<tr valign="top"><td><code>release</code></td>
<td>

<p>The Ensembl release to query e.g. 89. If set to <code>NA</code> (the default),
the current release is used.
</p>
</td></tr>
<tr valign="top"><td><code>circ_seqs</code></td>
<td>

<p>A character vector to list out which chromosomes should be marked
as circular.
</p>
</td></tr>
<tr valign="top"><td><code>server</code></td>
<td>

<p>The name of the MySQL server to query.
See <a href="https://www.ensembl.org/info/data/mysql.html">https://www.ensembl.org/info/data/mysql.html</a> for the list of
Ensembl public MySQL servers.
Make sure to use the server nearest to you. It can make a big difference!
</p>
</td></tr>
<tr valign="top"><td><code>username</code></td>
<td>

<p>Login username for the MySQL server.
</p>
</td></tr>
<tr valign="top"><td><code>password</code></td>
<td>

<p>Login password for the MySQL server.
</p>
</td></tr>
<tr valign="top"><td><code>port</code></td>
<td>

<p>Port of the MySQL server.
</p>
</td></tr>
<tr valign="top"><td><code>tx_attrib</code></td>
<td>

<p>If not <code>NULL</code>, only select transcripts with an attribute of
the given code, a string, like <code>"gencode_basic"</code>.
</p>
</td></tr>
</table>


<h3>Value</h3>

<p>A <a href="TxDb-class.html">TxDb</a> object.
</p>


<h3>Note</h3>

<p><code>makeTxDbFromEnsembl</code> tends to be faster and more reliable than
<code><a href="makeTxDbFromBiomart.html">makeTxDbFromBiomart</a></code>.
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="makeTxDbFromUCSC.html">makeTxDbFromUCSC</a></code> and <code><a href="makeTxDbFromBiomart.html">makeTxDbFromBiomart</a></code>
for making a <a href="TxDb-class.html">TxDb</a> object from other online resources.
</p>
</li>
<li> <p><code><a href="makeTxDbFromGRanges.html">makeTxDbFromGRanges</a></code> and <code><a href="makeTxDbFromGFF.html">makeTxDbFromGFF</a></code>
for making a <a href="TxDb-class.html">TxDb</a> object from a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a>
object, or from a GFF or GTF file.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li>
<li> <p><code><a href="makeTxDb.html">makeTxDb</a></code> for the low-level function used by the
<code>makeTxDbFrom*</code> functions to make the <a href="TxDb-class.html">TxDb</a> object
returned to the user.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## Not run: 
txdb &lt;- makeTxDbFromEnsembl("Saccharomyces cerevisiae",
                            server="useastdb.ensembl.org")
txdb

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
