<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Making a TxDb object from user supplied annotations</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for makeTxDb {GenomicFeatures}"><tr><td>makeTxDb {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Making a TxDb object from user supplied annotations
</h2>

<h3>Description</h3>

<p><code>makeTxDb</code> is a low-level constructor for making
a <a href="TxDb-class.html">TxDb</a> object from user supplied transcript annotations.
</p>
<p>Note that the end user will rarely need to use <code>makeTxDb</code> directly
but will typically use one of the high-level constructors
<code><a href="makeTxDbFromUCSC.html">makeTxDbFromUCSC</a></code>, <code><a href="makeTxDbFromEnsembl.html">makeTxDbFromEnsembl</a></code>,
or <code><a href="makeTxDbFromGFF.html">makeTxDbFromGFF</a></code>.
</p>


<h3>Usage</h3>

<pre>
makeTxDb(transcripts, splicings, genes=NULL,
         chrominfo=NULL, metadata=NULL,
         reassign.ids=FALSE, on.foreign.transcripts=c("error", "drop"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>transcripts</code></td>
<td>

<p>Data frame containing the genomic locations of a set of transcripts.
</p>
</td></tr>
<tr valign="top"><td><code>splicings</code></td>
<td>

<p>Data frame containing the exon and CDS locations of a set of transcripts.
</p>
</td></tr>
<tr valign="top"><td><code>genes</code></td>
<td>

<p>Data frame containing the genes associated to a set of transcripts.
</p>
</td></tr>
<tr valign="top"><td><code>chrominfo</code></td>
<td>

<p>Data frame containing information about the chromosomes hosting
the set of transcripts.
</p>
</td></tr>
<tr valign="top"><td><code>metadata</code></td>
<td>

<p>2-column data frame containing meta information about this set of
transcripts like organism, genome, UCSC table, etc...
The names of the columns must be <code>"name"</code> and <code>"value"</code>
and their type must be character.
</p>
</td></tr>
<tr valign="top"><td><code>reassign.ids</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>.
Controls how internal ids should be assigned for each type of feature
i.e. for transcripts, exons, and CDS. For each type, if <code>reassign.ids</code>
is <code>FALSE</code> (the default) and if the ids are supplied, then they
are used as the internal ids, otherwise the internal ids are assigned
in a way that is compatible with the order defined by ordering the
features first by chromosome, then by strand, then by start, and
finally by end.
</p>
</td></tr>
<tr valign="top"><td><code>on.foreign.transcripts</code></td>
<td>

<p>Controls what to do when the input contains <em>foreign transcripts</em>
i.e. transcripts that are on sequences not in <code>chrominfo</code>.
If set to <code>"error"</code> (the default)
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The <code>transcripts</code> (required), <code>splicings</code> (required)
and <code>genes</code> (optional) arguments must be data frames that
describe a set of transcripts and the genomic features related
to them (exons, CDS and genes at the moment).
The <code>chrominfo</code> (optional) argument must be a data frame
containing chromosome information like the length of each chromosome.
</p>
<p><code>transcripts</code> must have 1 row per transcript and the following
columns:
</p>

<ul>
<li> <p><code>tx_id</code>: Transcript ID. Integer vector. No NAs. No duplicates.
</p>
</li>
<li> <p><code>tx_chrom</code>: Transcript chromosome. Character vector (or factor)
with no NAs.
</p>
</li>
<li> <p><code>tx_strand</code>: Transcript strand. Character vector (or factor)
with no NAs where each element is either <code>"+"</code> or <code>"-"</code>.
</p>
</li>
<li> <p><code>tx_start</code>, <code>tx_end</code>: Transcript start and end.
Integer vectors with no NAs.
</p>
</li>
<li> <p><code>tx_name</code>: [optional] Transcript name. Character vector (or
factor). NAs and/or duplicates are ok.
</p>
</li>
<li> <p><code>tx_type</code>: [optional] Transcript type (e.g. mRNA, ncRNA, snoRNA,
etc...). Character vector (or factor). NAs and/or duplicates are ok.
</p>
</li>
<li> <p><code>gene_id</code>: [optional] Associated gene. Character vector (or
factor). NAs and/or duplicates are ok.
</p>
</li></ul>

<p>Other columns, if any, are ignored (with a warning).
</p>
<p><code>splicings</code> must have N rows per transcript, where N is the nb
of exons in the transcript. Each row describes an exon plus, optionally,
the CDS contained in this exon. Its columns must be:
</p>

<ul>
<li> <p><code>tx_id</code>: Foreign key that links each row in the <code>splicings</code>
data frame to a unique row in the <code>transcripts</code> data frame.
Note that more than 1 row in <code>splicings</code> can be linked to the
same row in <code>transcripts</code> (many-to-one relationship).
Same type as <code>transcripts$tx_id</code> (integer vector). No NAs.
All the values in this column must be present in
<code>transcripts$tx_id</code>.
</p>
</li>
<li> <p><code>exon_rank</code>: The rank of the exon in the transcript.
Integer vector with no NAs. (<code>tx_id</code>, <code>exon_rank</code>)
pairs must be unique.
</p>
</li>
<li> <p><code>exon_id</code>: [optional] Exon ID.
Integer vector with no NAs.
</p>
</li>
<li> <p><code>exon_name</code>: [optional] Exon name.  Character vector (or factor).
NAs and/or duplicates are ok.
</p>
</li>
<li> <p><code>exon_chrom</code>: [optional] Exon chromosome.
Character vector (or factor) with no NAs.
If missing then <code>transcripts$tx_chrom</code> is used.
If present then <code>exon_strand</code> must also be present.
</p>
</li>
<li> <p><code>exon_strand</code>: [optional] Exon strand.
Character vector (or factor) with no NAs.
If missing then <code>transcripts$tx_strand</code> is used
and <code>exon_chrom</code> must also be missing.
</p>
</li>
<li> <p><code>exon_start</code>, <code>exon_end</code>: Exon start and end.
Integer vectors with no NAs.
</p>
</li>
<li> <p><code>cds_id</code>: [optional] CDS ID. Integer vector.
If present then <code>cds_start</code> and <code>cds_end</code> must also
be present.
NAs are allowed and must match those in <code>cds_start</code> and
<code>cds_end</code>.
</p>
</li>
<li> <p><code>cds_name</code>: [optional] CDS name. Character vector (or factor).
If present then <code>cds_start</code> and <code>cds_end</code> must also be
present. NAs and/or duplicates are ok. Must contain NAs at least
where <code>cds_start</code> and <code>cds_end</code> contain them.
</p>
</li>
<li> <p><code>cds_start</code>, <code>cds_end</code>: [optional] CDS start and end.
Integer vectors.
If one of the 2 columns is missing then all <code>cds_*</code> columns
must be missing.
NAs are allowed and must occur at the same positions in
<code>cds_start</code> and <code>cds_end</code>.
</p>
</li>
<li> <p><code>cds_phase</code>: [optional] CDS phase. Integer vector.
If present then <code>cds_start</code> and <code>cds_end</code> must also
be present.
NAs are allowed and must match those in <code>cds_start</code> and
<code>cds_end</code>.
</p>
</li></ul>

<p>Other columns, if any, are ignored (with a warning).
</p>
<p><code>genes</code> should not be supplied if <code>transcripts</code> has a
<code>gene_id</code> column. If supplied, it must have N rows per transcript,
where N is the nb of genes linked to the transcript (N will be 1 most
of the time). Its columns must be:
</p>

<ul>
<li> <p><code>tx_id</code>: [optional] <code>genes</code> must have either a
<code>tx_id</code> or a <code>tx_name</code> column but not both.
Like <code>splicings$tx_id</code>, this is a foreign key that
links each row in the <code>genes</code> data frame to a unique
row in the <code>transcripts</code> data frame.
</p>
</li>
<li> <p><code>tx_name</code>: [optional]
Can be used as an alternative to the <code>genes$tx_id</code>
foreign key.
</p>
</li>
<li> <p><code>gene_id</code>: Gene ID. Character vector (or factor). No NAs.
</p>
</li></ul>

<p>Other columns, if any, are ignored (with a warning).
</p>
<p><code>chrominfo</code> must have 1 row per chromosome and the following
columns:
</p>

<ul>
<li> <p><code>chrom</code>: Chromosome name.
Character vector (or factor) with no NAs and no duplicates.
</p>
</li>
<li> <p><code>length</code>: Chromosome length.
Integer vector with either all NAs or no NAs.
</p>
</li>
<li> <p><code>is_circular</code>: [optional] Chromosome circularity flag.
Logical vector. NAs are ok.
</p>
</li></ul>

<p>Other columns, if any, are ignored (with a warning).
</p>


<h3>Value</h3>

<p>A <a href="TxDb-class.html">TxDb</a> object.</p>


<h3>Author(s)</h3>

<p>Hervé Pagès</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="makeTxDbFromUCSC.html">makeTxDbFromUCSC</a></code>, <code><a href="makeTxDbFromBiomart.html">makeTxDbFromBiomart</a></code>,
and <code><a href="makeTxDbFromEnsembl.html">makeTxDbFromEnsembl</a></code>, for making a <a href="TxDb-class.html">TxDb</a>
object from online resources.
</p>
</li>
<li> <p><code><a href="makeTxDbFromGRanges.html">makeTxDbFromGRanges</a></code> and <code><a href="makeTxDbFromGFF.html">makeTxDbFromGFF</a></code>
for making a <a href="TxDb-class.html">TxDb</a> object from a <a href="../../GenomicRanges/html/GRanges.html">GRanges</a>
object, or from a GFF or GTF file.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li>
<li> <p><code><a href="../../AnnotationDbi/html/saveDb.html">saveDb</a></code> and
<code><a href="../../AnnotationDbi/html/loadDb.html">loadDb</a></code> in the <span class="pkg">AnnotationDbi</span>
package for saving and loading a <a href="TxDb-class.html">TxDb</a> object as an SQLite
file.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
transcripts &lt;- data.frame(
                   tx_id=1:3,
                   tx_chrom="chr1",
                   tx_strand=c("-", "+", "+"),
                   tx_start=c(1, 2001, 2001),
                   tx_end=c(999, 2199, 2199))
splicings &lt;-  data.frame(
                   tx_id=c(1L, 2L, 2L, 2L, 3L, 3L),
                   exon_rank=c(1, 1, 2, 3, 1, 2),
                   exon_start=c(1, 2001, 2101, 2131, 2001, 2131),
                   exon_end=c(999, 2085, 2144, 2199, 2085, 2199),
                   cds_start=c(1, 2022, 2101, 2131, NA, NA),
                   cds_end=c(999, 2085, 2144, 2193, NA, NA),
                   cds_phase=c(0, 0, 2, 0, NA, NA))

txdb &lt;- makeTxDb(transcripts, splicings)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
