<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Finding the nearest genomic range neighbor in a TxDb</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for nearest-methods {GenomicFeatures}"><tr><td>nearest-methods {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Finding the nearest genomic range neighbor in a TxDb</h2>

<h3>Description</h3>

<p>The <code>distance</code> methods for TxDb objects and subclasses.
</p>


<h3>Usage</h3>

<pre>
## S4 method for signature 'GenomicRanges,TxDb'
distance(x, y, ignore.strand=FALSE,
    ..., id, type=c("gene", "tx", "exon", "cds"))
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>The query <a href="../../GenomicRanges/html/GRanges-class.html">GenomicRanges</a> instance.
</p>
</td></tr>
<tr valign="top"><td><code>y</code></td>
<td>
<p>For <code>distance</code>, a <a href="TxDb-class.html">TxDb</a> instance. The
<code>id</code> is used to extract ranges from the <a href="TxDb-class.html">TxDb</a>
which are then used to compute the distance from <code>x</code>.
</p>
</td></tr>
<tr valign="top"><td><code>id</code></td>
<td>
<p>A <code>character</code> vector the same length as <code>x</code>.
The <code>id</code> must be identifiers in the <a href="TxDb-class.html">TxDb</a> object.
<code>type</code> indicates what type of identifier <code>id</code> is. 
</p>
</td></tr>
<tr valign="top"><td><code>type</code></td>
<td>
<p>A <code>character(1)</code> describing the <code>id</code>. 
Must be one of &lsquo;gene&rsquo;, &lsquo;tx&rsquo;, &lsquo;exon&rsquo; or
&lsquo;cds&rsquo;.
</p>
</td></tr>
<tr valign="top"><td><code>ignore.strand</code></td>
<td>

<p>A <code>logical</code> indicating if the strand of the ranges
should be ignored. When <code>TRUE</code>, strand is set to <code>'+'</code>.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Additional arguments for methods.
</p>
</td></tr>
</table>


<h3>Details</h3>


<ul>
<li><p>distance: 
Returns the distance for each range in <code>x</code> to the range 
extracted from the <a href="TxDb-class.html">TxDb</a> object <code>y</code>. Values in 
<code>id</code> are matched to one of &lsquo;gene_id&rsquo;, &lsquo;tx_id&rsquo;, 
&lsquo;exon_id&rsquo; or &lsquo;cds_id&rsquo; identifiers in the <a href="TxDb-class.html">TxDb</a>
and the corresponding ranges are extracted. The <code>type</code> argument 
specifies which identifier is represented in <code>id</code>. The extracted 
ranges are used in the distance calculation with the ranges in 
<code>x</code>.
</p>
<p>The method returns <code>NA</code> values when the genomic region defined
by <code>id</code> cannot be collapsed into a single range (e.g.,
when a gene spans multiple chromosomes) or if the <code>id</code>
is not found in <code>y</code>.
</p>
<p>The behavior of <code>distance</code> with respect to zero-width ranges
has changed in Bioconductor 2.12. See the man page <code>?distance</code> 
in IRanges for details.

</p>
</li></ul>



<h3>Value</h3>

<p>For <code>distance</code>, an integer vector of distances between the ranges
in <code>x</code> and <code>y</code>.
</p>


<h3>Author(s)</h3>

<p>Valerie Obenchain &lt;vobencha@fhcrc.org&gt;</p>


<h3>See Also</h3>


<ul>
<li> <p><a href="../../IRanges/html/nearest-methods.html">nearest-methods</a> man page in IRanges.
</p>
</li>
<li> <p><a href="../../GenomicRanges/html/nearest-methods.html">nearest-methods</a> man page in GenomicRanges.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
  ## -----------------------------------------------------------
  ## distance()
  ## -----------------------------------------------------------

  library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
  txdb &lt;- TxDb.Dmelanogaster.UCSC.dm3.ensGene
  gr &lt;- GRanges(c("chr2L", "chr2R"),
                IRanges(c(100000, 200000),  width=100))
  distance(gr, txdb, id=c("FBgn0259717", "FBgn0261501"), type="gene")
  distance(gr, txdb, id=c("10000", "23000"), type="cds")

  ## The id's must be in the appropriate order with respect to 'x'.
  distance(gr, txdb, id=c("4", "4097"), type="tx")

  ## 'id' "4" is on chr2L and "4097" is on chr2R. 
  transcripts(txdb, filter=list(tx_id=c("4", "4097")))

  ## If we reverse the 'id' the chromosomes are incompatable with gr.
  distance(gr, txdb, id=c("4097", "4"), type="tx")

  ## distance() compares each 'x' to the corresponding 'y'.
  ## If an 'id' is not found in the TxDb 'y' will not 
  ## be the same lenth as 'x' and an error is thrown.
  ## Not run:  
  distance(gr, txdb, id=c("FBgn0000008", "INVALID"), type="gene") ## will fail
  
## End(Not run) 
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
