<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract genomic features from a TxDb-like object</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for transcripts {GenomicFeatures}"><tr><td>transcripts {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Extract genomic features from a TxDb-like object
</h2>

<h3>Description</h3>

<p>Generic functions to extract genomic features from a TxDb-like object.
This page documents the methods for <a href="TxDb-class.html">TxDb</a> objects only.
</p>


<h3>Usage</h3>

<pre>
transcripts(x, ...)
## S4 method for signature 'TxDb'
transcripts(x, columns=c("tx_id", "tx_name"), filter=NULL, use.names=FALSE)

exons(x, ...)
## S4 method for signature 'TxDb'
exons(x, columns="exon_id", filter=NULL, use.names=FALSE)

cds(x, ...)
## S4 method for signature 'TxDb'
cds(x, columns="cds_id", filter=NULL, use.names=FALSE)

genes(x, ...)
## S4 method for signature 'TxDb'
genes(x, columns="gene_id", filter=NULL, single.strand.genes.only=TRUE)

## S4 method for signature 'TxDb'
promoters(x, upstream=2000, downstream=200, use.names=TRUE, ...)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <a href="TxDb-class.html">TxDb</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>

<p>For the <code>transcripts</code>, <code>exons</code>, <code>cds</code>, and <code>genes</code>
generic functions: arguments to be passed to methods.
</p>
<p>For the <code>promoters</code> method for <a href="TxDb-class.html">TxDb</a> objects: arguments
to be passed to the internal call to <code>transcripts</code>.
</p>
</td></tr>
<tr valign="top"><td><code>columns</code></td>
<td>

<p>Columns to include in the output.
Must be <code>NULL</code> or a character vector as given by the
<code>columns</code> method. With the following restrictions:
</p>

<ul>
<li> <p><code>"TXCHROM"</code> and <code>"TXSTRAND"</code> are not allowed
for <code>transcripts</code>.
</p>
</li>
<li> <p><code>"EXONCHROM"</code> and <code>"EXONSTRAND"</code> are not allowed
for <code>exons</code>.
</p>
</li>
<li> <p><code>"CDSCHROM"</code> and <code>"CDSSTRAND"</code> are not allowed
for <code>cds</code>.
</p>
</li></ul>

<p>If the vector is named, those names are used for the corresponding
column in the element metadata of the returned object.
</p>
</td></tr>
<tr valign="top"><td><code>filter</code></td>
<td>

<p>Either <code>NULL</code> or a named list of vectors to be used to
restrict the output. Valid names for this list are:
<code>"gene_id"</code>,
<code>"tx_id"</code>, <code>"tx_name"</code>,
<code>"tx_chrom"</code>, <code>"tx_strand"</code>,
<code>"exon_id"</code>, <code>"exon_name"</code>,
<code>"exon_chrom"</code>, <code>"exon_strand"</code>,
<code>"cds_id"</code>, <code>"cds_name"</code>,
<code>"cds_chrom"</code>, <code>"cds_strand"</code> and <code>"exon_rank"</code>.
</p>
</td></tr>
<tr valign="top"><td><code>use.names</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>. If <code>TRUE</code>, the feature names are
set as the names of the returned object, with NAs being replaced with
empty strings.
</p>
</td></tr>
<tr valign="top"><td><code>single.strand.genes.only</code></td>
<td>

<p><code>TRUE</code> or <code>FALSE</code>.
If <code>TRUE</code> (the default), then genes are returned in a
<a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object and those genes that cannot be
represented by a single genomic range (because they have exons located
on both strands of the same reference sequence or on more than one
reference sequence) are dropped with a message.
</p>
<p>If <code>FALSE</code>, then all the genes are returned in a
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object with the columns specified
thru the <code>columns</code> argument set as <em>top level</em> metadata
columns. (Please keep in mind that the <em>top level</em> metadata
columns of a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object are not displayed
by the <code>show()</code> method.)
</p>
</td></tr>
<tr valign="top"><td><code>upstream</code></td>
<td>

<p>For <code>promoters</code> : An <code>integer(1)</code> value indicating the
number of bases upstream from the transcription start site. For
additional details see <code>?`promoters,GRanges-method`</code>.
</p>
</td></tr>
<tr valign="top"><td><code>downstream</code></td>
<td>

<p>For <code>promoters</code> : An <code>integer(1)</code> value indicating the
number of bases downstream from the transcription start site. For
additional details see <code>?`promoters,GRanges-method`</code>.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>These are the main functions for extracting transcript information
from a <a href="TxDb-class.html">TxDb</a>-like object. These methods can restrict the output
based on categorical information. To restrict the output based on interval
information, use the <code><a href="transcriptsByOverlaps.html">transcriptsByOverlaps</a></code>,
<code><a href="transcriptsByOverlaps.html">exonsByOverlaps</a></code>, and <code><a href="transcriptsByOverlaps.html">cdsByOverlaps</a></code>
functions.
</p>
<p>The <code>promoters</code> function computes user-defined promoter regions
for the transcripts in a <a href="TxDb-class.html">TxDb</a>-like object. The return object is
a <code>GRanges</code> of promoter regions around the transcription start
site the span of which is defined by <code>upstream</code> and <code>downstream</code>.
For additional details on how the promoter range is computed and the
handling of <code>+</code> and <code>-</code> strands see
<code>?`promoters,GRanges-method`</code>.
</p>


<h3>Value</h3>

<p>A <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object. The only exception being
when <code>genes</code> is used with <code>single.strand.genes.only=FALSE</code>,
in which case a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object is returned.
</p>


<h3>Author(s)</h3>

<p>M. Carlson, P. Aboyoun and H. Pag√®s
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="transcriptsBy.html">transcriptsBy</a></code> and <code><a href="transcriptsByOverlaps.html">transcriptsByOverlaps</a></code>
for more ways to extract genomic features
from a <a href="TxDb-class.html">TxDb</a>-like object.
</p>
</li>
<li> <p><code><a href="transcriptLengths.html">transcriptLengths</a></code> for extracting the transcript
lengths (and other metrics) from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="exonicParts.html">exonicParts</a></code> and <code><a href="exonicParts.html">intronicParts</a></code> for
extracting non-overlapping exonic or intronic parts from a
TxDb-like object.
</p>
</li>
<li> <p><code><a href="extractTranscriptSeqs.html">extractTranscriptSeqs</a></code> for extracting transcript
(or CDS) sequences from reference sequences.
</p>
</li>
<li> <p><code><a href="coverageByTranscript.html">coverageByTranscript</a></code> for computing coverage by
transcript (or CDS) of a set of ranges.
</p>
</li>
<li> <p><a href="select-methods.html">select-methods</a> for how to use the
simple &quot;select&quot; interface to extract information from a
<a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="microRNAs.html">microRNAs</a></code> and <code><a href="microRNAs.html">tRNAs</a></code> for extracting
microRNA or tRNA genomic ranges from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="id2name.html">id2name</a></code> for mapping <a href="TxDb-class.html">TxDb</a> internal ids
to external names for a given feature type.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
txdb_file &lt;- system.file("extdata", "hg19_knownGene_sample.sqlite",
                         package="GenomicFeatures")
txdb &lt;- loadDb(txdb_file)

## ---------------------------------------------------------------------
## transcripts()
## ---------------------------------------------------------------------

tx1 &lt;- transcripts(txdb)
tx1

transcripts(txdb, use.names=TRUE)
transcripts(txdb, columns=NULL, use.names=TRUE)

filter &lt;- list(tx_chrom = c("chr3", "chr5"), tx_strand = "+")
tx2 &lt;- transcripts(txdb, filter=filter)
tx2

## Sanity checks:
stopifnot(
  identical(mcols(tx1)$tx_id, seq_along(tx1)),
  identical(tx2, tx1[seqnames(tx1) == "chr3" &amp; strand(tx1) == "+"])
)

## ---------------------------------------------------------------------
## exons()
## ---------------------------------------------------------------------

exons(txdb, columns=c("EXONID", "TXNAME"),
            filter=list(exon_id=1))
exons(txdb, columns=c("EXONID", "TXNAME"),
            filter=list(tx_name="uc009vip.1"))

## ---------------------------------------------------------------------
## genes()
## ---------------------------------------------------------------------

genes(txdb)  # a GRanges object
cols &lt;- c("tx_id", "tx_chrom", "tx_strand",
          "exon_id", "exon_chrom", "exon_strand")
## By default, genes are returned in a GRanges object and those that
## cannot be represented by a single genomic range (because they have
## exons located on both strands of the same reference sequence or on
## more than one reference sequence) are dropped with a message:
single_strand_genes &lt;- genes(txdb, columns=cols)

## Because we've returned single strand genes only, the "tx_chrom"
## and "exon_chrom" metadata columns are guaranteed to match
## 'seqnames(single_strand_genes)':
stopifnot(identical(as.character(seqnames(single_strand_genes)),
                    as.character(mcols(single_strand_genes)$tx_chrom)))
stopifnot(identical(as.character(seqnames(single_strand_genes)),
                    as.character(mcols(single_strand_genes)$exon_chrom)))

## and also the "tx_strand" and "exon_strand" metadata columns are
## guaranteed to match 'strand(single_strand_genes)':
stopifnot(identical(as.character(strand(single_strand_genes)),
                    as.character(mcols(single_strand_genes)$tx_strand)))
stopifnot(identical(as.character(strand(single_strand_genes)),
                    as.character(mcols(single_strand_genes)$exon_strand)))

all_genes &lt;- genes(txdb, columns=cols, single.strand.genes.only=FALSE)
all_genes  # a GRangesList object
multiple_strand_genes &lt;- all_genes[elementNROWS(all_genes) &gt;= 2]
multiple_strand_genes
mcols(multiple_strand_genes)

## ---------------------------------------------------------------------
## promoters()
## ---------------------------------------------------------------------

## This:
promoters(txdb, upstream=100, downstream=50)
## is equivalent to:
promoters(transcripts(txdb, use.names=TRUE), upstream=100, downstream=50)

## Extra arguments are passed to transcripts(). So this:
columns &lt;- c("tx_name", "gene_id")
promoters(txdb, upstream=100, downstream=50, columns=columns)
## is equivalent to:
promoters(transcripts(txdb, columns=columns, use.names=TRUE),
          upstream=100, downstream=50)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
