<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract and group genomic features of a given type from a...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for transcriptsBy {GenomicFeatures}"><tr><td>transcriptsBy {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Extract and group genomic features of a given type from a TxDb-like object
</h2>

<h3>Description</h3>

<p>Generic functions to extract genomic features of a given type
grouped based on another type of genomic feature.
This page documents the methods for <a href="TxDb-class.html">TxDb</a> objects only.
</p>


<h3>Usage</h3>

<pre>
transcriptsBy(x, by=c("gene", "exon", "cds"), ...)
## S4 method for signature 'TxDb'
transcriptsBy(x, by=c("gene", "exon", "cds"), use.names=FALSE)

exonsBy(x, by=c("tx", "gene"), ...)
## S4 method for signature 'TxDb'
exonsBy(x, by=c("tx", "gene"), use.names=FALSE)

cdsBy(x, by=c("tx", "gene"), ...)
## S4 method for signature 'TxDb'
cdsBy(x, by=c("tx", "gene"), use.names=FALSE)

intronsByTranscript(x, ...)
## S4 method for signature 'TxDb'
intronsByTranscript(x, use.names=FALSE)

fiveUTRsByTranscript(x, ...)
## S4 method for signature 'TxDb'
fiveUTRsByTranscript(x, use.names=FALSE)

threeUTRsByTranscript(x, ...)
## S4 method for signature 'TxDb'
threeUTRsByTranscript(x, use.names=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>A <a href="TxDb-class.html">TxDb</a> object.</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Arguments to be passed to or from methods.</p>
</td></tr>
<tr valign="top"><td><code>by</code></td>
<td>
<p>One of <code>"gene"</code>, <code>"exon"</code>, <code>"cds"</code> or <code>"tx"</code>.
Determines the grouping.</p>
</td></tr>
<tr valign="top"><td><code>use.names</code></td>
<td>
<p>Controls how to set the names of the returned
<a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
These functions return all the features of a given type (e.g.
all the exons) grouped by another feature type (e.g. grouped by
transcript) in a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.
By default (i.e. if <code>use.names</code> is <code>FALSE</code>), the
names of this <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object
(aka the group names) are the internal ids of the features
used for grouping (aka the grouping features), which are
guaranteed to be unique.
If <code>use.names</code> is <code>TRUE</code>, then the names of the
grouping features are used instead of their internal ids.
For example, when grouping by transcript (<code>by="tx"</code>),
the default group names are the transcript internal ids
(<code>"tx_id"</code>). But, if <code>use.names=TRUE</code>, the group
names are the transcript names (<code>"tx_name"</code>).
Note that, unlike the feature ids, the feature names are not
guaranteed to be unique or even defined (they could be all
<code>NA</code>s). A warning is issued when this happens.
See <code>?<a href="id2name.html">id2name</a></code> for more information about
feature internal ids and feature external names and how
to map the formers to the latters.
</p>
<p>Finally, <code>use.names=TRUE</code> cannot be used when grouping
by gene <code>by="gene"</code>. This is because, unlike for the
other features, the gene ids are external ids (e.g. Entrez
Gene or Ensembl ids) so the db doesn't have a <code>"gene_name"</code>
column for storing alternate gene names.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>These functions return a <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object
where the ranges within each of the elements are ordered according
to the following rule:
</p>
<p>When using <code>exonsBy</code> or <code>cdsBy</code> with <code>by = "tx"</code>,
the returned exons or CDS are ordered by ascending rank for each
transcript, that is, by their position in the transcript.
In all other cases, the ranges will be ordered by chromosome, strand,
start, and end values.
</p>


<h3>Value</h3>

<p>A <a href="../../GenomicRanges/html/GRangesList.html">GRangesList</a> object.</p>


<h3>Author(s)</h3>

<p>M. Carlson, P. Aboyoun and H. Pag√®s
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="transcripts.html">transcripts</a></code> and <code><a href="transcriptsByOverlaps.html">transcriptsByOverlaps</a></code>
for more ways to extract genomic features
from a <a href="TxDb-class.html">TxDb</a>-like object.
</p>
</li>
<li> <p><code><a href="transcriptLengths.html">transcriptLengths</a></code> for extracting the transcript
lengths (and other metrics) from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="exonicParts.html">exonicParts</a></code> and <code><a href="exonicParts.html">intronicParts</a></code> for
extracting non-overlapping exonic or intronic parts from a
TxDb-like object.
</p>
</li>
<li> <p><code><a href="extractTranscriptSeqs.html">extractTranscriptSeqs</a></code> for extracting transcript
(or CDS) sequences from chromosome sequences.
</p>
</li>
<li> <p><code><a href="coverageByTranscript.html">coverageByTranscript</a></code> for computing coverage by
transcript (or CDS) of a set of ranges.
</p>
</li>
<li> <p><a href="select-methods.html">select-methods</a> for how to use the
simple &quot;select&quot; interface to extract information from a
<a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="id2name.html">id2name</a></code> for mapping <a href="TxDb-class.html">TxDb</a> internal ids
to external names for a given feature type.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
txdb_file &lt;- system.file("extdata", "hg19_knownGene_sample.sqlite",
                         package="GenomicFeatures")
txdb &lt;- loadDb(txdb_file)

## Get the transcripts grouped by gene:
transcriptsBy(txdb, "gene")

## Get the exons grouped by gene:
exonsBy(txdb, "gene")

## Get the CDS grouped by transcript:
cds_by_tx0 &lt;- cdsBy(txdb, "tx")
## With more informative group names:
cds_by_tx1 &lt;- cdsBy(txdb, "tx", use.names=TRUE)
## Note that 'cds_by_tx1' can also be obtained with:
names(cds_by_tx0) &lt;- id2name(txdb, feature.type="tx")[names(cds_by_tx0)]
stopifnot(identical(cds_by_tx0, cds_by_tx1))

## Get the introns grouped by transcript:
intronsByTranscript(txdb)

## Get the 5' UTRs grouped by transcript:
fiveUTRsByTranscript(txdb)
fiveUTRsByTranscript(txdb, use.names=TRUE)  # more informative group names
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
