<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Extract genomic features from a TxDb-like object based on...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for transcriptsByOverlaps {GenomicFeatures}"><tr><td>transcriptsByOverlaps {GenomicFeatures}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>
Extract genomic features from a TxDb-like object based on their
genomic location
</h2>

<h3>Description</h3>

<p>Generic functions to extract genomic features for specified genomic
locations.
This page documents the methods for <a href="TxDb-class.html">TxDb</a> objects only.
</p>


<h3>Usage</h3>

<pre>
transcriptsByOverlaps(x, ranges,
                      maxgap = -1L, minoverlap = 0L,
                      type = c("any", "start", "end"), ...)
## S4 method for signature 'TxDb'
transcriptsByOverlaps(x, ranges,
                      maxgap = -1L, minoverlap = 0L,
                      type = c("any", "start", "end"),
                      columns = c("tx_id", "tx_name"))

exonsByOverlaps(x, ranges,
                maxgap = -1L, minoverlap = 0L,
                type = c("any", "start", "end"), ...)
## S4 method for signature 'TxDb'
exonsByOverlaps(x, ranges,
                maxgap = -1L, minoverlap = 0L,
                type = c("any", "start", "end"),
                columns = "exon_id")

cdsByOverlaps(x, ranges,
              maxgap = -1L, minoverlap = 0L,
              type = c("any", "start", "end"), ...)
## S4 method for signature 'TxDb'
cdsByOverlaps(x, ranges,
              maxgap = -1L, minoverlap = 0L,
              type = c("any", "start", "end"),
              columns = "cds_id")
</pre>


<h3>Arguments</h3>

  
<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>
<p>A <a href="TxDb-class.html">TxDb</a> object.</p>
</td></tr>
<tr valign="top"><td><code>ranges</code></td>
<td>
<p>A <a href="../../GenomicRanges/html/GRanges.html">GRanges</a> object to restrict the output.</p>
</td></tr>
<tr valign="top"><td><code>maxgap,minoverlap,type</code></td>
<td>

<p>Used in the internal call to <code>findOverlaps()</code> to detect overlaps.
See <code>?<a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code> in the <span class="pkg">IRanges</span> package
for a description of these arguments.
</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>Arguments to be passed to or from methods.</p>
</td></tr>
<tr valign="top"><td><code>columns</code></td>
<td>
<p>Columns to include in the output.
See <code>?<a href="transcripts.html">transcripts</a></code> for the possible values.</p>
</td></tr>
</table>


<h3>Details</h3>

<p>These functions subset the results of <code><a href="transcripts.html">transcripts</a></code>,
<code><a href="transcripts.html">exons</a></code>, and <code><a href="transcripts.html">cds</a></code> function calls with
using the results of <code><a href="../../IRanges/html/findOverlaps.html">findOverlaps</a></code>
calls based on the specified <code>ranges</code>.
</p>


<h3>Value</h3>

<p> a GRanges object </p>


<h3>Author(s)</h3>

<p>P. Aboyoun
</p>


<h3>See Also</h3>


<ul>
<li> <p><code><a href="transcripts.html">transcripts</a></code> and <code><a href="transcriptsBy.html">transcriptsBy</a></code>
for more ways to extract genomic features
from a <a href="TxDb-class.html">TxDb</a>-like object.
</p>
</li>
<li> <p><code><a href="transcriptLengths.html">transcriptLengths</a></code> for extracting the transcript
lengths (and other metrics) from a <a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="exonicParts.html">exonicParts</a></code> and <code><a href="exonicParts.html">intronicParts</a></code> for
extracting non-overlapping exonic or intronic parts from a
TxDb-like object.
</p>
</li>
<li> <p><code><a href="extractTranscriptSeqs.html">extractTranscriptSeqs</a></code> for extracting transcript
(or CDS) sequences from chromosome sequences.
</p>
</li>
<li> <p><code><a href="coverageByTranscript.html">coverageByTranscript</a></code> for computing coverage by
transcript (or CDS) of a set of ranges.
</p>
</li>
<li> <p><a href="select-methods.html">select-methods</a> for how to use the
simple &quot;select&quot; interface to extract information from a
<a href="TxDb-class.html">TxDb</a> object.
</p>
</li>
<li> <p><code><a href="id2name.html">id2name</a></code> for mapping <a href="TxDb-class.html">TxDb</a> internal ids
to external names for a given feature type.
</p>
</li>
<li><p> The <a href="TxDb-class.html">TxDb</a> class.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
txdb &lt;- loadDb(system.file("extdata", "hg19_knownGene_sample.sqlite",
                           package="GenomicFeatures"))
gr &lt;- GRanges(Rle("chr1", 2),
              IRanges(c(500,10500), c(10000,30000)),
              strand = Rle("-", 2))
transcriptsByOverlaps(txdb, gr)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicFeatures</em> version 1.42.3 <a href="00Index.html">Index</a>]</div>
</body></html>
