<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Transform genomic ranges into &quot;absolute&quot; ranges</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for absoluteRanges {GenomicRanges}"><tr><td>absoluteRanges {GenomicRanges}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Transform genomic ranges into &quot;absolute&quot; ranges</h2>

<h3>Description</h3>

<p><code>absoluteRanges</code> transforms the genomic ranges in <code>x</code> into
<em>absolute</em> ranges i.e. into ranges counted from the beginning of
the virtual sequence obtained by concatenating all the sequences in the
underlying genome (in the order reported by <code>seqlevels(x)</code>).
</p>
<p><code>relativeRanges</code> performs the reverse transformation.
</p>
<p>NOTE: These functions only work on <em>small</em> genomes. See Details
section below.
</p>


<h3>Usage</h3>

<pre>
absoluteRanges(x)
relativeRanges(x, seqlengths)

## Related utility:
isSmallGenome(seqlengths)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>For <code>absoluteRanges</code>: a <a href="GRanges-class.html">GenomicRanges</a> object with ranges
defined on a <em>small</em> genome (see Details section below).
</p>
<p>For <code>relativeRanges</code>: an <a href="../../IRanges/html/IntegerRanges-class.html">IntegerRanges</a> object.
</p>
</td></tr>
<tr valign="top"><td><code>seqlengths</code></td>
<td>

<p>An object holding sequence lengths. This can be a named integer
(or numeric) vector with no duplicated names as returned by
<code><a href="../../GenomeInfoDb/html/seqlengths.html">seqlengths</a>()</code>, or any object from
which sequence lengths can be extracted with
<code><a href="../../GenomeInfoDb/html/seqlengths.html">seqlengths</a>()</code>.
</p>
<p>For <code>relativeRanges</code>, <code>seqlengths</code> must describe a <em>small</em>
genome (see Details section below).
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>Because <code>absoluteRanges</code> returns the <em>absolute</em> ranges in an
<a href="../../IRanges/html/IRanges.html">IRanges</a> object, and because an <a href="../../IRanges/html/IRanges.html">IRanges</a>
object cannot hold ranges with an end &gt; <code>.Machine$integer.max</code>
(i.e. &gt;= 2^31 on most platforms), <code>absoluteRanges</code> cannot be used
if the size of the underlying genome (i.e. the total length of the
sequences in it) is &gt; <code>.Machine$integer.max</code>. Utility function
<code>isSmallGenome</code> is provided as a mean for the user to check
upfront whether the genome is <em>small</em> (i.e. its size is &lt;=
<code>.Machine$integer.max</code>) or not, and thus compatible with
<code>absoluteRanges</code> or not.
</p>
<p><code>relativeRanges</code> applies the same restriction by looking at the
<code>seqlengths</code> argument.
</p>


<h3>Value</h3>

<p>An <a href="../../IRanges/html/IRanges.html">IRanges</a> object for <code>absoluteRanges</code>.
</p>
<p>A <a href="GRanges-class.html">GRanges</a> object for <code>relativeRanges</code>.
</p>
<p><code>absoluteRanges</code> and <code>relativeRanges</code> both return an object that
is <em>parallel</em> to the input object (i.e. same length and names).
</p>
<p><code>isSmallGenome</code> returns TRUE if the total length of the underlying
sequences is &lt;= <code>.Machine$integer.max</code> (e.g. Fly genome),
FALSE if not (e.g. Human genome), or NA if it cannot be computed (because
some sequence lengths are NA).
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s
</p>


<h3>See Also</h3>


<ul>
<li> <p><a href="GRanges-class.html">GRanges</a> objects.
</p>
</li>
<li> <p><a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> objects in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><a href="../../GenomeInfoDb/html/Seqinfo-class.html">Seqinfo</a> objects and the <code><a href="../../GenomeInfoDb/html/seqinfo.html">seqlengths</a></code> getter in
the <span class="pkg">GenomeInfoDb</span> package.
</p>
</li>
<li> <p><a href="genomicvars.html">genomicvars</a> for manipulating genomic variables.
</p>
</li>
<li><p> The <code><a href="tileGenome.html">tileGenome</a></code> function for putting tiles on a
genome.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## TOY EXAMPLE
## ---------------------------------------------------------------------

gr &lt;- GRanges(Rle(c("chr2", "chr1", "chr3", "chr1"), 4:1),
              IRanges(1:10, width=5),
              seqinfo=Seqinfo(c("chr1", "chr2", "chr3"), c(100, 50, 20)))

ar &lt;- absoluteRanges(gr)
ar

gr2 &lt;- relativeRanges(ar, seqlengths(gr))
gr2

## Sanity check:
stopifnot(all(gr == gr2))

## ---------------------------------------------------------------------
## ON REAL DATA
## ---------------------------------------------------------------------

## With a "small" genome

library(TxDb.Dmelanogaster.UCSC.dm3.ensGene)
txdb &lt;- TxDb.Dmelanogaster.UCSC.dm3.ensGene
ex &lt;- exons(txdb)
ex

isSmallGenome(ex)

## Note that because isSmallGenome() can return NA (see Value section
## above), its result should always be wrapped inside isTRUE() when
## used in an if statement:
if (isTRUE(isSmallGenome(ex))) {
    ar &lt;- absoluteRanges(ex)
    ar

    ex2 &lt;- relativeRanges(ar, seqlengths(ex))
    ex2  # original strand is not restored

    ## Sanity check:
    strand(ex2) &lt;- strand(ex)  # restore the strand
    stopifnot(all(ex == ex2))
}

## With a "big" genome (but we can reduce it)

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb &lt;- TxDb.Hsapiens.UCSC.hg19.knownGene
ex &lt;- exons(txdb)
isSmallGenome(ex)
## Not run: 
    absoluteRanges(ex)  # error!

## End(Not run)

## However, if we are only interested in some chromosomes, we might
## still be able to use absoluteRanges():
seqlevels(ex, pruning.mode="coarse") &lt;- paste0("chr", 1:10)
isSmallGenome(ex)  # TRUE!
ar &lt;- absoluteRanges(ex)
ex2 &lt;- relativeRanges(ar, seqlengths(ex))

## Sanity check:
strand(ex2) &lt;- strand(ex) 
stopifnot(all(ex == ex2))
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicRanges</em> version 1.42.0 <a href="00Index.html">Index</a>]</div>
</body></html>
