<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Manipulating genomic variables</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for genomicvars {GenomicRanges}"><tr><td>genomicvars {GenomicRanges}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Manipulating genomic variables</h2>

<h3>Description</h3>

<p>A <em>genomic variable</em> is a variable defined along a genome.
Here are 2 ways a genomic variable is generally represented in
Bioconductor:
</p>

<ol>
<li><p> as a named <a href="../../IRanges/html/RleList.html">RleList</a> object with one list element
per chromosome;
</p>
</li>
<li><p> as a metadata column on a <em>disjoint</em> GRanges object.
</p>
</li></ol>

<p>This man page documents tools for switching from one form to the other.
</p>


<h3>Usage</h3>

<pre>
bindAsGRanges(...)
mcolAsRleList(x, varname)
binnedAverage(bins, numvar, varname, na.rm=FALSE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>...</code></td>
<td>

<p>One or more genomic variables in the form of named RleList objects.
</p>
</td></tr>
<tr valign="top"><td><code>x</code></td>
<td>

<p>A <em>disjoint</em> <a href="GRanges-class.html">GRanges</a> object with metadata columns on it.
A <a href="GRanges-class.html">GRanges</a> object is said to be <em>disjoint</em> if it contains
ranges that do not overlap with each other. This can be tested with
<code>isDisjoint</code>. See <code>?`<a href="inter-range-methods.html">isDisjoint,GenomicRanges-method</a>`</code>
for more information about the <code>isDisjoint</code> method for
<a href="GRanges-class.html">GRanges</a> objects.
</p>
</td></tr>
<tr valign="top"><td><code>varname</code></td>
<td>

<p>The name of the genomic variable.
</p>
<p>For <code>mcolAsRleList</code> this must be the name of the metadata column
on <code>x</code> to be turned into an <a href="../../IRanges/html/RleList.html">RleList</a> object.
</p>
<p>For <code>binnedAverage</code> this will be the name of the metadata column
that contains the binned average in the returned object.
</p>
</td></tr>
<tr valign="top"><td><code>bins</code></td>
<td>

<p>A <a href="GRanges-class.html">GRanges</a> object representing the genomic bins. Typically
obtained by calling <code><a href="tileGenome.html">tileGenome</a></code> with
<code>cut.last.tile.in.chrom=TRUE</code>.
</p>
</td></tr>
<tr valign="top"><td><code>numvar</code></td>
<td>

<p>A named <a href="../../IRanges/html/RleList.html">RleList</a> object representing a numerical variable
defined along the genome covered by <code>bins</code> (which is the genome
described by <code>seqinfo(bins)</code>).
</p>
</td></tr>
<tr valign="top"><td><code>na.rm</code></td>
<td>

<p>A logical value indicating whether <code>NA</code> values should be
stripped before the average is computed.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p><code>bindAsGRanges</code> allows to switch the representation of one or
more genomic variables from the <em>named RleList</em> form to the
<em>metadata column on a disjoint GRanges object</em> form by binding
the supplied named RleList objects together and putting them
on the same <a href="GRanges-class.html">GRanges</a> object. This transformation is lossless.
</p>
<p><code>mcolAsRleList</code> performs the opposite transformation and is also
lossless (however the circularity flags and genome information in
<code>seqinfo(x)</code> won't propagate). It works for any metadata column on
<code>x</code> that can be put in <a href="../../S4Vectors/html/Rle.html">Rle</a> form i.e. that is an
atomic vector or a factor.
</p>
<p><code>binnedAverage</code> computes the binned average of a numerical variable
defined along a genome.
</p>


<h3>Value</h3>

<p>For <code>bindAsGRanges</code>: a <a href="GRanges-class.html">GRanges</a> object with 1 metadata column
per supplied genomic variable.
</p>
<p>For <code>mcolAsRleList</code>: a named <a href="../../IRanges/html/RleList.html">RleList</a> object with
1 list element per seqlevel in <code>x</code>.
</p>
<p>For <code>binnedAverage</code>: input <a href="GRanges-class.html">GRanges</a> object <code>bins</code> with
an additional metadata column named <code>varname</code> containing the binned
average.
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s
</p>


<h3>See Also</h3>


<ul>
<li> <p><a href="../../IRanges/html/RleList.html">RleList</a> objects in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><a href="coverage-methods.html">coverage,GenomicRanges-method</a> for computing the coverage
of a <a href="GRanges-class.html">GRanges</a> object.
</p>
</li>
<li><p> The <code><a href="tileGenome.html">tileGenome</a></code> function for putting tiles on a
genome.
</p>
</li>
<li> <p><a href="GRanges-class.html">GRanges</a> objects and <a href="inter-range-methods.html">isDisjoint,GenomicRanges-method</a>
for the <code>isDisjoint</code> method for <a href="GRanges-class.html">GenomicRanges</a> objects.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## ---------------------------------------------------------------------
## A. TWO WAYS TO REPRESENT A GENOMIC VARIABLE
## -----------------------------------------------------------------

## 1) As a named RleList object
## ----------------------------
## Let's create a genomic variable in the "named RleList" form:
library(BSgenome.Scerevisiae.UCSC.sacCer2)
set.seed(55)
my_var &lt;- RleList(
    lapply(seqlengths(Scerevisiae),
        function(seqlen) {
            tmp &lt;- sample(50L, seqlen, replace=TRUE) 
            Rle(cumsum(tmp - rev(tmp)))
        }
    ),
    compress=FALSE)
my_var

## 2) As a metadata column on a disjoint GRanges object
## ----------------------------------------------------
gr1 &lt;- bindAsGRanges(my_var=my_var)
gr1

gr2 &lt;- GRanges(c("chrI:1-150",
                 "chrI:211-285",
                 "chrI:291-377",
                 "chrV:51-60"),
               score=c(0.4, 8, -10, 2.2),
               id=letters[1:4],
               seqinfo=seqinfo(Scerevisiae))
gr2

## Going back to the "named RleList" form:
mcolAsRleList(gr1, "my_var")
score &lt;- mcolAsRleList(gr2, "score")
score
id &lt;- mcolAsRleList(gr2, "id")
id
bindAsGRanges(score=score, id=id)

## Bind 'my_var', 'score', and 'id' together:
gr3 &lt;- bindAsGRanges(my_var=my_var, score=score, id=id)

## Sanity checks:
stopifnot(identical(my_var, mcolAsRleList(gr3, "my_var")))
stopifnot(identical(score, mcolAsRleList(gr3, "score")))
stopifnot(identical(id, mcolAsRleList(gr3, "id")))
gr2b &lt;- bindAsGRanges(score=score, id=id)
seqinfo(gr2b) &lt;- seqinfo(gr2)
stopifnot(identical(gr2, gr2b))

## ---------------------------------------------------------------------
## B. BIND TOGETHER THE COVERAGES OF SEVERAL BAM FILES
## ---------------------------------------------------------------------

library(pasillaBamSubset)
library(GenomicAlignments)
untreated1_cvg &lt;- coverage(BamFile(untreated1_chr4()))
untreated3_cvg &lt;- coverage(BamFile(untreated3_chr4()))
all_cvg &lt;- bindAsGRanges(untreated1=untreated1_cvg,
                         untreated3=untreated3_cvg)

## Keep regions with coverage:
all_cvg[with(mcols(all_cvg), untreated1 + untreated3 &gt;= 1)]

## Plot the coverage profiles with the Gviz package:
library(Gviz)
plotNumvars &lt;- function(numvars, region, name="numvars", ...)
{
    stopifnot(is(numvars, "GRanges"))
    stopifnot(is(region, "GRanges"), length(region) == 1L)
    gtrack &lt;- GenomeAxisTrack()
    dtrack &lt;- DataTrack(numvars,
                        chromosome=as.character(seqnames(region)),
                        name=name,
                        groups=colnames(mcols(numvars)), type="l", ...)
    plotTracks(list(gtrack, dtrack), from=start(region), to=end(region))
}
plotNumvars(all_cvg, GRanges("chr4:1-25000"),
            "coverage", col=c("red", "blue"))
plotNumvars(all_cvg, GRanges("chr4:1.03e6-1.08e6"),
            "coverage", col=c("red", "blue"))

## Sanity checks:
stopifnot(identical(untreated1_cvg, mcolAsRleList(all_cvg, "untreated1")))
stopifnot(identical(untreated3_cvg, mcolAsRleList(all_cvg, "untreated3")))

## ---------------------------------------------------------------------
## C. COMPUTE THE BINNED AVERAGE OF A NUMERICAL VARIABLE DEFINED ALONG A
##    GENOME
## ---------------------------------------------------------------------

## In some applications (e.g. visualization), there is the need to compute
## the average of a genomic variable for a set of predefined fixed-width
## regions (sometimes called "bins").
## Let's use tileGenome() to create such a set of bins:
bins1 &lt;- tileGenome(seqinfo(Scerevisiae), tilewidth=100,
                    cut.last.tile.in.chrom=TRUE)

## Compute the binned average for 'my_var' and 'score':
bins1 &lt;- binnedAverage(bins1, my_var, "binned_var")
bins1
bins1 &lt;- binnedAverage(bins1, score, "binned_score")
bins1

## Binned average in "named RleList" form:
binned_var1 &lt;- mcolAsRleList(bins1, "binned_var")
binned_var1
stopifnot(all.equal(mean(my_var), mean(binned_var1)))  # sanity check

mcolAsRleList(bins1, "binned_score")

## With bigger bins:
bins2 &lt;- tileGenome(seqinfo(Scerevisiae), tilewidth=50000,
                    cut.last.tile.in.chrom=TRUE)
bins2 &lt;- binnedAverage(bins2, my_var, "binned_var")
bins2 &lt;- binnedAverage(bins2, score, "binned_score")
bins2

binned_var2 &lt;- mcolAsRleList(bins2, "binned_var")
binned_var2
stopifnot(all.equal(mean(my_var), mean(binned_var2)))  # sanity check

mcolAsRleList(bins2, "binned_score")

## Not surprisingly, the "binned" variables are much more compact in
## memory than the original variables (they contain much less runs):
object.size(my_var)
object.size(binned_var1)
object.size(binned_var2)

## ---------------------------------------------------------------------
## D. SANITY CHECKS
## ---------------------------------------------------------------------

bins3 &lt;- tileGenome(c(chr1=10, chr2=8), tilewidth=5,
                    cut.last.tile.in.chrom=TRUE)
my_var3 &lt;- RleList(chr1=Rle(c(1:3, NA, 5:7)), chr2=Rle(c(-3, NA, -3, NaN)))
bins3 &lt;- binnedAverage(bins3, my_var3, "binned_var3", na.rm=TRUE)
binned_var3 &lt;- mcols(bins3)$binned_var3
stopifnot(
  identical(mean(my_var3$chr1[1:5], na.rm=TRUE),
            binned_var3[1]),
  identical(mean(c(my_var3$chr1, 0, 0, 0)[6:10], na.rm=TRUE),
            binned_var3[2]),
  #identical(mean(c(my_var3$chr2, 0), na.rm=TRUE),
  #          binned_var3[3]),
  identical(0, binned_var3[4])
)
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicRanges</em> version 1.42.0 <a href="00Index.html">Index</a>]</div>
</body></html>
