<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: GNCList objects</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for GNCList-class {GenomicRanges}"><tr><td>GNCList-class {GenomicRanges}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>GNCList objects</h2>

<h3>Description</h3>

<p>The GNCList class is a container for storing the Nested Containment List
representation of a vector of genomic ranges (typically represented as
a <a href="GRanges-class.html">GRanges</a> object).
To preprocess a <a href="GRanges-class.html">GRanges</a> object, simply call the <code>GNCList</code>
constructor function on it. The resulting GNCList object can then be used
for efficient overlap-based operations on the genomic ranges.
</p>


<h3>Usage</h3>

<pre>
GNCList(x)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>x</code></td>
<td>

<p>The <a href="GRanges-class.html">GRanges</a> (or more generally <a href="GRanges-class.html">GenomicRanges</a>) object to
preprocess.
</p>
</td></tr>
</table>


<h3>Details</h3>

<p>The <span class="pkg">IRanges</span> package also defines the <code><a href="../../IRanges/html/NCList.html">NCList</a></code>
and <code><a href="../../IRanges/html/NCLists.html">NCLists</a></code> constructors and classes for
preprocessing and representing a <a href="../../IRanges/html/IntegerRanges.html">IntegerRanges</a> or
<a href="../../IRanges/html/IntegerRangesList.html">IntegerRangesList</a> object as a data structure based on
Nested Containment Lists.
</p>
<p>Note that GNCList objects (introduced in BioC 3.1) are replacements for
GIntervalTree objects (BioC &lt; 3.1).
</p>
<p>See <code>?<a href="../../IRanges/html/NCList.html">NCList</a></code> in the <span class="pkg">IRanges</span> package for
some important differences between the new algorithm based on Nested
Containment Lists and the old algorithm based on interval trees.
In particular, the new algorithm supports preprocessing of a
<a href="GRanges-class.html">GenomicRanges</a> object with ranges defined on circular sequences
(e.g. on the mitochnodrial chromosome). See below for some examples.
</p>


<h3>Value</h3>

<p>A GNCList object.
</p>


<h3>Author(s)</h3>

<p>H. Pag√®s</p>


<h3>References</h3>

<p>Alexander V. Alekseyenko and Christopher J. Lee &ndash;
Nested Containment List (NCList): a new algorithm for accelerating interval
query of genome alignment and interval databases.
Bioinformatics (2007) 23 (11): 1386-1393.
doi: 10.1093/bioinformatics/btl647
</p>


<h3>See Also</h3>


<ul>
<li><p> The <code><a href="../../IRanges/html/NCList.html">NCList</a></code> and <code><a href="../../IRanges/html/NCLists.html">NCLists</a></code>
constructors and classs defined in the <span class="pkg">IRanges</span> package.
</p>
</li>
<li> <p><code><a href="findOverlaps-methods.html">findOverlaps</a></code> for finding/counting interval overlaps
between two <em>range-based</em> objects.
</p>
</li>
<li> <p><a href="GRanges-class.html">GRanges</a> objects.
</p>
</li></ul>



<h3>Examples</h3>

<pre>
## The examples below are for illustration purpose only and do NOT
## reflect typical usage. This is because, for a one time use, it is
## NOT advised to explicitely preprocess the input for findOverlaps()
## or countOverlaps(). These functions will take care of it and do a
## better job at it (by preprocessing only what's needed when it's
## needed, and release memory as they go).

## ---------------------------------------------------------------------
## PREPROCESS QUERY OR SUBJECT
## ---------------------------------------------------------------------

query &lt;- GRanges(Rle(c("chrM", "chr1", "chrM", "chr1"), 4:1),
                 IRanges(1:10, width=5), strand=rep(c("+", "-"), 5))
subject &lt;- GRanges(Rle(c("chr1", "chr2", "chrM"), 3:1),
                   IRanges(6:1, width=5), strand="+")

## Either the query or the subject of findOverlaps() can be preprocessed:

ppsubject &lt;- GNCList(subject)
hits1a &lt;- findOverlaps(query, ppsubject)
hits1a
hits1b &lt;- findOverlaps(query, ppsubject, ignore.strand=TRUE)
hits1b

ppquery &lt;- GNCList(query)
hits2a &lt;- findOverlaps(ppquery, subject)
hits2a
hits2b &lt;- findOverlaps(ppquery, subject, ignore.strand=TRUE)
hits2b

## Note that 'hits1a' and 'hits2a' contain the same hits but not
## necessarily in the same order.
stopifnot(identical(sort(hits1a), sort(hits2a)))
## Same for 'hits1b' and 'hits2b'.
stopifnot(identical(sort(hits1b), sort(hits2b)))

## ---------------------------------------------------------------------
## WITH CIRCULAR SEQUENCES
## ---------------------------------------------------------------------

seqinfo &lt;- Seqinfo(c("chr1", "chr2", "chrM"),
                   seqlengths=c(100, 50, 10),
                   isCircular=c(FALSE, FALSE, TRUE))
seqinfo(query) &lt;- seqinfo[seqlevels(query)]
seqinfo(subject) &lt;- seqinfo[seqlevels(subject)]

ppsubject &lt;- GNCList(subject)
hits3 &lt;- findOverlaps(query, ppsubject)
hits3

## Circularity introduces more hits:

stopifnot(all(hits1a %in% hits3))
new_hits &lt;- setdiff(hits3, hits1a)
new_hits  # 1 new hit
query[queryHits(new_hits)]
subject[subjectHits(new_hits)]  # positions 11:13 on chrM are the same
                                # as positions 1:3

## Sanity checks:
stopifnot(identical(new_hits, Hits(9, 6, 10, 6, sort.by.query=TRUE)))
ppquery &lt;- GNCList(query)
hits4 &lt;- findOverlaps(ppquery, subject)
stopifnot(identical(sort(hits3), sort(hits4)))
</pre>

<hr /><div style="text-align: center;">[Package <em>GenomicRanges</em> version 1.42.0 <a href="00Index.html">Index</a>]</div>
</body></html>
